{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse patient A7 data\n",
    "\n",
    "## First use the paper's reported clusters (from SciClone) to generate input files \n",
    "Adapted from MACHNA (El-Kebir et. al.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "confidence=0.95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_id(line):\n",
    "    return \"_\".join(map(str, [line['chromosome_name'], line['start'], line['stop']]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome_name', 'start', 'stop', 'reference', 'variant', 'type',\n",
       "       'gene_name', 'transcript_name', 'transcript_species',\n",
       "       'transcript_source', 'transcript_version', 'strand',\n",
       "       'transcript_status', 'trv_type', 'c_position', 'amino_acid_change',\n",
       "       'ucsc_cons', 'domain', 'all_domains', 'deletion_substructures',\n",
       "       'transcript_error', 'brain.rcnt.llr3_ref', 'brain.rcnt.llr3_var',\n",
       "       'brain.rcnt.llr3_VAF', 'kidney.rcnt.llr3_ref', 'kidney.rcnt.llr3_var',\n",
       "       'kidney.rcnt.llr3_VAF', 'liver.rcnt.llr3_ref', 'liver.rcnt.llr3_var',\n",
       "       'liver.rcnt.llr3_VAF', 'lung.rcnt.llr3_ref', 'lung.rcnt.llr3_var',\n",
       "       'lung.rcnt.llr3_VAF', 'rib.rcnt.llr3_ref', 'rib.rcnt.llr3_var',\n",
       "       'rib.rcnt.llr3_VAF', 'tumor.rcnt.llr3_ref', 'tumor.rcnt.llr3_var',\n",
       "       'tumor.rcnt.llr3_VAF', 'cluster'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "REPO_DIR = os.path.join(os.getcwd(), \"../\") \n",
    "HOADLEY_DATA_DIR = os.path.join(REPO_DIR, 'data', 'hoadley_breast_cancer_2016/')\n",
    "table = pd.read_table(os.path.join(HOADLEY_DATA_DIR,\"A7/A7_raw.tsv\"))\n",
    "table['id']=table.apply(get_id, axis=1)\n",
    "table = table.set_index('id')\n",
    "table.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reassign clusters of mutations 7_12163423_12163423 and 7_57562948_57562948\n",
    "\n",
    "#### details for why in MACHINA paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "table.loc['7_12163423_12163423', 'cluster']=2\n",
    "table.loc['7_57562948_57562948', 'cluster']=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def label_snv(row):\n",
    "    label = []\n",
    "    if row['gene_name'] != \"-\":\n",
    "        label.append(row['gene_name'])\n",
    "    label += [str(row['chromosome_name']), str(row['start']), str(row['stop'])]\n",
    "    return (\":\").join(label)\n",
    "table['character_label'] = table.apply(lambda row: label_snv(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_species</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>...</th>\n",
       "      <th>lung.rcnt.llr3_var</th>\n",
       "      <th>lung.rcnt.llr3_VAF</th>\n",
       "      <th>rib.rcnt.llr3_ref</th>\n",
       "      <th>rib.rcnt.llr3_var</th>\n",
       "      <th>rib.rcnt.llr3_VAF</th>\n",
       "      <th>tumor.rcnt.llr3_ref</th>\n",
       "      <th>tumor.rcnt.llr3_var</th>\n",
       "      <th>tumor.rcnt.llr3_VAF</th>\n",
       "      <th>cluster</th>\n",
       "      <th>character_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_2866642_2866642</th>\n",
       "      <td>1</td>\n",
       "      <td>2866642</td>\n",
       "      <td>2866642</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENSG00000177133</td>\n",
       "      <td>ENST00000321399</td>\n",
       "      <td>human</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>...</td>\n",
       "      <td>57</td>\n",
       "      <td>33.33</td>\n",
       "      <td>102</td>\n",
       "      <td>85</td>\n",
       "      <td>45.45</td>\n",
       "      <td>125</td>\n",
       "      <td>42</td>\n",
       "      <td>25.15</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000177133:1:2866642:2866642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_2984058_2984058</th>\n",
       "      <td>1</td>\n",
       "      <td>2984058</td>\n",
       "      <td>2984058</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>NM_022114.2</td>\n",
       "      <td>human</td>\n",
       "      <td>genbank</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.56</td>\n",
       "      <td>212</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>205</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>PRDM16:1:2984058:2984058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_3814510_3814510</th>\n",
       "      <td>1</td>\n",
       "      <td>3814510</td>\n",
       "      <td>3814510</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C1orf174</td>\n",
       "      <td>NM_207356.2</td>\n",
       "      <td>human</td>\n",
       "      <td>genbank</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1.96</td>\n",
       "      <td>82</td>\n",
       "      <td>58</td>\n",
       "      <td>41.43</td>\n",
       "      <td>170</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6</td>\n",
       "      <td>C1orf174:1:3814510:3814510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_4293949_4293949</th>\n",
       "      <td>1</td>\n",
       "      <td>4293949</td>\n",
       "      <td>4293949</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>28.12</td>\n",
       "      <td>51</td>\n",
       "      <td>45</td>\n",
       "      <td>46.88</td>\n",
       "      <td>100</td>\n",
       "      <td>45</td>\n",
       "      <td>31.03</td>\n",
       "      <td>1</td>\n",
       "      <td>1:4293949:4293949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_4367117_4367117</th>\n",
       "      <td>1</td>\n",
       "      <td>4367117</td>\n",
       "      <td>4367117</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>179</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>187</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9</td>\n",
       "      <td>1:4367117:4367117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chromosome_name    start     stop reference variant type  \\\n",
       "id                                                                            \n",
       "1_2866642_2866642                1  2866642  2866642         G       A  SNP   \n",
       "1_2984058_2984058                1  2984058  2984058         G       A  SNP   \n",
       "1_3814510_3814510                1  3814510  3814510         A       C  SNP   \n",
       "1_4293949_4293949                1  4293949  4293949         G       T  SNP   \n",
       "1_4367117_4367117                1  4367117  4367117         A       C  SNP   \n",
       "\n",
       "                         gene_name  transcript_name transcript_species  \\\n",
       "id                                                                       \n",
       "1_2866642_2866642  ENSG00000177133  ENST00000321399              human   \n",
       "1_2984058_2984058           PRDM16      NM_022114.2              human   \n",
       "1_3814510_3814510         C1orf174      NM_207356.2              human   \n",
       "1_4293949_4293949                -                -                  -   \n",
       "1_4367117_4367117                -                -                  -   \n",
       "\n",
       "                  transcript_source  ... lung.rcnt.llr3_var  \\\n",
       "id                                   ...                      \n",
       "1_2866642_2866642           ensembl  ...                 57   \n",
       "1_2984058_2984058           genbank  ...                  1   \n",
       "1_3814510_3814510           genbank  ...                  3   \n",
       "1_4293949_4293949                 -  ...                 27   \n",
       "1_4367117_4367117                 -  ...                  0   \n",
       "\n",
       "                  lung.rcnt.llr3_VAF rib.rcnt.llr3_ref rib.rcnt.llr3_var  \\\n",
       "id                                                                         \n",
       "1_2866642_2866642              33.33               102                85   \n",
       "1_2984058_2984058               0.56               212                 0   \n",
       "1_3814510_3814510               1.96                82                58   \n",
       "1_4293949_4293949              28.12                51                45   \n",
       "1_4367117_4367117               0.00               179                 0   \n",
       "\n",
       "                  rib.rcnt.llr3_VAF tumor.rcnt.llr3_ref tumor.rcnt.llr3_var  \\\n",
       "id                                                                            \n",
       "1_2866642_2866642             45.45                 125                  42   \n",
       "1_2984058_2984058              0.00                 205                   0   \n",
       "1_3814510_3814510             41.43                 170                   0   \n",
       "1_4293949_4293949             46.88                 100                  45   \n",
       "1_4367117_4367117              0.00                 187                   0   \n",
       "\n",
       "                  tumor.rcnt.llr3_VAF cluster  \\\n",
       "id                                              \n",
       "1_2866642_2866642               25.15       1   \n",
       "1_2984058_2984058                0.00       3   \n",
       "1_3814510_3814510                0.00       6   \n",
       "1_4293949_4293949               31.03       1   \n",
       "1_4367117_4367117                0.00       9   \n",
       "\n",
       "                                     character_label  \n",
       "id                                                    \n",
       "1_2866642_2866642  ENSG00000177133:1:2866642:2866642  \n",
       "1_2984058_2984058           PRDM16:1:2984058:2984058  \n",
       "1_3814510_3814510         C1orf174:1:3814510:3814510  \n",
       "1_4293949_4293949                  1:4293949:4293949  \n",
       "1_4367117_4367117                  1:4367117:4367117  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_table = table.copy()\n",
    "raw_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate migration history input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_cols = ['tumor.rcnt.llr3_ref','brain.rcnt.llr3_ref', 'kidney.rcnt.llr3_ref', 'liver.rcnt.llr3_ref', 'lung.rcnt.llr3_ref', 'rib.rcnt.llr3_ref']\n",
    "var_cols = ['tumor.rcnt.llr3_var','brain.rcnt.llr3_var', 'kidney.rcnt.llr3_var', 'liver.rcnt.llr3_var', 'lung.rcnt.llr3_var', 'rib.rcnt.llr3_var']\n",
    "\n",
    "cols = ['breast', 'brain', 'kidney', 'liver', 'lung', 'rib']\n",
    "table = table[['cluster']+ref_cols+var_cols]\n",
    "table.columns = ['cluster']+['ref-'+c for c in cols] + ['var-'+c for c in cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9991666666666666\n"
     ]
    }
   ],
   "source": [
    "ctable = table.groupby('cluster').sum()\n",
    "ctable.head()\n",
    "\n",
    "global corrected_confidence\n",
    "nsamples = len([c for c in ctable.columns if c.startswith('ref')])\n",
    "nclusters = len(ctable)\n",
    "corrected_confidence = 1-((1.-confidence)/(nsamples*nclusters))\n",
    "print(corrected_confidence)\n",
    "\n",
    "assert(corrected_confidence > confidence)\n",
    "assert(corrected_confidence < 1.0)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "from scipy.stats import beta\n",
    "from scipy.stats import norm\n",
    "\n",
    "def binomial_hpdr(n, N, pct, a=1, b=1, n_pbins=1e3):\n",
    "    \"\"\"\n",
    "    Function computes the posterior mode along with the upper and lower bounds of the\n",
    "    **Highest Posterior Density Region**.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n: number of successes \n",
    "    N: sample size \n",
    "    pct: the size of the confidence interval (between 0 and 1)\n",
    "    a: the alpha hyper-parameter for the Beta distribution used as a prior (Default=1)\n",
    "    b: the beta hyper-parameter for the Beta distribution used as a prior (Default=1)\n",
    "    n_pbins: the number of bins to segment the p_range into (Default=1e3)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    A tuple that contains the mode as well as the lower and upper bounds of the interval\n",
    "    (mode, lower, upper)\n",
    "\n",
    "    \"\"\"\n",
    "    # fixed random variable object for posterior Beta distribution\n",
    "    rv = beta(n+a, N-n+b)\n",
    "    # determine the mode and standard deviation of the posterior\n",
    "    stdev = rv.stats('v')**0.5\n",
    "    mode = (n+a-1.)/(N+a+b-2.)\n",
    "    # compute the number of sigma that corresponds to this confidence\n",
    "    # this is used to set the rough range of possible success probabilities\n",
    "    n_sigma = numpy.ceil(norm.ppf( (1+pct)/2. ))+1\n",
    "    # set the min and max values for success probability \n",
    "    max_p = mode + n_sigma * stdev\n",
    "    if max_p > 1:\n",
    "        max_p = 1.\n",
    "    min_p = mode - n_sigma * stdev\n",
    "    if min_p > 1:\n",
    "        min_p = 1.\n",
    "    # make the range of success probabilities\n",
    "    p_range = numpy.linspace(min_p, max_p, int(n_pbins+1))\n",
    "    # construct the probability mass function over the given range\n",
    "    if mode > 0.5:\n",
    "        sf = rv.sf(p_range)\n",
    "        pmf = sf[:-1] - sf[1:]\n",
    "    else:\n",
    "        cdf = rv.cdf(p_range)\n",
    "        pmf = cdf[1:] - cdf[:-1]\n",
    "    # find the upper and lower bounds of the interval \n",
    "    sorted_idxs = numpy.argsort( pmf )[::-1]\n",
    "    cumsum = numpy.cumsum( numpy.sort(pmf)[::-1] )\n",
    "    j = numpy.argmin( numpy.abs(cumsum - pct) )\n",
    "    upper = p_range[ (sorted_idxs[:j+1]).max()+1 ]\n",
    "    lower = p_range[ (sorted_idxs[:j+1]).min() ]    \n",
    "\n",
    "    return (mode, lower, upper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### \n",
    "def get_ub(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    return v[2]\n",
    "    \n",
    "\n",
    "def get_lb(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[1]\n",
    "    #if mval < 0.01: mval = 0\n",
    "    return mval\n",
    "\n",
    "def get_mean(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[0]\n",
    "    return mval\n",
    "\n",
    "ctable = table.groupby('cluster').sum()\n",
    "for sam in cols:\n",
    "    ctable['ub-'+sam]= ctable.apply(get_ub, args=[sam], axis=1)\n",
    "    ctable['lb-'+sam]= ctable.apply(get_lb, args=[sam], axis=1)\n",
    "    ctable[sam]= ctable.apply(get_mean, args=[sam], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ub(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    return v[2]\n",
    "    \n",
    "\n",
    "def get_lb(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[1]\n",
    "    if mval < 0.01: mval = 0\n",
    "    return mval\n",
    "\n",
    "def get_mean(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[0]\n",
    "    return mval\n",
    "\n",
    "ctable_cutoff = table.groupby('cluster').sum()\n",
    "for sam in cols:\n",
    "    ctable_cutoff['ub-'+sam]= ctable.apply(get_ub, args=[sam], axis=1)\n",
    "    ctable_cutoff['lb-'+sam]= ctable.apply(get_lb, args=[sam], axis=1)\n",
    "    ctable_cutoff[sam]= ctable.apply(get_mean, args=[sam], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vaf(row, sam):\n",
    "    return float(row['var-'+sam])/float(row['var-'+sam]+row['ref-'+sam])\n",
    "\n",
    "vafs = pd.DataFrame()\n",
    "for sam in cols:\n",
    "    vafs[sam] = table.apply(get_vaf, args=[sam], axis=1)\n",
    "vafs['cluster'] = table['cluster']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref-breast</th>\n",
       "      <th>ref-brain</th>\n",
       "      <th>ref-kidney</th>\n",
       "      <th>ref-liver</th>\n",
       "      <th>ref-lung</th>\n",
       "      <th>ref-rib</th>\n",
       "      <th>var-breast</th>\n",
       "      <th>var-brain</th>\n",
       "      <th>var-kidney</th>\n",
       "      <th>var-liver</th>\n",
       "      <th>...</th>\n",
       "      <th>ub-liver</th>\n",
       "      <th>lb-liver</th>\n",
       "      <th>liver</th>\n",
       "      <th>ub-lung</th>\n",
       "      <th>lb-lung</th>\n",
       "      <th>lung</th>\n",
       "      <th>ub-rib</th>\n",
       "      <th>lb-rib</th>\n",
       "      <th>rib</th>\n",
       "      <th>character_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19957</td>\n",
       "      <td>24490</td>\n",
       "      <td>34894</td>\n",
       "      <td>17945</td>\n",
       "      <td>17201</td>\n",
       "      <td>18757</td>\n",
       "      <td>7090</td>\n",
       "      <td>11892</td>\n",
       "      <td>9405</td>\n",
       "      <td>12987</td>\n",
       "      <td>...</td>\n",
       "      <td>0.429257</td>\n",
       "      <td>0.410512</td>\n",
       "      <td>0.419856</td>\n",
       "      <td>0.357974</td>\n",
       "      <td>0.338386</td>\n",
       "      <td>0.348151</td>\n",
       "      <td>0.374843</td>\n",
       "      <td>0.356134</td>\n",
       "      <td>0.365460</td>\n",
       "      <td>ENSG00000177133:1:2866642:2866642_1:4293949:42...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2833</td>\n",
       "      <td>3627</td>\n",
       "      <td>3483</td>\n",
       "      <td>1747</td>\n",
       "      <td>2649</td>\n",
       "      <td>1862</td>\n",
       "      <td>5</td>\n",
       "      <td>39</td>\n",
       "      <td>935</td>\n",
       "      <td>1409</td>\n",
       "      <td>...</td>\n",
       "      <td>0.476082</td>\n",
       "      <td>0.416998</td>\n",
       "      <td>0.446451</td>\n",
       "      <td>0.028580</td>\n",
       "      <td>0.011115</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.387565</td>\n",
       "      <td>0.328122</td>\n",
       "      <td>0.357488</td>\n",
       "      <td>ENSG00000189015:1:28295035:28295035_1:56623876...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5860</td>\n",
       "      <td>7725</td>\n",
       "      <td>7242</td>\n",
       "      <td>3535</td>\n",
       "      <td>5652</td>\n",
       "      <td>6225</td>\n",
       "      <td>4</td>\n",
       "      <td>68</td>\n",
       "      <td>1931</td>\n",
       "      <td>2844</td>\n",
       "      <td>...</td>\n",
       "      <td>0.466682</td>\n",
       "      <td>0.425119</td>\n",
       "      <td>0.445838</td>\n",
       "      <td>0.002973</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000884</td>\n",
       "      <td>0.001296</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000161</td>\n",
       "      <td>PRDM16:1:2984058:2984058_FHAD1:1:15514080:1551...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6681</td>\n",
       "      <td>5938</td>\n",
       "      <td>10865</td>\n",
       "      <td>7874</td>\n",
       "      <td>4371</td>\n",
       "      <td>7259</td>\n",
       "      <td>5</td>\n",
       "      <td>2826</td>\n",
       "      <td>13</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000508</td>\n",
       "      <td>0.345283</td>\n",
       "      <td>0.306408</td>\n",
       "      <td>0.325671</td>\n",
       "      <td>0.003190</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001238</td>\n",
       "      <td>1:5200289:5200289_KCNAB2:1:6078316:6078316_PRD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1601</td>\n",
       "      <td>2064</td>\n",
       "      <td>1894</td>\n",
       "      <td>996</td>\n",
       "      <td>1483</td>\n",
       "      <td>1620</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>490</td>\n",
       "      <td>631</td>\n",
       "      <td>...</td>\n",
       "      <td>0.428625</td>\n",
       "      <td>0.348122</td>\n",
       "      <td>0.387830</td>\n",
       "      <td>0.005428</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000674</td>\n",
       "      <td>0.004970</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>PRDM2:1:14012851:14012851_ZSCAN20:1:33685697:3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>15478</td>\n",
       "      <td>20498</td>\n",
       "      <td>24541</td>\n",
       "      <td>18141</td>\n",
       "      <td>14737</td>\n",
       "      <td>10482</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000551</td>\n",
       "      <td>0.017513</td>\n",
       "      <td>0.011068</td>\n",
       "      <td>0.014050</td>\n",
       "      <td>0.356147</td>\n",
       "      <td>0.331045</td>\n",
       "      <td>0.343521</td>\n",
       "      <td>C1orf174:1:3814510:3814510_1:5461427:5461427_K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1786</td>\n",
       "      <td>2503</td>\n",
       "      <td>3143</td>\n",
       "      <td>2338</td>\n",
       "      <td>1189</td>\n",
       "      <td>1911</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.005545</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001282</td>\n",
       "      <td>0.331037</td>\n",
       "      <td>0.257056</td>\n",
       "      <td>0.293103</td>\n",
       "      <td>0.005563</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001045</td>\n",
       "      <td>ENSG00000219178:1:29727073:29727073_ENSG000002...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3346</td>\n",
       "      <td>4457</td>\n",
       "      <td>4158</td>\n",
       "      <td>4056</td>\n",
       "      <td>3258</td>\n",
       "      <td>3556</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1105</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000739</td>\n",
       "      <td>0.003267</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000613</td>\n",
       "      <td>0.002267</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>AADACL3:1:12659643:12659643_KIAA1026:1:1516176...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>13466</td>\n",
       "      <td>18097</td>\n",
       "      <td>21928</td>\n",
       "      <td>12496</td>\n",
       "      <td>13187</td>\n",
       "      <td>14786</td>\n",
       "      <td>5</td>\n",
       "      <td>28</td>\n",
       "      <td>12</td>\n",
       "      <td>2886</td>\n",
       "      <td>...</td>\n",
       "      <td>0.198293</td>\n",
       "      <td>0.177266</td>\n",
       "      <td>0.187622</td>\n",
       "      <td>0.001145</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>0.001022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000270</td>\n",
       "      <td>1:4367117:4367117_CLCN6:1:11799962:11799962_PA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>915</td>\n",
       "      <td>1029</td>\n",
       "      <td>1394</td>\n",
       "      <td>1015</td>\n",
       "      <td>855</td>\n",
       "      <td>903</td>\n",
       "      <td>0</td>\n",
       "      <td>227</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004912</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005519</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AGBL4:1:49111131:49111131_AGBL4:1:49210849:492...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ref-breast  ref-brain  ref-kidney  ref-liver  ref-lung  ref-rib  \\\n",
       "cluster                                                                    \n",
       "1             19957      24490       34894      17945     17201    18757   \n",
       "2              2833       3627        3483       1747      2649     1862   \n",
       "3              5860       7725        7242       3535      5652     6225   \n",
       "4              6681       5938       10865       7874      4371     7259   \n",
       "5              1601       2064        1894        996      1483     1620   \n",
       "6             15478      20498       24541      18141     14737    10482   \n",
       "7              1786       2503        3143       2338      1189     1911   \n",
       "8              3346       4457        4158       4056      3258     3556   \n",
       "9             13466      18097       21928      12496     13187    14786   \n",
       "10              915       1029        1394       1015       855      903   \n",
       "\n",
       "         var-breast  var-brain  var-kidney  var-liver  ...  ub-liver  \\\n",
       "cluster                                                ...             \n",
       "1              7090      11892        9405      12987  ...  0.429257   \n",
       "2                 5         39         935       1409  ...  0.476082   \n",
       "3                 4         68        1931       2844  ...  0.466682   \n",
       "4                 5       2826          13          4  ...  0.001918   \n",
       "5                 1          0         490        631  ...  0.428625   \n",
       "6                 7          5          17         10  ...  0.001362   \n",
       "7                 0          3           2          3  ...  0.005545   \n",
       "8                 4          1        1105          3  ...  0.003200   \n",
       "9                 5         28          12       2886  ...  0.198293   \n",
       "10                0        227           1          0  ...  0.004912   \n",
       "\n",
       "         lb-liver     liver   ub-lung   lb-lung      lung    ub-rib    lb-rib  \\\n",
       "cluster                                                                         \n",
       "1        0.410512  0.419856  0.357974  0.338386  0.348151  0.374843  0.356134   \n",
       "2        0.416998  0.446451  0.028580  0.011115  0.018525  0.387565  0.328122   \n",
       "3        0.425119  0.445838  0.002973  0.000000  0.000884  0.001296  0.000000   \n",
       "4        0.000000  0.000508  0.345283  0.306408  0.325671  0.003190  0.000000   \n",
       "5        0.348122  0.387830  0.005428  0.000000  0.000674  0.004970  0.000000   \n",
       "6        0.000000  0.000551  0.017513  0.011068  0.014050  0.356147  0.331045   \n",
       "7        0.000000  0.001282  0.331037  0.257056  0.293103  0.005563  0.000000   \n",
       "8        0.000000  0.000739  0.003267  0.000000  0.000613  0.002267  0.000000   \n",
       "9        0.177266  0.187622  0.001145  0.000000  0.000303  0.001022  0.000000   \n",
       "10       0.000000  0.000000  0.005828  0.000000  0.000000  0.005519  0.000000   \n",
       "\n",
       "              rib                                    character_label  \n",
       "cluster                                                               \n",
       "1        0.365460  ENSG00000177133:1:2866642:2866642_1:4293949:42...  \n",
       "2        0.357488  ENSG00000189015:1:28295035:28295035_1:56623876...  \n",
       "3        0.000161  PRDM16:1:2984058:2984058_FHAD1:1:15514080:1551...  \n",
       "4        0.001238  1:5200289:5200289_KCNAB2:1:6078316:6078316_PRD...  \n",
       "5        0.000617  PRDM2:1:14012851:14012851_ZSCAN20:1:33685697:3...  \n",
       "6        0.343521  C1orf174:1:3814510:3814510_1:5461427:5461427_K...  \n",
       "7        0.001045  ENSG00000219178:1:29727073:29727073_ENSG000002...  \n",
       "8        0.000281  AADACL3:1:12659643:12659643_KIAA1026:1:1516176...  \n",
       "9        0.000270  1:4367117:4367117_CLCN6:1:11799962:11799962_PA...  \n",
       "10       0.000000  AGBL4:1:49111131:49111131_AGBL4:1:49210849:492...  \n",
       "\n",
       "[10 rows x 31 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def add_char_label(row): \n",
    "    return \"_\".join(list(raw_table[raw_table['cluster']==row.name]['character_label']))\n",
    "\n",
    "ctable_cutoff['character_label'] = ctable_cutoff.apply(lambda row: add_char_label(row), axis=1)\n",
    "ctable_cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "rows = [\"6 #anatomical sites\\n6 #samples\\n10 #mutations\\n#sample_index\\tsample_label\\tanatomical_site_index\\tanatomical_site_label\\tcharacter_index\\tcharacter_label\\tf_lb\\tf_ub\\tref\\tvar\\n\",]\n",
    "def print_char(row, sam):\n",
    "    return \"\\t\".join(map(str,[i, sam, i, sam, row.name-1, row[\"character_label\"], max(row['lb-'+sam] * 2, 0), min(1, 2 * row['ub-'+sam]), int(row['ref-'+sam]), int(row['var-'+sam])]))+\"\\n\"\n",
    "\n",
    "for i, sam in enumerate(cols):\n",
    "    rows += list(ctable_cutoff.apply(print_char, args=[sam], axis=1))\n",
    "\n",
    "with open(os.path.join(HOADLEY_DATA_DIR,\"A7/A7_\"+str(confidence)+\".tsv\"), 'w') as f:\n",
    "    for line in rows:\n",
    "        f.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prep inputs from raw data for PyClone clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "pyclone_dir = os.path.join(HOADLEY_DATA_DIR, \"pyclone_analysis\", \"A7\")\n",
    "if not os.path.exists(pyclone_dir): \n",
    "    os.makedirs(pyclone_dir)\n",
    "    \n",
    "raw_sample_names = ['tumor.rcnt.llr3','brain.rcnt.llr3', 'kidney.rcnt.llr3', 'liver.rcnt.llr3', 'lung.rcnt.llr3', 'rib.rcnt.llr3']\n",
    "final_sample_names = ['breast', 'brain', 'kidney', 'liver', 'lung', 'rib']\n",
    "\n",
    "# put df in format used in PyClone, with columns:\n",
    "#'mutation_id', 'ref_counts','var_counts', 'normal_cn', 'minor_cn', 'major_cn'\n",
    "for raw_sample, final_sample in zip(raw_sample_names, final_sample_names):\n",
    "    data = []\n",
    "    for i, row in raw_table.iterrows():\n",
    "        data.append([row['character_label'], row[f'{raw_sample}_ref'], row[f'{raw_sample}_var'], 2, 1, 1])\n",
    "    sample_subset = pd.DataFrame(data, columns=['mutation_id', 'ref_counts','var_counts', 'normal_cn', 'minor_cn', 'major_cn'])\n",
    "    sample_subset.to_csv(os.path.join(pyclone_dir,  f\"A7_{final_sample}.tsv\"), index=False, sep=\"\\t\")    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prep inputs for orchard tree inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CUDA GPU: False\n"
     ]
    }
   ],
   "source": [
    "from metient.util import data_extraction_util as dutil\n",
    "import json\n",
    "cluster_id_to_mut_names, mutation_names = dutil.load_pyclone_clusters(os.path.join(pyclone_dir, \"tables\", \"loci.tsv\"))\n",
    "pairtree_dir = os.path.join(HOADLEY_DATA_DIR, \"orchard_trees\")\n",
    "header = [\"id\", \"name\", \"var_reads\", \"total_reads\", \"var_read_prob\"]\n",
    "\n",
    "mut_name_to_mut_id = {}\n",
    "with open(os.path.join(pairtree_dir, f\"A7.ssm\"), \"w\") as f:\n",
    "    f.write(\"\\t\".join(header))\n",
    "    f.write(\"\\n\")  \n",
    "    for i, mut in enumerate(mutation_names):\n",
    "        mut_name_to_mut_id[mut] = f\"m{i}\"\n",
    "        row = [f\"m{i}\", mut]\n",
    "        \n",
    "        mut_row = raw_table[raw_table['character_label']==mut]\n",
    "        var_reads = []\n",
    "        total_reads = []\n",
    "        var_read_probs = []\n",
    "        for sample in raw_sample_names:\n",
    "            var = mut_row[f'{sample}_var'].item()\n",
    "            ref = mut_row[f'{sample}_ref'].item()\n",
    "            var_reads.append(str(var))\n",
    "            total_reads.append(str(var+ref))\n",
    "            var_read_probs.append(str(0.5))\n",
    "\n",
    "        row += [\",\".join(var_reads), \",\".join(total_reads), \",\".join(var_read_probs)]\n",
    "        f.write(\"\\t\".join(row))\n",
    "        f.write(\"\\n\")\n",
    "json_data = {\"samples\": final_sample_names, \"clusters\": [], \"garbage\": []}\n",
    "for x in range(0,len(cluster_id_to_mut_names)):\n",
    "    json_data[\"clusters\"].append([mut_name_to_mut_id[t] for t in cluster_id_to_mut_names[x]])\n",
    "\n",
    "with open(os.path.join(pairtree_dir, f\"A7.params.json\"), 'w', encoding='utf-8') as f:\n",
    "    json.dump(json_data, f, ensure_ascii=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare data for migration history inference (tsvs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need a tsv for each patient with ['anatomical_site_index','anatomical_site_label', 'character_index', 'character_label', 'ref', 'var']\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "cols = ['anatomical_site_index','anatomical_site_label', 'character_index', 'character_label', 'ref', 'var', 'var_read_prob', 'site_category']\n",
    "\n",
    "mut_name_to_clstr_id = {}\n",
    "clstr_id_to_name = {}\n",
    "for cid in cluster_id_to_mut_names:\n",
    "    for mut in cluster_id_to_mut_names[cid]:\n",
    "        mut_name_to_clstr_id[mut] = cid\n",
    "    clstr_id_to_name[cid] = \";\".join(cluster_id_to_mut_names[cid])\n",
    "\n",
    "data = []\n",
    "for midx, mut in enumerate(mutation_names):\n",
    "\n",
    "    mut_row = raw_table[raw_table['character_label']==mut]\n",
    "    for sidx, sample in enumerate(raw_sample_names):\n",
    "        var = mut_row[f'{sample}_var'].item()\n",
    "        ref = mut_row[f'{sample}_ref'].item()\n",
    "        var_read_probs.append(str(0.5))\n",
    "        site_category = 'primary' if final_sample_names[sidx] == 'breast' else 'metastasis'\n",
    "        data.append([sidx, final_sample_names[sidx], midx, mut, ref, var, 0.5, site_category])\n",
    "            \n",
    "patient_df = pd.DataFrame(data, columns=cols)\n",
    "\n",
    "dutil.write_pooled_tsv_from_clusters(patient_df, mut_name_to_clstr_id, clstr_id_to_name, {}, \n",
    "                                     os.path.join(HOADLEY_DATA_DIR,\"pyclone_clustered_tsvs\"), \"A7\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add extra required fields\n",
    "tsv_fn = os.path.join(HOADLEY_DATA_DIR,\"pyclone_clustered_tsvs\", f\"A7_clustered_SNVs.tsv\")\n",
    "df = pd.read_csv(tsv_fn, sep=\"\\t\")\n",
    "df['num_mutations'] = df.apply(lambda row: len(row['character_label'].split(\";\")), axis=1)\n",
    "df['full_label'] = df['character_label']\n",
    "df['character_label'] = df.apply(lambda row:dutil.get_pruned_mut_label(row['character_label'], \";\", \":\"), axis=1)\n",
    "df.to_csv(tsv_fn, sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:met]",
   "language": "python",
   "name": "conda-env-met-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
