{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb419e5e",
   "metadata": {},
   "source": [
    "### NOTE: this was done for validation, we use the McPherson published trees/clonal presences to "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "04e0eccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "repo_dir = os.path.join(os.getcwd(), \"../../\")\n",
    "data_dir = os.path.join(repo_dir, 'data', 'mcpherson_ovarian_2016')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d36464b",
   "metadata": {},
   "source": [
    "## (1) Read in data from McPherson et.al. supplemental tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee814dd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient IDs: [ 1  2  3  4  7  9 10]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>malignant</th>\n",
       "      <th>primer_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref_counts</th>\n",
       "      <th>alt_counts</th>\n",
       "      <th>depth</th>\n",
       "      <th>alt_freq</th>\n",
       "      <th>background_average_alt_freq</th>\n",
       "      <th>ref_p_value</th>\n",
       "      <th>alt_p_value</th>\n",
       "      <th>status</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snpeff_impact</th>\n",
       "      <th>character_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>normal_blood</td>\n",
       "      <td>1</td>\n",
       "      <td>no</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>641</td>\n",
       "      <td>0</td>\n",
       "      <td>641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>normal_sample</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>omentum_site_1</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>449</td>\n",
       "      <td>0</td>\n",
       "      <td>449</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>395</td>\n",
       "      <td>64</td>\n",
       "      <td>459</td>\n",
       "      <td>0.139434</td>\n",
       "      <td>0.002375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.800000e-90</td>\n",
       "      <td>somatic</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>416</td>\n",
       "      <td>0.002404</td>\n",
       "      <td>0.002754</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.825260e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>609</td>\n",
       "      <td>4</td>\n",
       "      <td>613</td>\n",
       "      <td>0.006525</td>\n",
       "      <td>0.002145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.440600e-02</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15293</th>\n",
       "      <td>normal_blood</td>\n",
       "      <td>10</td>\n",
       "      <td>no</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3669</td>\n",
       "      <td>10</td>\n",
       "      <td>3679</td>\n",
       "      <td>0.002718</td>\n",
       "      <td>0.002239</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.130860e-01</td>\n",
       "      <td>normal_sample</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15294</th>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2037</td>\n",
       "      <td>295</td>\n",
       "      <td>2332</td>\n",
       "      <td>0.126501</td>\n",
       "      <td>0.002306</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>somatic</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15295</th>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1369</td>\n",
       "      <td>4</td>\n",
       "      <td>1373</td>\n",
       "      <td>0.002913</td>\n",
       "      <td>0.002442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.314750e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15296</th>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2675</td>\n",
       "      <td>4</td>\n",
       "      <td>2679</td>\n",
       "      <td>0.001493</td>\n",
       "      <td>0.002189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.364490e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15297</th>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3044</td>\n",
       "      <td>20</td>\n",
       "      <td>3064</td>\n",
       "      <td>0.006527</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.430000e-06</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13600 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample_id  patient_id malignant        primer_id chrom  \\\n",
       "0            normal_blood           1        no   1_b_amplicrazy     1   \n",
       "1          omentum_site_1           1       yes   1_b_amplicrazy     1   \n",
       "2      right_ovary_site_1           1       yes   1_b_amplicrazy     1   \n",
       "3      right_ovary_site_2           1       yes   1_b_amplicrazy     1   \n",
       "4      right_ovary_site_3           1       yes   1_b_amplicrazy     1   \n",
       "...                   ...         ...       ...              ...   ...   \n",
       "15293        normal_blood          10        no  10_b_amplicrazy     X   \n",
       "15294  right_ovary_site_1          10       yes  10_b_amplicrazy     X   \n",
       "15295  right_ovary_site_2          10       yes  10_b_amplicrazy     X   \n",
       "15296  right_ovary_site_3          10       yes  10_b_amplicrazy     X   \n",
       "15297  right_ovary_site_4          10       yes  10_b_amplicrazy     X   \n",
       "\n",
       "             coord ref alt  ref_counts  alt_counts  depth  alt_freq  \\\n",
       "0       17090971.0   G   A         641           0    641  0.000000   \n",
       "1       17090971.0   G   A         449           0    449  0.000000   \n",
       "2       17090971.0   G   A         395          64    459  0.139434   \n",
       "3       17090971.0   G   A         415           1    416  0.002404   \n",
       "4       17090971.0   G   A         609           4    613  0.006525   \n",
       "...            ...  ..  ..         ...         ...    ...       ...   \n",
       "15293  110000000.0   A   G        3669          10   3679  0.002718   \n",
       "15294  110000000.0   A   G        2037         295   2332  0.126501   \n",
       "15295  110000000.0   A   G        1369           4   1373  0.002913   \n",
       "15296  110000000.0   A   G        2675           4   2679  0.001493   \n",
       "15297  110000000.0   A   G        3044          20   3064  0.006527   \n",
       "\n",
       "       background_average_alt_freq  ref_p_value   alt_p_value         status  \\\n",
       "0                         0.002034          0.0  1.000000e+00  normal_sample   \n",
       "1                         0.001597          0.0  1.000000e+00       wildtype   \n",
       "2                         0.002375          0.0  7.800000e-90        somatic   \n",
       "3                         0.002754          0.0  6.825260e-01       wildtype   \n",
       "4                         0.002145          0.0  4.440600e-02       wildtype   \n",
       "...                            ...          ...           ...            ...   \n",
       "15293                     0.002239          0.0  3.130860e-01  normal_sample   \n",
       "15294                     0.002306          0.0  0.000000e+00        somatic   \n",
       "15295                     0.002442          0.0  4.314750e-01       wildtype   \n",
       "15296                     0.002189          0.0  8.364490e-01       wildtype   \n",
       "15297                     0.001969          0.0  5.430000e-06       wildtype   \n",
       "\n",
       "      gene_name snpeff_impact    character_label  \n",
       "0           nan           NaN         1:17090971  \n",
       "1           nan           NaN         1:17090971  \n",
       "2           nan           NaN         1:17090971  \n",
       "3           nan           NaN         1:17090971  \n",
       "4           nan           NaN         1:17090971  \n",
       "...         ...           ...                ...  \n",
       "15293     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15294     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15295     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15296     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15297     RGAG1           LOW  RGAG1:X:110000000  \n",
       "\n",
       "[13600 rows x 19 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First, read in WGS reads data from McPherson and reformat into \n",
    "# file structure needed for MACHINA's and PyClone's clustering\n",
    "# TODO: how to handle \"status\" column? -> shows somatic, wildtype, germline?\n",
    "\n",
    "# Taken directly from supplementary table 10 of McPherson et.al.\n",
    "reads_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_10.csv'))\n",
    "patient_ids = reads_df['patient_id'].unique()\n",
    "print(\"patient IDs:\", patient_ids)\n",
    "reads_df['gene_name'] = reads_df['gene_name'].astype(str)\n",
    "def label_snv(row):\n",
    "    label = []\n",
    "    if row['gene_name'] != \"nan\":\n",
    "        label.append(row['gene_name'])\n",
    "    label += [str(row['chrom']), str(int(row['coord']))]\n",
    "    return (\":\").join(label)\n",
    "reads_df['character_label'] = reads_df.apply(lambda row: label_snv(row), axis=1)\n",
    "# Remove germline variants\n",
    "reads_df = reads_df[reads_df['status'] != 'germline']\n",
    "reads_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e5fd4091",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['left_fallopian_tube_site_b2', 'normal_blood', 'omentum_site_c1',\n",
       "       'right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3',\n",
       "       'right_ovary_site_4', 'right_ovary_site_a4', 'right_ovary_site_a9'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads_df[reads_df['patient_id']==10]['sample_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fcb86f6f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>paper_id</th>\n",
       "      <th>total_reads</th>\n",
       "      <th>aligned_reads</th>\n",
       "      <th>coverage</th>\n",
       "      <th>mutation_seq_snvs</th>\n",
       "      <th>strelka_snvs</th>\n",
       "      <th>all_snvs</th>\n",
       "      <th>high_quality_snvs</th>\n",
       "      <th>validated_snvs</th>\n",
       "      <th>ploidy</th>\n",
       "      <th>tumour_cell_proportion</th>\n",
       "      <th>subclone_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>normal_blood</td>\n",
       "      <td>Nml</td>\n",
       "      <td>1.290000e+09</td>\n",
       "      <td>1.050000e+09</td>\n",
       "      <td>34.47713</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>omentum_site_1</td>\n",
       "      <td>Om1</td>\n",
       "      <td>9.880000e+08</td>\n",
       "      <td>8.850000e+08</td>\n",
       "      <td>28.78891</td>\n",
       "      <td>5715.0</td>\n",
       "      <td>6665.0</td>\n",
       "      <td>7281.0</td>\n",
       "      <td>3550.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>3.606594</td>\n",
       "      <td>0.463669</td>\n",
       "      <td>0.318225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>1.260000e+09</td>\n",
       "      <td>1.080000e+09</td>\n",
       "      <td>34.99681</td>\n",
       "      <td>6276.0</td>\n",
       "      <td>7902.0</td>\n",
       "      <td>8514.0</td>\n",
       "      <td>3986.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>1.880553</td>\n",
       "      <td>0.826552</td>\n",
       "      <td>0.251911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>ROv2</td>\n",
       "      <td>1.390000e+09</td>\n",
       "      <td>1.190000e+09</td>\n",
       "      <td>38.78504</td>\n",
       "      <td>6004.0</td>\n",
       "      <td>7256.0</td>\n",
       "      <td>7912.0</td>\n",
       "      <td>3757.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>1.847366</td>\n",
       "      <td>0.695751</td>\n",
       "      <td>0.268036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>ROv3</td>\n",
       "      <td>1.210000e+09</td>\n",
       "      <td>1.060000e+09</td>\n",
       "      <td>33.31238</td>\n",
       "      <td>5803.0</td>\n",
       "      <td>6756.0</td>\n",
       "      <td>7410.0</td>\n",
       "      <td>3612.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>1.854135</td>\n",
       "      <td>0.678331</td>\n",
       "      <td>0.265984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>ROv4</td>\n",
       "      <td>1.480000e+09</td>\n",
       "      <td>1.300000e+09</td>\n",
       "      <td>36.79460</td>\n",
       "      <td>6501.0</td>\n",
       "      <td>7920.0</td>\n",
       "      <td>8664.0</td>\n",
       "      <td>4017.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>1.857220</td>\n",
       "      <td>0.702755</td>\n",
       "      <td>0.240692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>small_bowel_site_1</td>\n",
       "      <td>SBwl</td>\n",
       "      <td>1.240000e+09</td>\n",
       "      <td>1.100000e+09</td>\n",
       "      <td>35.65017</td>\n",
       "      <td>7100.0</td>\n",
       "      <td>8323.0</td>\n",
       "      <td>9137.0</td>\n",
       "      <td>4357.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>3.611431</td>\n",
       "      <td>0.448979</td>\n",
       "      <td>0.270676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>normal_blood</td>\n",
       "      <td>Nml</td>\n",
       "      <td>1.400000e+09</td>\n",
       "      <td>1.190000e+09</td>\n",
       "      <td>37.44915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>omentum_site_1</td>\n",
       "      <td>Om1</td>\n",
       "      <td>1.060000e+09</td>\n",
       "      <td>9.370000e+08</td>\n",
       "      <td>30.63196</td>\n",
       "      <td>1860.0</td>\n",
       "      <td>2705.0</td>\n",
       "      <td>2849.0</td>\n",
       "      <td>1228.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>3.016544</td>\n",
       "      <td>0.196412</td>\n",
       "      <td>0.347316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>omentum_site_2</td>\n",
       "      <td>Om2</td>\n",
       "      <td>1.190000e+09</td>\n",
       "      <td>1.070000e+09</td>\n",
       "      <td>34.93282</td>\n",
       "      <td>3046.0</td>\n",
       "      <td>4036.0</td>\n",
       "      <td>4242.0</td>\n",
       "      <td>1980.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>3.014453</td>\n",
       "      <td>0.239950</td>\n",
       "      <td>0.348032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>1.110000e+09</td>\n",
       "      <td>9.820000e+08</td>\n",
       "      <td>31.70880</td>\n",
       "      <td>6946.0</td>\n",
       "      <td>8342.0</td>\n",
       "      <td>8787.0</td>\n",
       "      <td>4494.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>2.994620</td>\n",
       "      <td>0.495999</td>\n",
       "      <td>0.272663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>ROv2</td>\n",
       "      <td>1.290000e+09</td>\n",
       "      <td>1.140000e+09</td>\n",
       "      <td>34.94380</td>\n",
       "      <td>5571.0</td>\n",
       "      <td>6979.0</td>\n",
       "      <td>7313.0</td>\n",
       "      <td>3630.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>2.945082</td>\n",
       "      <td>0.458403</td>\n",
       "      <td>0.341355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3</td>\n",
       "      <td>adnexa_site_1</td>\n",
       "      <td>Adnx</td>\n",
       "      <td>1.350000e+09</td>\n",
       "      <td>1.190000e+09</td>\n",
       "      <td>37.49350</td>\n",
       "      <td>6167.0</td>\n",
       "      <td>7515.0</td>\n",
       "      <td>8089.0</td>\n",
       "      <td>3883.0</td>\n",
       "      <td>161.0</td>\n",
       "      <td>3.438031</td>\n",
       "      <td>0.545925</td>\n",
       "      <td>0.273037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>3</td>\n",
       "      <td>normal_blood</td>\n",
       "      <td>Nml</td>\n",
       "      <td>1.160000e+09</td>\n",
       "      <td>1.010000e+09</td>\n",
       "      <td>31.97733</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3</td>\n",
       "      <td>omentum_site_1</td>\n",
       "      <td>Om1</td>\n",
       "      <td>1.090000e+09</td>\n",
       "      <td>9.600000e+08</td>\n",
       "      <td>31.42286</td>\n",
       "      <td>3489.0</td>\n",
       "      <td>4709.0</td>\n",
       "      <td>5022.0</td>\n",
       "      <td>2217.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>3.429465</td>\n",
       "      <td>0.373117</td>\n",
       "      <td>0.384717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>9.610000e+08</td>\n",
       "      <td>8.230000e+08</td>\n",
       "      <td>27.01643</td>\n",
       "      <td>4264.0</td>\n",
       "      <td>5501.0</td>\n",
       "      <td>5877.0</td>\n",
       "      <td>2684.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>3.442529</td>\n",
       "      <td>0.513049</td>\n",
       "      <td>0.330682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3</td>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>ROv2</td>\n",
       "      <td>1.300000e+09</td>\n",
       "      <td>1.100000e+09</td>\n",
       "      <td>35.89313</td>\n",
       "      <td>5775.0</td>\n",
       "      <td>6890.0</td>\n",
       "      <td>7454.0</td>\n",
       "      <td>3612.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>3.504195</td>\n",
       "      <td>0.495587</td>\n",
       "      <td>0.279543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>4</td>\n",
       "      <td>left_pelvic_sidewall_site_1</td>\n",
       "      <td>LPvS</td>\n",
       "      <td>1.330000e+09</td>\n",
       "      <td>1.110000e+09</td>\n",
       "      <td>35.28034</td>\n",
       "      <td>4036.0</td>\n",
       "      <td>5537.0</td>\n",
       "      <td>5736.0</td>\n",
       "      <td>2771.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>1.843066</td>\n",
       "      <td>0.390764</td>\n",
       "      <td>0.255537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>4</td>\n",
       "      <td>normal_blood</td>\n",
       "      <td>Nml</td>\n",
       "      <td>1.340000e+09</td>\n",
       "      <td>1.120000e+09</td>\n",
       "      <td>36.11832</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>4</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>1.260000e+09</td>\n",
       "      <td>1.110000e+09</td>\n",
       "      <td>33.92379</td>\n",
       "      <td>5931.0</td>\n",
       "      <td>7193.0</td>\n",
       "      <td>7415.0</td>\n",
       "      <td>3984.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>1.938613</td>\n",
       "      <td>0.652612</td>\n",
       "      <td>0.188791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>4</td>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>ROv2</td>\n",
       "      <td>1.270000e+09</td>\n",
       "      <td>1.090000e+09</td>\n",
       "      <td>34.69970</td>\n",
       "      <td>4844.0</td>\n",
       "      <td>5932.0</td>\n",
       "      <td>6195.0</td>\n",
       "      <td>3265.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>1.939805</td>\n",
       "      <td>0.405332</td>\n",
       "      <td>0.286485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>4</td>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>ROv3</td>\n",
       "      <td>1.140000e+09</td>\n",
       "      <td>9.960000e+08</td>\n",
       "      <td>31.86368</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10854.0</td>\n",
       "      <td>10854.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>170.0</td>\n",
       "      <td>1.932581</td>\n",
       "      <td>0.680065</td>\n",
       "      <td>0.265265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>4</td>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>ROv4</td>\n",
       "      <td>1.060000e+09</td>\n",
       "      <td>9.410000e+08</td>\n",
       "      <td>29.52893</td>\n",
       "      <td>5670.0</td>\n",
       "      <td>6908.0</td>\n",
       "      <td>7193.0</td>\n",
       "      <td>3798.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>1.867389</td>\n",
       "      <td>0.544647</td>\n",
       "      <td>0.274072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>7</td>\n",
       "      <td>brain_metastasis</td>\n",
       "      <td>BrnM</td>\n",
       "      <td>1.450000e+09</td>\n",
       "      <td>1.220000e+09</td>\n",
       "      <td>38.80455</td>\n",
       "      <td>9689.0</td>\n",
       "      <td>11601.0</td>\n",
       "      <td>12620.0</td>\n",
       "      <td>6023.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>3.470541</td>\n",
       "      <td>0.611919</td>\n",
       "      <td>0.302643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>7</td>\n",
       "      <td>left_ovary_site_1</td>\n",
       "      <td>LOv1</td>\n",
       "      <td>1.290000e+09</td>\n",
       "      <td>1.100000e+09</td>\n",
       "      <td>35.95753</td>\n",
       "      <td>6361.0</td>\n",
       "      <td>8134.0</td>\n",
       "      <td>8683.0</td>\n",
       "      <td>4085.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>2.263655</td>\n",
       "      <td>0.829766</td>\n",
       "      <td>0.251688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>7</td>\n",
       "      <td>normal_blood</td>\n",
       "      <td>Nml</td>\n",
       "      <td>1.070000e+09</td>\n",
       "      <td>9.100000e+08</td>\n",
       "      <td>29.25048</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>7</td>\n",
       "      <td>righ_pelvic_mass</td>\n",
       "      <td>RPvM</td>\n",
       "      <td>1.020000e+09</td>\n",
       "      <td>7.160000e+08</td>\n",
       "      <td>22.97519</td>\n",
       "      <td>4951.0</td>\n",
       "      <td>6580.0</td>\n",
       "      <td>7276.0</td>\n",
       "      <td>2962.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>1.939543</td>\n",
       "      <td>0.367508</td>\n",
       "      <td>0.293328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>9</td>\n",
       "      <td>left_ovary_site_1</td>\n",
       "      <td>LOv1</td>\n",
       "      <td>9.060000e+08</td>\n",
       "      <td>6.840000e+08</td>\n",
       "      <td>21.89483</td>\n",
       "      <td>2712.0</td>\n",
       "      <td>4304.0</td>\n",
       "      <td>4448.0</td>\n",
       "      <td>1772.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>3.521985</td>\n",
       "      <td>0.761987</td>\n",
       "      <td>0.292346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>9</td>\n",
       "      <td>left_ovary_site_2</td>\n",
       "      <td>LOv2</td>\n",
       "      <td>1.020000e+09</td>\n",
       "      <td>8.010000e+08</td>\n",
       "      <td>25.10629</td>\n",
       "      <td>2892.0</td>\n",
       "      <td>4115.0</td>\n",
       "      <td>4295.0</td>\n",
       "      <td>1878.0</td>\n",
       "      <td>142.0</td>\n",
       "      <td>3.515547</td>\n",
       "      <td>0.654521</td>\n",
       "      <td>0.363734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>9</td>\n",
       "      <td>normal_blood</td>\n",
       "      <td>Nml</td>\n",
       "      <td>1.230000e+09</td>\n",
       "      <td>1.050000e+09</td>\n",
       "      <td>39.68317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>9</td>\n",
       "      <td>omentum_site_1</td>\n",
       "      <td>Om1</td>\n",
       "      <td>9.900000e+08</td>\n",
       "      <td>8.510000e+08</td>\n",
       "      <td>27.17733</td>\n",
       "      <td>1083.0</td>\n",
       "      <td>1655.0</td>\n",
       "      <td>1771.0</td>\n",
       "      <td>682.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>3.460575</td>\n",
       "      <td>0.178824</td>\n",
       "      <td>0.313948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>9</td>\n",
       "      <td>omentum_site_2</td>\n",
       "      <td>Om2</td>\n",
       "      <td>6.610000e+08</td>\n",
       "      <td>5.570000e+08</td>\n",
       "      <td>17.76029</td>\n",
       "      <td>2546.0</td>\n",
       "      <td>4135.0</td>\n",
       "      <td>4264.0</td>\n",
       "      <td>1641.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>3.366123</td>\n",
       "      <td>0.672038</td>\n",
       "      <td>0.275741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>9</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>1.070000e+09</td>\n",
       "      <td>8.940000e+08</td>\n",
       "      <td>27.95664</td>\n",
       "      <td>3531.0</td>\n",
       "      <td>4814.0</td>\n",
       "      <td>5013.0</td>\n",
       "      <td>2322.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>3.374040</td>\n",
       "      <td>0.601017</td>\n",
       "      <td>0.307162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>10</td>\n",
       "      <td>normal_blood</td>\n",
       "      <td>Nml</td>\n",
       "      <td>1.230000e+09</td>\n",
       "      <td>1.070000e+09</td>\n",
       "      <td>33.96793</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>1.210000e+09</td>\n",
       "      <td>9.970000e+08</td>\n",
       "      <td>31.78397</td>\n",
       "      <td>12213.0</td>\n",
       "      <td>15362.0</td>\n",
       "      <td>16195.0</td>\n",
       "      <td>7948.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>2.857267</td>\n",
       "      <td>0.445714</td>\n",
       "      <td>0.317931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>ROv2</td>\n",
       "      <td>1.160000e+09</td>\n",
       "      <td>9.790000e+08</td>\n",
       "      <td>31.25871</td>\n",
       "      <td>16798.0</td>\n",
       "      <td>20907.0</td>\n",
       "      <td>21952.0</td>\n",
       "      <td>10890.0</td>\n",
       "      <td>231.0</td>\n",
       "      <td>3.175412</td>\n",
       "      <td>0.696471</td>\n",
       "      <td>0.301428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>ROv3</td>\n",
       "      <td>1.050000e+09</td>\n",
       "      <td>8.470000e+08</td>\n",
       "      <td>27.19881</td>\n",
       "      <td>6499.0</td>\n",
       "      <td>8540.0</td>\n",
       "      <td>9038.0</td>\n",
       "      <td>4224.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>2.893883</td>\n",
       "      <td>0.270729</td>\n",
       "      <td>0.313770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>ROv4</td>\n",
       "      <td>1.290000e+09</td>\n",
       "      <td>1.100000e+09</td>\n",
       "      <td>33.50294</td>\n",
       "      <td>10129.0</td>\n",
       "      <td>12891.0</td>\n",
       "      <td>13576.0</td>\n",
       "      <td>6797.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>2.891926</td>\n",
       "      <td>0.323749</td>\n",
       "      <td>0.322962</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    patient_id                    sample_id paper_id   total_reads  \\\n",
       "0            1                 normal_blood      Nml  1.290000e+09   \n",
       "1            1               omentum_site_1      Om1  9.880000e+08   \n",
       "2            1           right_ovary_site_1     ROv1  1.260000e+09   \n",
       "3            1           right_ovary_site_2     ROv2  1.390000e+09   \n",
       "4            1           right_ovary_site_3     ROv3  1.210000e+09   \n",
       "5            1           right_ovary_site_4     ROv4  1.480000e+09   \n",
       "6            1           small_bowel_site_1     SBwl  1.240000e+09   \n",
       "7            2                 normal_blood      Nml  1.400000e+09   \n",
       "8            2               omentum_site_1      Om1  1.060000e+09   \n",
       "9            2               omentum_site_2      Om2  1.190000e+09   \n",
       "10           2           right_ovary_site_1     ROv1  1.110000e+09   \n",
       "11           2           right_ovary_site_2     ROv2  1.290000e+09   \n",
       "12           3                adnexa_site_1     Adnx  1.350000e+09   \n",
       "13           3                 normal_blood      Nml  1.160000e+09   \n",
       "14           3               omentum_site_1      Om1  1.090000e+09   \n",
       "15           3           right_ovary_site_1     ROv1  9.610000e+08   \n",
       "16           3           right_ovary_site_2     ROv2  1.300000e+09   \n",
       "17           4  left_pelvic_sidewall_site_1     LPvS  1.330000e+09   \n",
       "18           4                 normal_blood      Nml  1.340000e+09   \n",
       "19           4           right_ovary_site_1     ROv1  1.260000e+09   \n",
       "20           4           right_ovary_site_2     ROv2  1.270000e+09   \n",
       "21           4           right_ovary_site_3     ROv3  1.140000e+09   \n",
       "22           4           right_ovary_site_4     ROv4  1.060000e+09   \n",
       "23           7             brain_metastasis     BrnM  1.450000e+09   \n",
       "24           7            left_ovary_site_1     LOv1  1.290000e+09   \n",
       "25           7                 normal_blood      Nml  1.070000e+09   \n",
       "26           7             righ_pelvic_mass     RPvM  1.020000e+09   \n",
       "27           9            left_ovary_site_1     LOv1  9.060000e+08   \n",
       "28           9            left_ovary_site_2     LOv2  1.020000e+09   \n",
       "29           9                 normal_blood      Nml  1.230000e+09   \n",
       "30           9               omentum_site_1      Om1  9.900000e+08   \n",
       "31           9               omentum_site_2      Om2  6.610000e+08   \n",
       "32           9           right_ovary_site_1     ROv1  1.070000e+09   \n",
       "33          10                 normal_blood      Nml  1.230000e+09   \n",
       "34          10           right_ovary_site_1     ROv1  1.210000e+09   \n",
       "35          10           right_ovary_site_2     ROv2  1.160000e+09   \n",
       "36          10           right_ovary_site_3     ROv3  1.050000e+09   \n",
       "37          10           right_ovary_site_4     ROv4  1.290000e+09   \n",
       "\n",
       "    aligned_reads  coverage  mutation_seq_snvs  strelka_snvs  all_snvs  \\\n",
       "0    1.050000e+09  34.47713                NaN           NaN       NaN   \n",
       "1    8.850000e+08  28.78891             5715.0        6665.0    7281.0   \n",
       "2    1.080000e+09  34.99681             6276.0        7902.0    8514.0   \n",
       "3    1.190000e+09  38.78504             6004.0        7256.0    7912.0   \n",
       "4    1.060000e+09  33.31238             5803.0        6756.0    7410.0   \n",
       "5    1.300000e+09  36.79460             6501.0        7920.0    8664.0   \n",
       "6    1.100000e+09  35.65017             7100.0        8323.0    9137.0   \n",
       "7    1.190000e+09  37.44915                NaN           NaN       NaN   \n",
       "8    9.370000e+08  30.63196             1860.0        2705.0    2849.0   \n",
       "9    1.070000e+09  34.93282             3046.0        4036.0    4242.0   \n",
       "10   9.820000e+08  31.70880             6946.0        8342.0    8787.0   \n",
       "11   1.140000e+09  34.94380             5571.0        6979.0    7313.0   \n",
       "12   1.190000e+09  37.49350             6167.0        7515.0    8089.0   \n",
       "13   1.010000e+09  31.97733                NaN           NaN       NaN   \n",
       "14   9.600000e+08  31.42286             3489.0        4709.0    5022.0   \n",
       "15   8.230000e+08  27.01643             4264.0        5501.0    5877.0   \n",
       "16   1.100000e+09  35.89313             5775.0        6890.0    7454.0   \n",
       "17   1.110000e+09  35.28034             4036.0        5537.0    5736.0   \n",
       "18   1.120000e+09  36.11832                NaN           NaN       NaN   \n",
       "19   1.110000e+09  33.92379             5931.0        7193.0    7415.0   \n",
       "20   1.090000e+09  34.69970             4844.0        5932.0    6195.0   \n",
       "21   9.960000e+08  31.86368                2.0       10854.0   10854.0   \n",
       "22   9.410000e+08  29.52893             5670.0        6908.0    7193.0   \n",
       "23   1.220000e+09  38.80455             9689.0       11601.0   12620.0   \n",
       "24   1.100000e+09  35.95753             6361.0        8134.0    8683.0   \n",
       "25   9.100000e+08  29.25048                NaN           NaN       NaN   \n",
       "26   7.160000e+08  22.97519             4951.0        6580.0    7276.0   \n",
       "27   6.840000e+08  21.89483             2712.0        4304.0    4448.0   \n",
       "28   8.010000e+08  25.10629             2892.0        4115.0    4295.0   \n",
       "29   1.050000e+09  39.68317                NaN           NaN       NaN   \n",
       "30   8.510000e+08  27.17733             1083.0        1655.0    1771.0   \n",
       "31   5.570000e+08  17.76029             2546.0        4135.0    4264.0   \n",
       "32   8.940000e+08  27.95664             3531.0        4814.0    5013.0   \n",
       "33   1.070000e+09  33.96793                NaN           NaN       NaN   \n",
       "34   9.970000e+08  31.78397            12213.0       15362.0   16195.0   \n",
       "35   9.790000e+08  31.25871            16798.0       20907.0   21952.0   \n",
       "36   8.470000e+08  27.19881             6499.0        8540.0    9038.0   \n",
       "37   1.100000e+09  33.50294            10129.0       12891.0   13576.0   \n",
       "\n",
       "    high_quality_snvs  validated_snvs    ploidy  tumour_cell_proportion  \\\n",
       "0                 NaN             NaN       NaN                     NaN   \n",
       "1              3550.0           120.0  3.606594                0.463669   \n",
       "2              3986.0           132.0  1.880553                0.826552   \n",
       "3              3757.0           111.0  1.847366                0.695751   \n",
       "4              3612.0           118.0  1.854135                0.678331   \n",
       "5              4017.0           112.0  1.857220                0.702755   \n",
       "6              4357.0           127.0  3.611431                0.448979   \n",
       "7                 NaN             NaN       NaN                     NaN   \n",
       "8              1228.0           106.0  3.016544                0.196412   \n",
       "9              1980.0           107.0  3.014453                0.239950   \n",
       "10             4494.0            87.0  2.994620                0.495999   \n",
       "11             3630.0           119.0  2.945082                0.458403   \n",
       "12             3883.0           161.0  3.438031                0.545925   \n",
       "13                NaN             NaN       NaN                     NaN   \n",
       "14             2217.0           159.0  3.429465                0.373117   \n",
       "15             2684.0           154.0  3.442529                0.513049   \n",
       "16             3612.0           153.0  3.504195                0.495587   \n",
       "17             2771.0           117.0  1.843066                0.390764   \n",
       "18                NaN             NaN       NaN                     NaN   \n",
       "19             3984.0           158.0  1.938613                0.652612   \n",
       "20             3265.0           156.0  1.939805                0.405332   \n",
       "21                1.0           170.0  1.932581                0.680065   \n",
       "22             3798.0           154.0  1.867389                0.544647   \n",
       "23             6023.0           182.0  3.470541                0.611919   \n",
       "24             4085.0           163.0  2.263655                0.829766   \n",
       "25                NaN             NaN       NaN                     NaN   \n",
       "26             2962.0           115.0  1.939543                0.367508   \n",
       "27             1772.0           119.0  3.521985                0.761987   \n",
       "28             1878.0           142.0  3.515547                0.654521   \n",
       "29                NaN             NaN       NaN                     NaN   \n",
       "30              682.0           108.0  3.460575                0.178824   \n",
       "31             1641.0           111.0  3.366123                0.672038   \n",
       "32             2322.0           112.0  3.374040                0.601017   \n",
       "33                NaN             NaN       NaN                     NaN   \n",
       "34             7948.0           168.0  2.857267                0.445714   \n",
       "35            10890.0           231.0  3.175412                0.696471   \n",
       "36             4224.0           157.0  2.893883                0.270729   \n",
       "37             6797.0           166.0  2.891926                0.323749   \n",
       "\n",
       "    subclone_frequency  \n",
       "0                  NaN  \n",
       "1             0.318225  \n",
       "2             0.251911  \n",
       "3             0.268036  \n",
       "4             0.265984  \n",
       "5             0.240692  \n",
       "6             0.270676  \n",
       "7                  NaN  \n",
       "8             0.347316  \n",
       "9             0.348032  \n",
       "10            0.272663  \n",
       "11            0.341355  \n",
       "12            0.273037  \n",
       "13                 NaN  \n",
       "14            0.384717  \n",
       "15            0.330682  \n",
       "16            0.279543  \n",
       "17            0.255537  \n",
       "18                 NaN  \n",
       "19            0.188791  \n",
       "20            0.286485  \n",
       "21            0.265265  \n",
       "22            0.274072  \n",
       "23            0.302643  \n",
       "24            0.251688  \n",
       "25                 NaN  \n",
       "26            0.293328  \n",
       "27            0.292346  \n",
       "28            0.363734  \n",
       "29                 NaN  \n",
       "30            0.313948  \n",
       "31            0.275741  \n",
       "32            0.307162  \n",
       "33                 NaN  \n",
       "34            0.317931  \n",
       "35            0.301428  \n",
       "36            0.313770  \n",
       "37            0.322962  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the table that contains which \"high-quality DNA extractions (discovery samples)\"\n",
    "# were used in the paper's analysis\n",
    "discovery_samples_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_2.csv'))\n",
    "discovery_samples_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ab653e9",
   "metadata": {},
   "source": [
    "## (2) Generate inputs for PyClone clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b201a3fd",
   "metadata": {},
   "source": [
    "### (2a) Create input for PyClone clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "376d6f8b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        patient_id           sample_id chrom        start          end  major  \\\n",
      "0                1  small_bowel_site_1     1          1.0      71283.0     16   \n",
      "1                1  small_bowel_site_1     1      73560.0      94700.0      7   \n",
      "2                1  small_bowel_site_1     1     101640.0     110479.0     13   \n",
      "3                1  small_bowel_site_1     1     110479.0     141357.0     22   \n",
      "4                1  small_bowel_site_1     1     144162.0     247926.0     24   \n",
      "...            ...                 ...   ...          ...          ...    ...   \n",
      "135154          10  right_ovary_site_4     X  153000000.0  153000000.0      0   \n",
      "135155          10  right_ovary_site_4     X  153000000.0  154000000.0      2   \n",
      "135156          10  right_ovary_site_4     X  154000000.0  155000000.0      2   \n",
      "135157          10  right_ovary_site_4     X  155000000.0  155000000.0      2   \n",
      "135158          10  right_ovary_site_4     X  155000000.0  155000000.0      2   \n",
      "\n",
      "        minor  major_sub  minor_sub  subclonal  \n",
      "0           7          0          0        NaN  \n",
      "1           4          0          0        NaN  \n",
      "2           9          0          0        NaN  \n",
      "3          14          0          0        NaN  \n",
      "4          24          0          0        NaN  \n",
      "...       ...        ...        ...        ...  \n",
      "135154      0          0          0        NaN  \n",
      "135155      0          0          0       0.22  \n",
      "135156      0          1          0       0.99  \n",
      "135157      0          0          0        NaN  \n",
      "135158      0          1          0       0.97  \n",
      "\n",
      "[135159 rows x 10 columns]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>malignant</th>\n",
       "      <th>primer_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref_counts</th>\n",
       "      <th>alt_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>alt_freq</th>\n",
       "      <th>background_average_alt_freq</th>\n",
       "      <th>ref_p_value</th>\n",
       "      <th>alt_p_value</th>\n",
       "      <th>status</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snpeff_impact</th>\n",
       "      <th>character_label</th>\n",
       "      <th>major_cn</th>\n",
       "      <th>minor_cn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>normal_blood</td>\n",
       "      <td>1</td>\n",
       "      <td>no</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>641</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>normal_sample</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>omentum_site_1</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>449</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>395</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>0.139434</td>\n",
       "      <td>0.002375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.800000e-90</td>\n",
       "      <td>somatic</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002404</td>\n",
       "      <td>0.002754</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.825260e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>609</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006525</td>\n",
       "      <td>0.002145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.440600e-02</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15293</th>\n",
       "      <td>normal_blood</td>\n",
       "      <td>10</td>\n",
       "      <td>no</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3669</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002718</td>\n",
       "      <td>0.002239</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.130860e-01</td>\n",
       "      <td>normal_sample</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15294</th>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2037</td>\n",
       "      <td>295</td>\n",
       "      <td>...</td>\n",
       "      <td>0.126501</td>\n",
       "      <td>0.002306</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>somatic</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15295</th>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1369</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002913</td>\n",
       "      <td>0.002442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.314750e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15296</th>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2675</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001493</td>\n",
       "      <td>0.002189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.364490e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15297</th>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3044</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006527</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.430000e-06</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13600 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample_id  patient_id malignant        primer_id chrom  \\\n",
       "0            normal_blood           1        no   1_b_amplicrazy     1   \n",
       "1          omentum_site_1           1       yes   1_b_amplicrazy     1   \n",
       "2      right_ovary_site_1           1       yes   1_b_amplicrazy     1   \n",
       "3      right_ovary_site_2           1       yes   1_b_amplicrazy     1   \n",
       "4      right_ovary_site_3           1       yes   1_b_amplicrazy     1   \n",
       "...                   ...         ...       ...              ...   ...   \n",
       "15293        normal_blood          10        no  10_b_amplicrazy     X   \n",
       "15294  right_ovary_site_1          10       yes  10_b_amplicrazy     X   \n",
       "15295  right_ovary_site_2          10       yes  10_b_amplicrazy     X   \n",
       "15296  right_ovary_site_3          10       yes  10_b_amplicrazy     X   \n",
       "15297  right_ovary_site_4          10       yes  10_b_amplicrazy     X   \n",
       "\n",
       "             coord ref alt  ref_counts  alt_counts  ...  alt_freq  \\\n",
       "0       17090971.0   G   A         641           0  ...  0.000000   \n",
       "1       17090971.0   G   A         449           0  ...  0.000000   \n",
       "2       17090971.0   G   A         395          64  ...  0.139434   \n",
       "3       17090971.0   G   A         415           1  ...  0.002404   \n",
       "4       17090971.0   G   A         609           4  ...  0.006525   \n",
       "...            ...  ..  ..         ...         ...  ...       ...   \n",
       "15293  110000000.0   A   G        3669          10  ...  0.002718   \n",
       "15294  110000000.0   A   G        2037         295  ...  0.126501   \n",
       "15295  110000000.0   A   G        1369           4  ...  0.002913   \n",
       "15296  110000000.0   A   G        2675           4  ...  0.001493   \n",
       "15297  110000000.0   A   G        3044          20  ...  0.006527   \n",
       "\n",
       "       background_average_alt_freq  ref_p_value   alt_p_value         status  \\\n",
       "0                         0.002034          0.0  1.000000e+00  normal_sample   \n",
       "1                         0.001597          0.0  1.000000e+00       wildtype   \n",
       "2                         0.002375          0.0  7.800000e-90        somatic   \n",
       "3                         0.002754          0.0  6.825260e-01       wildtype   \n",
       "4                         0.002145          0.0  4.440600e-02       wildtype   \n",
       "...                            ...          ...           ...            ...   \n",
       "15293                     0.002239          0.0  3.130860e-01  normal_sample   \n",
       "15294                     0.002306          0.0  0.000000e+00        somatic   \n",
       "15295                     0.002442          0.0  4.314750e-01       wildtype   \n",
       "15296                     0.002189          0.0  8.364490e-01       wildtype   \n",
       "15297                     0.001969          0.0  5.430000e-06       wildtype   \n",
       "\n",
       "      gene_name snpeff_impact    character_label major_cn  minor_cn  \n",
       "0           nan           NaN         1:17090971        0         0  \n",
       "1           nan           NaN         1:17090971       13         5  \n",
       "2           nan           NaN         1:17090971        4         2  \n",
       "3           nan           NaN         1:17090971        6         2  \n",
       "4           nan           NaN         1:17090971        6         2  \n",
       "...         ...           ...                ...      ...       ...  \n",
       "15293     RGAG1           LOW  RGAG1:X:110000000        0         0  \n",
       "15294     RGAG1           LOW  RGAG1:X:110000000        1         0  \n",
       "15295     RGAG1           LOW  RGAG1:X:110000000        1         0  \n",
       "15296     RGAG1           LOW  RGAG1:X:110000000        1         0  \n",
       "15297     RGAG1           LOW  RGAG1:X:110000000        1         0  \n",
       "\n",
       "[13600 rows x 21 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# get CNA info from supplement of McPherson et. al.\n",
    "cna_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_7.csv'))\n",
    "print(cna_df)\n",
    "\n",
    "def subset_cna(row):\n",
    "    chrom = row['chrom']\n",
    "    coord = row['coord']\n",
    "    patient_id = row['patient_id']\n",
    "    sample_id = row['sample_id']\n",
    "    df = cna_df[(cna_df['patient_id'] == patient_id) & (cna_df['sample_id'] == sample_id) & (cna_df['chrom'] == chrom) & (cna_df['start'] <= coord) & (cna_df['end'] >= coord)]\n",
    "    return df\n",
    "\n",
    "def major_cn(row):\n",
    "    coord_df = subset_cna(row)\n",
    "    if len(coord_df) != 1:\n",
    "        #print(\"Not found in CNA:\\n\", row['chrom'], row['coord'])\n",
    "        return 0\n",
    "    return int(coord_df['major'])\n",
    "\n",
    "def minor_cn(row):\n",
    "    coord_df = subset_cna(row)\n",
    "    if len(coord_df) != 1:\n",
    "        return 0\n",
    "    return int(coord_df['minor'])\n",
    "\n",
    "\n",
    "# Add CNAs to reads_df\n",
    "cna_reads_df = reads_df.copy(deep=True)\n",
    "cna_reads_df['major_cn'] =  cna_reads_df.apply(lambda row: major_cn(row), axis=1)\n",
    "cna_reads_df['minor_cn'] =  cna_reads_df.apply(lambda row: minor_cn(row), axis=1)\n",
    "cna_reads_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebd57557",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do some pre-processing to remove any mutations where CNA info is not available or\n",
    "# the major allele CN is 0 (if the malignant cells have no copies of the region \n",
    "# overlapping the mutation, the mutation cannot exist.) or minor allele CN > major allele CN\n",
    "cna_reads_df = cna_reads_df[cna_reads_df['major_cn'] != 0]\n",
    "cna_reads_df = cna_reads_df[(cna_reads_df['minor_cn'] <= cna_reads_df['major_cn'])]\n",
    "cna_reads_df['normal_cn'] = 2 # copy number of the locus in normal cells is 2 (no male chromosomes here)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "056f5685",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pyclone_dir = os.path.join(data_dir, \"pyclone_analysis\")\n",
    "\n",
    "cols = ['sample_id', 'ref_counts', 'alt_counts', 'character_label', \"chrom\", \"coord\", \"major_cn\", \"minor_cn\", \"normal_cn\"]\n",
    "# put df in format used in PyClone, with columns:\n",
    "#'mutation_id', 'ref_counts','var_counts', 'normal_cn', 'minor_cn', 'major_cn'\n",
    "for patient_id in patient_ids:\n",
    "    patient_dir = os.path.join(pyclone_dir, f\"patient_{patient_id}\")\n",
    "    if not os.path.exists(patient_dir): \n",
    "        os.makedirs(patient_dir)\n",
    "    patient_subset = cna_reads_df[cna_reads_df['patient_id'] == patient_id][cols]\n",
    "    # only keep the high quality discovery samples used in the paper\n",
    "    discovery_samples = discovery_samples_df[discovery_samples_df['patient_id']==patient_id]['sample_id']\n",
    "    patient_subset = patient_subset[patient_subset['sample_id'].isin(discovery_samples)]\n",
    "    # remove normal samples\n",
    "    patient_subset = patient_subset[patient_subset['sample_id'] != 'normal_blood'] \n",
    "    sample_ids = list(patient_subset['sample_id'].unique())\n",
    "    patient_subset = patient_subset.rename(columns={\"ref_counts\": \"ref_counts\", \"alt_counts\": \"var_counts\", \"character_label\": \"mutation_id\"})\n",
    "    \n",
    "    for sample_id in sample_ids:\n",
    "        sample_subset = patient_subset[patient_subset['sample_id'] == sample_id]\n",
    "        sample_subset = sample_subset[['mutation_id', 'ref_counts', 'var_counts', 'normal_cn', 'minor_cn', 'major_cn']]\n",
    "        sample_subset.to_csv(os.path.join(patient_dir,  f\"patient{patient_id}_{sample_id}.tsv\"), index=False, sep=\"\\t\")    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "299c9590",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_1.log -e error_1.log PyClone run_analysis_pipeline --in_files /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_1/patient1_omentum_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_1/patient1_right_ovary_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_1/patient1_right_ovary_site_2.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_1/patient1_right_ovary_site_3.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_1/patient1_right_ovary_site_4.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_1/patient1_small_bowel_site_1.tsv --working_dir /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_1 --tumour_contents 0.463669 0.826552 0.695751 0.678331 0.702755 0.448979 --samples omentum_site_1 right_ovary_site_1 right_ovary_site_2 right_ovary_site_3 right_ovary_site_4 small_bowel_site_1 --num_iters 100000 --burnin 50000\n",
      "bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_2.log -e error_2.log PyClone run_analysis_pipeline --in_files /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_2/patient2_omentum_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_2/patient2_omentum_site_2.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_2/patient2_right_ovary_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_2/patient2_right_ovary_site_2.tsv --working_dir /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_2 --tumour_contents 0.196412 0.23995 0.495999 0.458403 --samples omentum_site_1 omentum_site_2 right_ovary_site_1 right_ovary_site_2 --num_iters 100000 --burnin 50000\n",
      "bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_3.log -e error_3.log PyClone run_analysis_pipeline --in_files /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_3/patient3_adnexa_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_3/patient3_omentum_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_3/patient3_right_ovary_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_3/patient3_right_ovary_site_2.tsv --working_dir /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_3 --tumour_contents 0.545925 0.373117 0.513049 0.495587 --samples adnexa_site_1 omentum_site_1 right_ovary_site_1 right_ovary_site_2 --num_iters 100000 --burnin 50000\n",
      "bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_4.log -e error_4.log PyClone run_analysis_pipeline --in_files /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_4/patient4_left_pelvic_sidewall_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_4/patient4_right_ovary_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_4/patient4_right_ovary_site_2.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_4/patient4_right_ovary_site_3.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_4/patient4_right_ovary_site_4.tsv --working_dir /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_4 --tumour_contents 0.390764 0.652612 0.405332 0.680065 0.544647 --samples left_pelvic_sidewall_site_1 right_ovary_site_1 right_ovary_site_2 right_ovary_site_3 right_ovary_site_4 --num_iters 100000 --burnin 50000\n",
      "bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_7.log -e error_7.log PyClone run_analysis_pipeline --in_files /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_7/patient7_brain_metastasis.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_7/patient7_left_ovary_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_7/patient7_righ_pelvic_mass.tsv --working_dir /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_7 --tumour_contents 0.611919 0.829766 0.367508 --samples brain_metastasis left_ovary_site_1 righ_pelvic_mass --num_iters 100000 --burnin 50000\n",
      "bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_9.log -e error_9.log PyClone run_analysis_pipeline --in_files /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_9/patient9_left_ovary_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_9/patient9_left_ovary_site_2.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_9/patient9_omentum_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_9/patient9_omentum_site_2.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_9/patient9_right_ovary_site_1.tsv --working_dir /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_9 --tumour_contents 0.761987 0.654521 0.178824 0.672038 0.601017 --samples left_ovary_site_1 left_ovary_site_2 omentum_site_1 omentum_site_2 right_ovary_site_1 --num_iters 100000 --burnin 50000\n",
      "bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_10.log -e error_10.log PyClone run_analysis_pipeline --in_files /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_10/patient10_right_ovary_site_1.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_10/patient10_right_ovary_site_2.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_10/patient10_right_ovary_site_3.tsv /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_10/patient10_right_ovary_site_4.tsv --working_dir /lila/data/morrisq/divyak/projects/metient/metient/jupyter_notebooks/../data/mcpherson_ovarian_2016/pyclone_analysis/patient_10 --tumour_contents 0.445714 0.696471 0.270729 0.323749 --samples right_ovary_site_1 right_ovary_site_2 right_ovary_site_3 right_ovary_site_4 --num_iters 100000 --burnin 50000\n"
     ]
    }
   ],
   "source": [
    "# Setup PyClone commands to run for each patient\n",
    "\n",
    "for patient_id in patient_ids:\n",
    "    cmd = [f\"bsub -n 8 -W 20:00 -R rusage[mem=8] -o output_{patient_id}.log -e error_{patient_id}.log\", \"PyClone run_analysis_pipeline\", \"--in_files\"]\n",
    "    patient_dir = os.path.join(pyclone_dir, f\"patient_{patient_id}\")\n",
    "    discovery_samples = list(discovery_samples_df[discovery_samples_df['patient_id']==patient_id]['sample_id'])\n",
    "    discovery_samples.remove(\"normal_blood\")\n",
    "    # Add tsvs for each sample\n",
    "    for sample_id in discovery_samples:\n",
    "        cmd.append(os.path.join(patient_dir,  f\"patient{patient_id}_{sample_id}.tsv\"))\n",
    "    cmd += [\"--working_dir\", patient_dir, \"--tumour_contents\"]\n",
    "    # Add tumour cell proportions for each sample\n",
    "    for sample_id in discovery_samples:\n",
    "        tumour_prop_subset = discovery_samples_df[(discovery_samples_df['patient_id']==patient_id) & (discovery_samples_df['sample_id']==sample_id)]\n",
    "        tumour_prop = tumour_prop_subset['tumour_cell_proportion']                                                                                            \n",
    "        cmd.append(str(tumour_prop.values[0]))\n",
    "    # Add sample names\n",
    "    cmd.append(\"--samples\")\n",
    "    for sample_id in discovery_samples:\n",
    "        cmd.append(sample_id)\n",
    "    cmd += [\"--num_iters\", \"100000\", \"--burnin\", \"50000\"]\n",
    "    print(\" \".join(cmd))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "163ff544",
   "metadata": {},
   "source": [
    "## (3) Generate inputs for tree inference algorithms (PairTree/Orchard)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59085513",
   "metadata": {},
   "source": [
    "## (3a) Create input for PairTree/Orchard (using PyClone clustering)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0c834a4e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/lila/home/koyyald/mambaforge/envs/met/lib/python3.8/site-packages/torch/cuda/__init__.py:138: UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 11030). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:108.)\n",
      "  return torch._C._cuda_getDeviceCount() > 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CUDA GPU: False\n",
      "13 clusters 177 mutations\n",
      "6 clusters 120 mutations\n",
      "9 clusters 168 mutations\n",
      "10 clusters 161 mutations\n",
      "8 clusters 221 mutations\n",
      "5 clusters 132 mutations\n",
      "3 clusters 189 mutations\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from metient.util import data_extraction_util as dutil\n",
    "\n",
    "MIN_MUT_THRES = 5\n",
    "pairtree_dir = os.path.join(data_dir, \"orchard_trees\")\n",
    "header = [\"id\", \"name\", \"var_reads\", \"total_reads\", \"var_read_prob\"]\n",
    "\n",
    "# id, name, var_reads, total_reads, var_read_prob\n",
    "for patient_id in patient_ids:\n",
    "    # Use the CNA reads df since this has filtered out the variants without CN information\n",
    "    # and was used as input to PyClone, which our clusters are made from\n",
    "    patient_subset = cna_reads_df[cna_reads_df['patient_id'] == patient_id]\n",
    "    # only keep the high quality discovery samples used in the paper\n",
    "    discovery_samples = discovery_samples_df[discovery_samples_df['patient_id']==patient_id]['sample_id']\n",
    "    patient_subset = patient_subset[patient_subset['sample_id'].isin(discovery_samples)]\n",
    "    # remove normal samples\n",
    "    patient_subset = patient_subset[patient_subset['sample_id'] != 'normal_blood'] \n",
    "    \n",
    "    patient_pyclone_dir = os.path.join(pyclone_dir, f\"patient_{patient_id}\", \"tables\", f\"loci.tsv\")\n",
    "    mut_name_to_clstr_id, clstr_id_to_name, mutation_names = dutil.get_mut_to_cluster_map_from_pyclone_output(patient_pyclone_dir, min_mut_thres=MIN_MUT_THRES)\n",
    "    cluster_id_to_mut_names = {}\n",
    "    for mut in mut_name_to_clstr_id:\n",
    "        cid = mut_name_to_clstr_id[mut]\n",
    "        if cid not in cluster_id_to_mut_names:\n",
    "            cluster_id_to_mut_names[cid]= []\n",
    "        cluster_id_to_mut_names[cid].append(mut)\n",
    "   \n",
    "    sample_names = list(patient_subset['sample_id'].unique())\n",
    "    print(len(cluster_id_to_mut_names), \"clusters\", len(mutation_names), \"mutations\")\n",
    "    #print(mutation_names)\n",
    "    mut_name_to_mut_id = {}\n",
    "    with open(os.path.join(pairtree_dir, f\"patient{patient_id}.ssm\"), \"w\") as f:\n",
    "        \n",
    "        f.write(\"\\t\".join(header))\n",
    "        f.write(\"\\n\")\n",
    "        for i, mut in enumerate(mutation_names):\n",
    "            mut_name_to_mut_id[mut] = f\"m{i}\"\n",
    "            row = [f\"m{i}\", mut]\n",
    "            mut_patient_subset = patient_subset[patient_subset['character_label'] == mut]\n",
    "            #print(mut_patient_subset)\n",
    "            var_reads = []\n",
    "            total_reads = []\n",
    "            var_read_probs = []\n",
    "            for sample in sample_names:\n",
    "                #print(patient_id, mut, sample)\n",
    "                mut_patient_sample = mut_patient_subset[mut_patient_subset['sample_id'] == sample]\n",
    "                #print(mut_patient_sample)\n",
    "                var = mut_patient_sample['alt_counts'].values[0]\n",
    "                ref = mut_patient_sample['ref_counts'].values[0]\n",
    "                var_reads.append(str(var))\n",
    "                total_reads.append(str(var+ref))\n",
    "                tumour_prop_subset = discovery_samples_df[(discovery_samples_df['patient_id']==patient_id) & (discovery_samples_df['sample_id']==sample)]\n",
    "                p = tumour_prop_subset['tumour_cell_proportion'].item() \n",
    "                major_cn = mut_patient_sample['major_cn'].values[0]\n",
    "                minor_cn = mut_patient_sample['minor_cn'].values[0]\n",
    "                var_read_prob = dutil.calc_var_read_prob(major_cn, minor_cn, p)\n",
    "                var_read_probs.append(str(var_read_prob))\n",
    "\n",
    "            row += [\",\".join(var_reads), \",\".join(total_reads), \",\".join(var_read_probs)]\n",
    "            f.write(\"\\t\".join(row))\n",
    "            f.write(\"\\n\")\n",
    "    json_data = {\"samples\": sample_names, \"clusters\": [], \"garbage\": []}\n",
    "    for x in range(0,len(cluster_id_to_mut_names)):\n",
    "        json_data[\"clusters\"].append([mut_name_to_mut_id[t] for t in cluster_id_to_mut_names[x]])\n",
    "    \n",
    "    with open(os.path.join(pairtree_dir, f\"patient{patient_id}.params.json\"), 'w', encoding='utf-8') as f:\n",
    "        json.dump(json_data, f, ensure_ascii=False)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3c4a70",
   "metadata": {},
   "source": [
    "## (4) Prepare data for migration history inference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5be21d1",
   "metadata": {},
   "source": [
    "## (4a) Take PyClone generated clusters and create csvs with ref and var counts pooled by cluster "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9d3ac3fe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient 1 13 clusters 177 mutations 6 samples\n",
      "['omentum_site_1', 'right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3', 'right_ovary_site_4', 'small_bowel_site_1']\n",
      "['omentum_site', 'small_bowel_site', 'right_ovary_site']\n",
      "patient 2 6 clusters 120 mutations 4 samples\n",
      "['omentum_site_1', 'omentum_site_2', 'right_ovary_site_1', 'right_ovary_site_2']\n",
      "['omentum_site', 'right_ovary_site']\n",
      "patient 3 9 clusters 168 mutations 4 samples\n",
      "['adnexa_site_1', 'omentum_site_1', 'right_ovary_site_1', 'right_ovary_site_2']\n",
      "['omentum_site', 'adnexa_site', 'right_ovary_site']\n",
      "patient 4 10 clusters 161 mutations 5 samples\n",
      "['left_pelvic_sidewall_site_1', 'right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3', 'right_ovary_site_4']\n",
      "['right_ovary_site', 'left_pelvic_sidewall_site']\n",
      "patient 7 8 clusters 221 mutations 3 samples\n",
      "['brain_metastasis', 'left_ovary_site_1', 'righ_pelvic_mass']\n",
      "['righ_pelvic_mass', 'left_ovary_site', 'brain_metastasis']\n",
      "patient 9 5 clusters 132 mutations 5 samples\n",
      "['left_ovary_site_1', 'left_ovary_site_2', 'omentum_site_1', 'omentum_site_2', 'right_ovary_site_1']\n",
      "['omentum_site', 'left_ovary_site', 'right_ovary_site']\n",
      "patient 10 3 clusters 189 mutations 4 samples\n",
      "['right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3', 'right_ovary_site_4']\n",
      "['right_ovary_site']\n"
     ]
    }
   ],
   "source": [
    "# Need a tsv for each patient with ['#sample_index', 'sample_label', 'anatomical_site_index','anatomical_site_label', 'character_index', 'character_label', 'ref', 'var']\n",
    "import re\n",
    "import numpy as np\n",
    "from metient.util import data_extraction_util as dutil\n",
    "pyclone_dir = os.path.join(data_dir, \"pyclone_analysis\")\n",
    "\n",
    "\n",
    "cols = [ 'anatomical_site_index','anatomical_site_label',  'cluster_index', 'character_index', 'character_label', 'ref', 'var', 'var_read_prob', 'site_category']\n",
    "#agg_rules = {'sample_label': lambda x: ';'.join(set(x))}\n",
    "\n",
    "for patient_id in patient_ids:\n",
    "    patient_pyclone_dir = os.path.join(pyclone_dir, f\"patient_{patient_id}\", \"tables\", f\"loci.tsv\")\n",
    "    mut_name_to_clstr_id, clstr_id_to_name, mutation_names = dutil.get_mut_to_cluster_map_from_pyclone_output(patient_pyclone_dir, min_mut_thres=MIN_MUT_THRES)\n",
    "  \n",
    "    patient_subset = cna_reads_df[cna_reads_df['patient_id'] == patient_id]\n",
    "    discovery_samples = discovery_samples_df[discovery_samples_df['patient_id']==patient_id]['sample_id']\n",
    "    patient_subset = patient_subset[(patient_subset['sample_id'].isin(discovery_samples)) & (patient_subset['sample_id'] != 'normal_blood')]   \n",
    "    sample_names = list(patient_subset['sample_id'].unique())\n",
    "    print(\"patient\", patient_id, len(clstr_id_to_name), \"clusters\", len(mutation_names), \"mutations\", len(sample_names), \"samples\")\n",
    "    print(sample_names)\n",
    "    data = []\n",
    "    \n",
    "    anat_sites = list(set([re.sub(r'_[0-9]+$', '', sample) for sample in sample_names]))\n",
    "    print(anat_sites)\n",
    "    for midx, mut in enumerate(mutation_names):\n",
    "        mut_patient_subset = patient_subset[patient_subset['character_label'] == mut]\n",
    "        for sidx, sample in enumerate(sample_names):\n",
    "            mut_patient_sample = mut_patient_subset[mut_patient_subset['sample_id'] == sample]\n",
    "            var = mut_patient_sample['alt_counts'].values[0]\n",
    "            ref = mut_patient_sample['ref_counts'].values[0]\n",
    "            tumour_prop_subset = discovery_samples_df[(discovery_samples_df['patient_id']==patient_id) & (discovery_samples_df['sample_id']==sample)]\n",
    "            p = tumour_prop_subset['tumour_cell_proportion'].item() \n",
    "            major_cn = mut_patient_sample['major_cn'].values[0]\n",
    "            minor_cn = mut_patient_sample['minor_cn'].values[0]\n",
    "            var_read_prob = dutil.calc_var_read_prob(major_cn, minor_cn, p)\n",
    "            \n",
    "            site_label = re.sub(r'_[0-9]+$', '', sample)\n",
    "            #print(sample, site_label, anat_sites.index(site_label))\n",
    "            category = 'primary' if 'ovary_site' in site_label else 'metastasis'\n",
    "            data.append([anat_sites.index(site_label), site_label, mut_name_to_clstr_id[mut], midx, mut.split(\":\")[0], ref, var, var_read_prob, category])\n",
    "            \n",
    "    patient_df = pd.DataFrame(data, columns=cols)\n",
    "    patient_df.to_csv(os.path.join(data_dir,\"pyclone_clustered_tsvs\", f\"{patient_id}_SNVs.tsv\"), sep=\"\\t\", index=False)\n",
    "\n",
    "#     dutil.write_pooled_tsv_from_clusters(patient_df, mut_name_to_clstr_id, clstr_id_to_name, agg_rules, \n",
    "#                                          os.path.join(data_dir,\"pyclone_clustered_tsvs\"), f\"patient{patient_id}\", \";\", \":\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4962e0c5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:met]",
   "language": "python",
   "name": "conda-env-met-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
