{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parse A1\n",
    "## Adapted from MACHNA (El-Kebir et. al.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "confidence=0.95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_id(line):\n",
    "    return \"_\".join(map(str, [line['chromosome_name'], line['start'], line['stop']]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome_name', 'start', 'stop', 'reference', 'variant', 'type',\n",
       "       'gene_name', 'transcript_name', 'transcript_species',\n",
       "       'transcript_source', 'transcript_version', 'strand',\n",
       "       'transcript_status', 'trv_type', 'c_position', 'amino_acid_change',\n",
       "       'ucsc_cons', 'domain', 'all_domains', 'deletion_substructures',\n",
       "       'transcript_error', 'adrenalmet.rcnt.llr3_ref',\n",
       "       'adrenalmet.rcnt.llr3_var', 'adrenalmet.rcnt.llr3_VAF',\n",
       "       'livermet.rcnt.llr3_ref', 'livermet.rcnt.llr3_var',\n",
       "       'livermet.rcnt.llr3_VAF', 'lungmet.rcnt.llr3_ref',\n",
       "       'lungmet.rcnt.llr3_var', 'lungmet.rcnt.llr3_VAF',\n",
       "       'spinalmet.rcnt.llr3_ref', 'spinalmet.rcnt.llr3_var',\n",
       "       'spinalmet.rcnt.llr3_VAF', 'tumor.rcnt.llr3_ref', 'tumor.rcnt.llr3_var',\n",
       "       'tumor.rcnt.llr3_VAF', 'cluster'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "HOADLEY_DATA_DIR = '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/data/hoadley_breast_cancer_2016/'\n",
    "table = pd.read_table(os.path.join(HOADLEY_DATA_DIR,\"A1/A1_raw.tsv\"))\n",
    "table['id']=table.apply(get_id, axis=1)\n",
    "table = table.set_index('id')\n",
    "table.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def label_snv(row):\n",
    "    label = []\n",
    "    if row['gene_name'] != \"-\":\n",
    "        label.append(row['gene_name'])\n",
    "    label += [str(row['chromosome_name']), str(row['start']), str(row['stop'])]\n",
    "    return (\":\").join(label)\n",
    "table['character_label'] = table.apply(lambda row: label_snv(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_species</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>...</th>\n",
       "      <th>lungmet.rcnt.llr3_var</th>\n",
       "      <th>lungmet.rcnt.llr3_VAF</th>\n",
       "      <th>spinalmet.rcnt.llr3_ref</th>\n",
       "      <th>spinalmet.rcnt.llr3_var</th>\n",
       "      <th>spinalmet.rcnt.llr3_VAF</th>\n",
       "      <th>tumor.rcnt.llr3_ref</th>\n",
       "      <th>tumor.rcnt.llr3_var</th>\n",
       "      <th>tumor.rcnt.llr3_VAF</th>\n",
       "      <th>cluster</th>\n",
       "      <th>character_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_110359957_110359957</th>\n",
       "      <td>1</td>\n",
       "      <td>110359957</td>\n",
       "      <td>110359957</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AHCYL1</td>\n",
       "      <td>NM_006621</td>\n",
       "      <td>human</td>\n",
       "      <td>genbank</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>142</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>150</td>\n",
       "      <td>4</td>\n",
       "      <td>2.58</td>\n",
       "      <td>7</td>\n",
       "      <td>AHCYL1:1:110359957:110359957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_48902663_48902663</th>\n",
       "      <td>6</td>\n",
       "      <td>48902663</td>\n",
       "      <td>48902663</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>ENSG00000221175</td>\n",
       "      <td>ENST00000408248</td>\n",
       "      <td>human</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0.68</td>\n",
       "      <td>178</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>189</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000221175:6:48902663:48902663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_49276467_49276467</th>\n",
       "      <td>6</td>\n",
       "      <td>49276467</td>\n",
       "      <td>49276467</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>63</td>\n",
       "      <td>17</td>\n",
       "      <td>21.25</td>\n",
       "      <td>117</td>\n",
       "      <td>1</td>\n",
       "      <td>0.85</td>\n",
       "      <td>9</td>\n",
       "      <td>6:49276467:49276467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_50173697_50173697</th>\n",
       "      <td>6</td>\n",
       "      <td>50173697</td>\n",
       "      <td>50173697</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>DEFB112</td>\n",
       "      <td>NM_001037498</td>\n",
       "      <td>human</td>\n",
       "      <td>genbank</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>115</td>\n",
       "      <td>30</td>\n",
       "      <td>20.69</td>\n",
       "      <td>179</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9</td>\n",
       "      <td>DEFB112:6:50173697:50173697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6_50179508_50179508</th>\n",
       "      <td>6</td>\n",
       "      <td>50179508</td>\n",
       "      <td>50179508</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>128</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>141</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6</td>\n",
       "      <td>6:50179508:50179508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chromosome_name      start       stop reference  \\\n",
       "id                                                                       \n",
       "1_110359957_110359957                1  110359957  110359957         C   \n",
       "6_48902663_48902663                  6   48902663   48902663         T   \n",
       "6_49276467_49276467                  6   49276467   49276467         G   \n",
       "6_50173697_50173697                  6   50173697   50173697         C   \n",
       "6_50179508_50179508                  6   50179508   50179508         A   \n",
       "\n",
       "                      variant type        gene_name  transcript_name  \\\n",
       "id                                                                     \n",
       "1_110359957_110359957       T  SNP           AHCYL1        NM_006621   \n",
       "6_48902663_48902663         G  SNP  ENSG00000221175  ENST00000408248   \n",
       "6_49276467_49276467         C  SNP                -                -   \n",
       "6_50173697_50173697         G  SNP          DEFB112     NM_001037498   \n",
       "6_50179508_50179508         T  SNP                -                -   \n",
       "\n",
       "                      transcript_species transcript_source  ...  \\\n",
       "id                                                          ...   \n",
       "1_110359957_110359957              human           genbank  ...   \n",
       "6_48902663_48902663                human           ensembl  ...   \n",
       "6_49276467_49276467                    -                 -  ...   \n",
       "6_50173697_50173697                human           genbank  ...   \n",
       "6_50179508_50179508                    -                 -  ...   \n",
       "\n",
       "                      lungmet.rcnt.llr3_var lungmet.rcnt.llr3_VAF  \\\n",
       "id                                                                  \n",
       "1_110359957_110359957                     0                  0.00   \n",
       "6_48902663_48902663                       1                  0.68   \n",
       "6_49276467_49276467                       0                  0.00   \n",
       "6_50173697_50173697                       0                  0.00   \n",
       "6_50179508_50179508                       0                  0.00   \n",
       "\n",
       "                      spinalmet.rcnt.llr3_ref spinalmet.rcnt.llr3_var  \\\n",
       "id                                                                      \n",
       "1_110359957_110359957                     142                       0   \n",
       "6_48902663_48902663                       178                       0   \n",
       "6_49276467_49276467                        63                      17   \n",
       "6_50173697_50173697                       115                      30   \n",
       "6_50179508_50179508                       128                       0   \n",
       "\n",
       "                      spinalmet.rcnt.llr3_VAF tumor.rcnt.llr3_ref  \\\n",
       "id                                                                  \n",
       "1_110359957_110359957                    0.00                 150   \n",
       "6_48902663_48902663                      0.00                 189   \n",
       "6_49276467_49276467                     21.25                 117   \n",
       "6_50173697_50173697                     20.69                 179   \n",
       "6_50179508_50179508                      0.00                 141   \n",
       "\n",
       "                      tumor.rcnt.llr3_var tumor.rcnt.llr3_VAF cluster  \\\n",
       "id                                                                      \n",
       "1_110359957_110359957                   4                2.58       7   \n",
       "6_48902663_48902663                     0                0.00       6   \n",
       "6_49276467_49276467                     1                0.85       9   \n",
       "6_50173697_50173697                     0                0.00       9   \n",
       "6_50179508_50179508                     0                0.00       6   \n",
       "\n",
       "                                           character_label  \n",
       "id                                                          \n",
       "1_110359957_110359957         AHCYL1:1:110359957:110359957  \n",
       "6_48902663_48902663    ENSG00000221175:6:48902663:48902663  \n",
       "6_49276467_49276467                    6:49276467:49276467  \n",
       "6_50173697_50173697            DEFB112:6:50173697:50173697  \n",
       "6_50179508_50179508                    6:50179508:50179508  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_table = table.copy()\n",
    "raw_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate migration history input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_cols = ['tumor.rcnt.llr3_ref','adrenalmet.rcnt.llr3_ref', 'livermet.rcnt.llr3_ref', 'lungmet.rcnt.llr3_ref', 'spinalmet.rcnt.llr3_ref']\n",
    "var_cols = ['tumor.rcnt.llr3_var','adrenalmet.rcnt.llr3_var', 'livermet.rcnt.llr3_var', 'lungmet.rcnt.llr3_var', 'spinalmet.rcnt.llr3_var']\n",
    "\n",
    "#breast,adrenal,liver,lung,spinal\n",
    "cols = ['breast', 'adrenal', 'liver', 'lung', 'spinal']\n",
    "table = table[['cluster']+ref_cols+var_cols]\n",
    "table.columns = ['cluster']+['ref-'+c for c in cols] + ['var-'+c for c in cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9988888888888889\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref-breast</th>\n",
       "      <th>ref-adrenal</th>\n",
       "      <th>ref-liver</th>\n",
       "      <th>ref-lung</th>\n",
       "      <th>ref-spinal</th>\n",
       "      <th>var-breast</th>\n",
       "      <th>var-adrenal</th>\n",
       "      <th>var-liver</th>\n",
       "      <th>var-lung</th>\n",
       "      <th>var-spinal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>657</td>\n",
       "      <td>571</td>\n",
       "      <td>551</td>\n",
       "      <td>612</td>\n",
       "      <td>698</td>\n",
       "      <td>267</td>\n",
       "      <td>307</td>\n",
       "      <td>437</td>\n",
       "      <td>147</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5458</td>\n",
       "      <td>3263</td>\n",
       "      <td>2850</td>\n",
       "      <td>4572</td>\n",
       "      <td>5357</td>\n",
       "      <td>322</td>\n",
       "      <td>2287</td>\n",
       "      <td>2256</td>\n",
       "      <td>70</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6948</td>\n",
       "      <td>8762</td>\n",
       "      <td>8990</td>\n",
       "      <td>6248</td>\n",
       "      <td>6861</td>\n",
       "      <td>2070</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>1092</td>\n",
       "      <td>1663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14589</td>\n",
       "      <td>10980</td>\n",
       "      <td>10950</td>\n",
       "      <td>12169</td>\n",
       "      <td>14195</td>\n",
       "      <td>486</td>\n",
       "      <td>3274</td>\n",
       "      <td>3625</td>\n",
       "      <td>47</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4613</td>\n",
       "      <td>5563</td>\n",
       "      <td>6004</td>\n",
       "      <td>4323</td>\n",
       "      <td>5600</td>\n",
       "      <td>1337</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>720</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4255</td>\n",
       "      <td>3175</td>\n",
       "      <td>3715</td>\n",
       "      <td>3465</td>\n",
       "      <td>4020</td>\n",
       "      <td>2</td>\n",
       "      <td>884</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5529</td>\n",
       "      <td>5354</td>\n",
       "      <td>4489</td>\n",
       "      <td>4693</td>\n",
       "      <td>5212</td>\n",
       "      <td>152</td>\n",
       "      <td>3</td>\n",
       "      <td>1256</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2156</td>\n",
       "      <td>2528</td>\n",
       "      <td>3007</td>\n",
       "      <td>2144</td>\n",
       "      <td>2608</td>\n",
       "      <td>543</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>12221</td>\n",
       "      <td>11841</td>\n",
       "      <td>11226</td>\n",
       "      <td>9962</td>\n",
       "      <td>9242</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>2129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ref-breast  ref-adrenal  ref-liver  ref-lung  ref-spinal  var-breast  \\\n",
       "cluster                                                                         \n",
       "1               657          571        551       612         698         267   \n",
       "2              5458         3263       2850      4572        5357         322   \n",
       "3              6948         8762       8990      6248        6861        2070   \n",
       "4             14589        10980      10950     12169       14195         486   \n",
       "5              4613         5563       6004      4323        5600        1337   \n",
       "6              4255         3175       3715      3465        4020           2   \n",
       "7              5529         5354       4489      4693        5212         152   \n",
       "8              2156         2528       3007      2144        2608         543   \n",
       "9             12221        11841      11226      9962        9242          10   \n",
       "\n",
       "         var-adrenal  var-liver  var-lung  var-spinal  \n",
       "cluster                                                \n",
       "1                307        437       147         170  \n",
       "2               2287       2256        70          12  \n",
       "3                  7          8      1092        1663  \n",
       "4               3274       3625        47          10  \n",
       "5                  8         11       720          26  \n",
       "6                884          5         4           2  \n",
       "7                  3       1256         4           0  \n",
       "8                  1          1        13           0  \n",
       "9                  8          3        12        2129  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ctable = table.groupby('cluster').sum()\n",
    "\n",
    "global corrected_confidence\n",
    "nsamples = len([c for c in ctable.columns if c.startswith('ref')])\n",
    "nclusters = len(ctable)\n",
    "corrected_confidence = 1-((1.-confidence)/(nsamples*nclusters))\n",
    "print(corrected_confidence)\n",
    "\n",
    "assert(corrected_confidence > confidence)\n",
    "assert(corrected_confidence < 1.0)\n",
    "ctable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "from scipy.stats import beta\n",
    "from scipy.stats import norm\n",
    "\n",
    "def binomial_hpdr(n, N, pct, a=1, b=1, n_pbins=1e3):\n",
    "    \"\"\"\n",
    "    Function computes the posterior mode along with the upper and lower bounds of the\n",
    "    **Highest Posterior Density Region**.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n: number of successes \n",
    "    N: sample size \n",
    "    pct: the size of the confidence interval (between 0 and 1)\n",
    "    a: the alpha hyper-parameter for the Beta distribution used as a prior (Default=1)\n",
    "    b: the beta hyper-parameter for the Beta distribution used as a prior (Default=1)\n",
    "    n_pbins: the number of bins to segment the p_range into (Default=1e3)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    A tuple that contains the mode as well as the lower and upper bounds of the interval\n",
    "    (mode, lower, upper)\n",
    "\n",
    "    \"\"\"\n",
    "    # fixed random variable object for posterior Beta distribution\n",
    "    rv = beta(n+a, N-n+b)\n",
    "    # determine the mode and standard deviation of the posterior\n",
    "    stdev = rv.stats('v')**0.5\n",
    "    mode = (n+a-1.)/(N+a+b-2.)\n",
    "    # compute the number of sigma that corresponds to this confidence\n",
    "    # this is used to set the rough range of possible success probabilities\n",
    "    n_sigma = numpy.ceil(norm.ppf( (1+pct)/2. ))+1\n",
    "    # set the min and max values for success probability \n",
    "    max_p = mode + n_sigma * stdev\n",
    "    if max_p > 1:\n",
    "        max_p = 1.\n",
    "    min_p = mode - n_sigma * stdev\n",
    "    if min_p > 1:\n",
    "        min_p = 1.\n",
    "    # make the range of success probabilities\n",
    "    p_range = numpy.linspace(min_p, max_p, int(n_pbins+1))\n",
    "    # construct the probability mass function over the given range\n",
    "    if mode > 0.5:\n",
    "        sf = rv.sf(p_range)\n",
    "        pmf = sf[:-1] - sf[1:]\n",
    "    else:\n",
    "        cdf = rv.cdf(p_range)\n",
    "        pmf = cdf[1:] - cdf[:-1]\n",
    "    # find the upper and lower bounds of the interval \n",
    "    sorted_idxs = numpy.argsort( pmf )[::-1]\n",
    "    cumsum = numpy.cumsum( numpy.sort(pmf)[::-1] )\n",
    "    j = numpy.argmin( numpy.abs(cumsum - pct) )\n",
    "    upper = p_range[ (sorted_idxs[:j+1]).max()+1 ]\n",
    "    lower = p_range[ (sorted_idxs[:j+1]).min() ]    \n",
    "\n",
    "    return (mode, lower, upper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ub(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    return v[2]\n",
    "    \n",
    "\n",
    "def get_lb(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[1]\n",
    "    #if mval < 0.01: mval = 0\n",
    "    return mval\n",
    "\n",
    "def get_mean(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[0]\n",
    "    return mval\n",
    "\n",
    "ctable = table.groupby('cluster').sum()\n",
    "\n",
    "for sam in cols:\n",
    "    ctable['ub-'+sam]= ctable.apply(get_ub, args=[sam], axis=1)\n",
    "    ctable['lb-'+sam]= ctable.apply(get_lb, args=[sam], axis=1)\n",
    "    ctable[sam]= ctable.apply(get_mean, args=[sam], axis=1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ub(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    return v[2]\n",
    "    \n",
    "\n",
    "def get_lb(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[1]\n",
    "    if mval < 0.01: mval = 0\n",
    "    return mval\n",
    "\n",
    "def get_mean(row, sam):\n",
    "    v=binomial_hpdr(row['var-'+sam], row['var-'+sam]+row['ref-'+sam], corrected_confidence)\n",
    "    mval = v[0]\n",
    "    return mval\n",
    "\n",
    "ctable_cutoff = table.groupby('cluster').sum()\n",
    "for sam in cols:\n",
    "    ctable_cutoff['ub-'+sam]= ctable.apply(get_ub, args=[sam], axis=1)\n",
    "    ctable_cutoff['lb-'+sam]= ctable.apply(get_lb, args=[sam], axis=1)\n",
    "    ctable_cutoff[sam]= ctable.apply(get_mean, args=[sam], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vaf(row, sam):\n",
    "    return float(row['var-'+sam])/float(row['var-'+sam]+row['ref-'+sam])\n",
    "\n",
    "vafs = pd.DataFrame()\n",
    "for sam in cols:\n",
    "    vafs[sam] = table.apply(get_vaf, args=[sam], axis=1)\n",
    "vafs['cluster'] = table['cluster']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref-breast</th>\n",
       "      <th>ref-adrenal</th>\n",
       "      <th>ref-liver</th>\n",
       "      <th>ref-lung</th>\n",
       "      <th>ref-spinal</th>\n",
       "      <th>var-breast</th>\n",
       "      <th>var-adrenal</th>\n",
       "      <th>var-liver</th>\n",
       "      <th>var-lung</th>\n",
       "      <th>var-spinal</th>\n",
       "      <th>...</th>\n",
       "      <th>ub-liver</th>\n",
       "      <th>lb-liver</th>\n",
       "      <th>liver</th>\n",
       "      <th>ub-lung</th>\n",
       "      <th>lb-lung</th>\n",
       "      <th>lung</th>\n",
       "      <th>ub-spinal</th>\n",
       "      <th>lb-spinal</th>\n",
       "      <th>spinal</th>\n",
       "      <th>character_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>657</td>\n",
       "      <td>571</td>\n",
       "      <td>551</td>\n",
       "      <td>612</td>\n",
       "      <td>698</td>\n",
       "      <td>267</td>\n",
       "      <td>307</td>\n",
       "      <td>437</td>\n",
       "      <td>147</td>\n",
       "      <td>170</td>\n",
       "      <td>...</td>\n",
       "      <td>0.494057</td>\n",
       "      <td>0.391347</td>\n",
       "      <td>0.442308</td>\n",
       "      <td>0.243143</td>\n",
       "      <td>0.149944</td>\n",
       "      <td>0.193676</td>\n",
       "      <td>0.242172</td>\n",
       "      <td>0.154515</td>\n",
       "      <td>0.195853</td>\n",
       "      <td>INVS:9:101950362:101950362_9:105825009:1058250...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5458</td>\n",
       "      <td>3263</td>\n",
       "      <td>2850</td>\n",
       "      <td>4572</td>\n",
       "      <td>5357</td>\n",
       "      <td>322</td>\n",
       "      <td>2287</td>\n",
       "      <td>2256</td>\n",
       "      <td>70</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>0.464552</td>\n",
       "      <td>0.419253</td>\n",
       "      <td>0.441833</td>\n",
       "      <td>0.021687</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015080</td>\n",
       "      <td>0.005058</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002235</td>\n",
       "      <td>6:50519104:50519104_ENSG00000214604:6:55918529...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6948</td>\n",
       "      <td>8762</td>\n",
       "      <td>8990</td>\n",
       "      <td>6248</td>\n",
       "      <td>6861</td>\n",
       "      <td>2070</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>1092</td>\n",
       "      <td>1663</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002358</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000889</td>\n",
       "      <td>0.162648</td>\n",
       "      <td>0.135564</td>\n",
       "      <td>0.148774</td>\n",
       "      <td>0.209346</td>\n",
       "      <td>0.181362</td>\n",
       "      <td>0.195096</td>\n",
       "      <td>6:50534952:50534952_RAB23:6:57180847:57180847_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14589</td>\n",
       "      <td>10980</td>\n",
       "      <td>10950</td>\n",
       "      <td>12169</td>\n",
       "      <td>14195</td>\n",
       "      <td>486</td>\n",
       "      <td>3274</td>\n",
       "      <td>3625</td>\n",
       "      <td>47</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0.260493</td>\n",
       "      <td>0.237149</td>\n",
       "      <td>0.248714</td>\n",
       "      <td>0.005975</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003847</td>\n",
       "      <td>0.001707</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000704</td>\n",
       "      <td>ICK:6:53035144:53035144_LOC727842:6:58602457:5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4613</td>\n",
       "      <td>5563</td>\n",
       "      <td>6004</td>\n",
       "      <td>4323</td>\n",
       "      <td>5600</td>\n",
       "      <td>1337</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>720</td>\n",
       "      <td>26</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004273</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001829</td>\n",
       "      <td>0.159327</td>\n",
       "      <td>0.127203</td>\n",
       "      <td>0.142772</td>\n",
       "      <td>0.008232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004621</td>\n",
       "      <td>6:50568504:50568504_6:99144022:99144022_ROS1:6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4255</td>\n",
       "      <td>3175</td>\n",
       "      <td>3715</td>\n",
       "      <td>3465</td>\n",
       "      <td>4020</td>\n",
       "      <td>2</td>\n",
       "      <td>884</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004415</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001344</td>\n",
       "      <td>0.004243</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001153</td>\n",
       "      <td>0.002648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000497</td>\n",
       "      <td>ENSG00000221175:6:48902663:48902663_6:50179508...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5529</td>\n",
       "      <td>5354</td>\n",
       "      <td>4489</td>\n",
       "      <td>4693</td>\n",
       "      <td>5212</td>\n",
       "      <td>152</td>\n",
       "      <td>3</td>\n",
       "      <td>1256</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.236727</td>\n",
       "      <td>0.201177</td>\n",
       "      <td>0.218625</td>\n",
       "      <td>0.003134</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000852</td>\n",
       "      <td>0.000959</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>AHCYL1:1:110359957:110359957_6:97912586:979125...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2156</td>\n",
       "      <td>2528</td>\n",
       "      <td>3007</td>\n",
       "      <td>2144</td>\n",
       "      <td>2608</td>\n",
       "      <td>543</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002680</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000332</td>\n",
       "      <td>0.013246</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006027</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>TFAP2B:6:50909762:50909762_ENPP3:6:132075413:1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>12221</td>\n",
       "      <td>11841</td>\n",
       "      <td>11226</td>\n",
       "      <td>9962</td>\n",
       "      <td>9242</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>2129</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001156</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.002727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001203</td>\n",
       "      <td>0.199339</td>\n",
       "      <td>0.175488</td>\n",
       "      <td>0.187231</td>\n",
       "      <td>6:49276467:49276467_DEFB112:6:50173697:5017369...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ref-breast  ref-adrenal  ref-liver  ref-lung  ref-spinal  var-breast  \\\n",
       "cluster                                                                         \n",
       "1               657          571        551       612         698         267   \n",
       "2              5458         3263       2850      4572        5357         322   \n",
       "3              6948         8762       8990      6248        6861        2070   \n",
       "4             14589        10980      10950     12169       14195         486   \n",
       "5              4613         5563       6004      4323        5600        1337   \n",
       "6              4255         3175       3715      3465        4020           2   \n",
       "7              5529         5354       4489      4693        5212         152   \n",
       "8              2156         2528       3007      2144        2608         543   \n",
       "9             12221        11841      11226      9962        9242          10   \n",
       "\n",
       "         var-adrenal  var-liver  var-lung  var-spinal  ...  ub-liver  \\\n",
       "cluster                                                ...             \n",
       "1                307        437       147         170  ...  0.494057   \n",
       "2               2287       2256        70          12  ...  0.464552   \n",
       "3                  7          8      1092        1663  ...  0.002358   \n",
       "4               3274       3625        47          10  ...  0.260493   \n",
       "5                  8         11       720          26  ...  0.004273   \n",
       "6                884          5         4           2  ...  0.004415   \n",
       "7                  3       1256         4           0  ...  0.236727   \n",
       "8                  1          1        13           0  ...  0.002680   \n",
       "9                  8          3        12        2129  ...  0.001156   \n",
       "\n",
       "         lb-liver     liver   ub-lung   lb-lung      lung  ub-spinal  \\\n",
       "cluster                                                                \n",
       "1        0.391347  0.442308  0.243143  0.149944  0.193676   0.242172   \n",
       "2        0.419253  0.441833  0.021687  0.000000  0.015080   0.005058   \n",
       "3        0.000000  0.000889  0.162648  0.135564  0.148774   0.209346   \n",
       "4        0.237149  0.248714  0.005975  0.000000  0.003847   0.001707   \n",
       "5        0.000000  0.001829  0.159327  0.127203  0.142772   0.008232   \n",
       "6        0.000000  0.001344  0.004243  0.000000  0.001153   0.002648   \n",
       "7        0.201177  0.218625  0.003134  0.000000  0.000852   0.000959   \n",
       "8        0.000000  0.000332  0.013246  0.000000  0.006027   0.001915   \n",
       "9        0.000000  0.000267  0.002727  0.000000  0.001203   0.199339   \n",
       "\n",
       "         lb-spinal    spinal  \\\n",
       "cluster                        \n",
       "1         0.154515  0.195853   \n",
       "2         0.000000  0.002235   \n",
       "3         0.181362  0.195096   \n",
       "4         0.000000  0.000704   \n",
       "5         0.000000  0.004621   \n",
       "6         0.000000  0.000497   \n",
       "7         0.000000  0.000000   \n",
       "8         0.000000  0.000000   \n",
       "9         0.175488  0.187231   \n",
       "\n",
       "                                           character_label  \n",
       "cluster                                                     \n",
       "1        INVS:9:101950362:101950362_9:105825009:1058250...  \n",
       "2        6:50519104:50519104_ENSG00000214604:6:55918529...  \n",
       "3        6:50534952:50534952_RAB23:6:57180847:57180847_...  \n",
       "4        ICK:6:53035144:53035144_LOC727842:6:58602457:5...  \n",
       "5        6:50568504:50568504_6:99144022:99144022_ROS1:6...  \n",
       "6        ENSG00000221175:6:48902663:48902663_6:50179508...  \n",
       "7        AHCYL1:1:110359957:110359957_6:97912586:979125...  \n",
       "8        TFAP2B:6:50909762:50909762_ENPP3:6:132075413:1...  \n",
       "9        6:49276467:49276467_DEFB112:6:50173697:5017369...  \n",
       "\n",
       "[9 rows x 26 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def add_char_label(row): \n",
    "    return \"_\".join(list(raw_table[raw_table['cluster']==row.name]['character_label']))\n",
    "\n",
    "ctable_cutoff['character_label'] = ctable_cutoff.apply(lambda row: add_char_label(row), axis=1)\n",
    "ctable_cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "rows = [\"5 #anatomical sites\\n5 #samples\\n9 #mutations\\n#sample_index\\tsample_label\\tanatomical_site_index\\tanatomical_site_label\\tcharacter_index\\tcharacter_label\\tf_lb\\tf_ub\\tref\\tvar\\n\",]\n",
    "def print_char(row, sam):\n",
    "    return \"\\t\".join(map(str,[i, sam, i, sam, row.name-1, str(row['character_label']), max(row['lb-'+sam] * 2, 0), min(1, 2 * row['ub-'+sam]), int(row['ref-'+sam]), int(row['var-'+sam])]))+\"\\n\"\n",
    "\n",
    "for i, sam in enumerate(cols):\n",
    "    rows += list(ctable_cutoff.apply(print_char, args=[sam], axis=1))\n",
    "\n",
    "with open(os.path.join(HOADLEY_DATA_DIR,\"A1/A1_\"+str(confidence)+\".tsv\"), 'w') as f:\n",
    "    for line in rows:\n",
    "        f.write(line)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
