{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8e982e15",
   "metadata": {},
   "source": [
    "### Load paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9472b7a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyreadr\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "\n",
    "REPO_DIR = os.path.join(os.getcwd(), \"../../\")\n",
    "os.chdir(REPO_DIR)\n",
    "\n",
    "TRACERX_DATA_DIR = os.path.join(REPO_DIR, \"src/data/tracerx_nsclc/\")\n",
    "OUTPUT_DIR = os.path.join(REPO_DIR, \"src/data/tracerx_nsclc/patient_data\")\n",
    "CONIPHER_DIR = os.path.join(REPO_DIR, \"src/data/tracerx_nsclc/conipher_inputs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f81725b",
   "metadata": {},
   "source": [
    "### Load sample overview information (from TRACERx publicly available data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "be0117c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sampleType\n",
      "primary       476\n",
      "metastasis    218\n",
      "Name: count, dtype: int64\n",
      "sampleTypeDetail\n",
      "primary            476\n",
      "LN                 148\n",
      "metachronousMet     67\n",
      "synchronousMet       3\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumour_id</th>\n",
       "      <th>region</th>\n",
       "      <th>sampleType</th>\n",
       "      <th>sampleTypeDetail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010_SU_T1.R1</td>\n",
       "      <td>primary</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010_SU_T1.R2</td>\n",
       "      <td>primary</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010_SU_FLN1</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>LN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010_BR_LN1</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>metachronousMet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>CRUK0010_BR_LN2</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>metachronousMet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872_SU_T1.R1</td>\n",
       "      <td>primary</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872_SU_T1.R2</td>\n",
       "      <td>primary</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872_SU_T1.R3</td>\n",
       "      <td>primary</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872_SU_T1.R4</td>\n",
       "      <td>primary</td>\n",
       "      <td>primary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693</th>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CRUK0872_SU_FLN1</td>\n",
       "      <td>metastasis</td>\n",
       "      <td>LN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>694 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    patient_id tumour_id             region  sampleType sampleTypeDetail\n",
       "0     CRUK0010  CRUK0010  CRUK0010_SU_T1.R1     primary          primary\n",
       "1     CRUK0010  CRUK0010  CRUK0010_SU_T1.R2     primary          primary\n",
       "2     CRUK0010  CRUK0010   CRUK0010_SU_FLN1  metastasis               LN\n",
       "3     CRUK0010  CRUK0010    CRUK0010_BR_LN1  metastasis  metachronousMet\n",
       "4     CRUK0010  CRUK0010    CRUK0010_BR_LN2  metastasis  metachronousMet\n",
       "..         ...       ...                ...         ...              ...\n",
       "689   CRUK0872  CRUK0872  CRUK0872_SU_T1.R1     primary          primary\n",
       "690   CRUK0872  CRUK0872  CRUK0872_SU_T1.R2     primary          primary\n",
       "691   CRUK0872  CRUK0872  CRUK0872_SU_T1.R3     primary          primary\n",
       "692   CRUK0872  CRUK0872  CRUK0872_SU_T1.R4     primary          primary\n",
       "693   CRUK0872  CRUK0872   CRUK0872_SU_FLN1  metastasis               LN\n",
       "\n",
       "[694 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info_df= pd.read_csv(os.path.join(TRACERX_DATA_DIR,\"sample_overview_original.txt\"), sep=\"\\t\")\n",
    "print(sample_info_df['sampleType'].value_counts())\n",
    "print(sample_info_df['sampleTypeDetail'].value_counts())\n",
    "sample_info_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0352cbd5",
   "metadata": {},
   "source": [
    "### Load purity and ploidy information (from TRACERx publicly available data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9a183098",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Sample</th>\n",
       "      <th>final.purity</th>\n",
       "      <th>final.ploidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CRUK0090</td>\n",
       "      <td>CRUK0090_SU_T1.R2</td>\n",
       "      <td>0.54</td>\n",
       "      <td>3.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRUK0090</td>\n",
       "      <td>CRUK0090_BP_T1.R1</td>\n",
       "      <td>0.74</td>\n",
       "      <td>3.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRUK0090</td>\n",
       "      <td>CRUK0090_BR_T1.R1</td>\n",
       "      <td>0.20</td>\n",
       "      <td>3.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRUK0090</td>\n",
       "      <td>CRUK0090_SU_T1.R1</td>\n",
       "      <td>0.58</td>\n",
       "      <td>3.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRUK0090</td>\n",
       "      <td>CRUK0090_BP_T2.R1</td>\n",
       "      <td>0.23</td>\n",
       "      <td>3.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>CRUK0810</td>\n",
       "      <td>CRUK0810_SU_T1.R4</td>\n",
       "      <td>0.49</td>\n",
       "      <td>3.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>CRUK0810</td>\n",
       "      <td>CRUK0810_SU_T1.R1</td>\n",
       "      <td>0.25</td>\n",
       "      <td>3.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>691</th>\n",
       "      <td>CRUK0810</td>\n",
       "      <td>CRUK0810_SU_T1.R8</td>\n",
       "      <td>0.20</td>\n",
       "      <td>3.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692</th>\n",
       "      <td>CRUK0810</td>\n",
       "      <td>CRUK0810_SU_T1.R6</td>\n",
       "      <td>0.39</td>\n",
       "      <td>3.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693</th>\n",
       "      <td>CRUK0810</td>\n",
       "      <td>CRUK0810_SU_T1.R5</td>\n",
       "      <td>0.51</td>\n",
       "      <td>3.75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>694 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Patient             Sample  final.purity  final.ploidy\n",
       "0    CRUK0090  CRUK0090_SU_T1.R2          0.54          3.15\n",
       "1    CRUK0090  CRUK0090_BP_T1.R1          0.74          3.25\n",
       "2    CRUK0090  CRUK0090_BR_T1.R1          0.20          3.30\n",
       "3    CRUK0090  CRUK0090_SU_T1.R1          0.58          3.20\n",
       "4    CRUK0090  CRUK0090_BP_T2.R1          0.23          3.40\n",
       "..        ...                ...           ...           ...\n",
       "689  CRUK0810  CRUK0810_SU_T1.R4          0.49          3.85\n",
       "690  CRUK0810  CRUK0810_SU_T1.R1          0.25          3.70\n",
       "691  CRUK0810  CRUK0810_SU_T1.R8          0.20          3.95\n",
       "692  CRUK0810  CRUK0810_SU_T1.R6          0.39          3.65\n",
       "693  CRUK0810  CRUK0810_SU_T1.R5          0.51          3.75\n",
       "\n",
       "[694 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "purity_ploidy_df = pd.read_csv(os.path.join(TRACERX_DATA_DIR, \"20220808_purityandploidy.txt\"), sep=\"\\t\")\n",
    "purity_ploidy_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65d35c5a",
   "metadata": {},
   "source": [
    "### Load all patient mutation data (from TRACERx publicly available data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "62a1b970",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient data columns:\n",
      "  Index(['mutation_id', 'patient_id', 'tumour_id', 'chr', 'start', 'stop', 'ref',\n",
      "       'var', 'Hugo_Symbol', 'func', 'exonic.func', 'NucleotideChange',\n",
      "       'AAChange', 'GL_VAF', 'GL_nAlt', 'GL_depth', 'ITHState', 'RegionSum',\n",
      "       'Is.present', 'PyCloneClonal_SC', 'PyCloneCluster_SC',\n",
      "       'cleanCluster_SC', 'PhyloCCF_SC', 'combTiming_SC', 'MutCPN_SC',\n",
      "       'MajorCPN_SC', 'MinorCPN_SC', 'DriverMut', 'treeClones',\n",
      "       'seedingClones', 'clonalClones', 'sharedClones', 'primaryClones',\n",
      "       'metClones'],\n",
      "      dtype='object')\n",
      "0              BR_LN1:40/168;BR_LN2:64/276;BR_LN3:47/235;SU_FLN1:96/518;SU_T1.R1:236/317;SU_T1.R2:80/227\n",
      "1               BR_LN1:14/106;BR_LN2:18/155;BR_LN3:18/138;SU_FLN1:12/102;SU_T1.R1:94/153;SU_T1.R2:27/175\n",
      "2              BR_LN1:44/191;BR_LN2:55/233;BR_LN3:75/260;SU_FLN1:86/288;SU_T1.R1:148/245;SU_T1.R2:92/270\n",
      "3             BR_LN1:71/525;BR_LN2:61/694;BR_LN3:60/675;SU_FLN1:57/485;SU_T1.R1:385/647;SU_T1.R2:103/654\n",
      "4               BR_LN1:16/116;BR_LN2:21/141;BR_LN3:27/180;SU_FLN1:49/281;SU_T1.R1:76/152;SU_T1.R2:31/139\n",
      "5            BR_LN1:55/456;BR_LN2:54/702;BR_LN3:84/653;SU_FLN1:131/863;SU_T1.R1:278/559;SU_T1.R2:152/626\n",
      "6    BR_LN1:136/815;BR_LN2:190/1083;BR_LN3:222/1104;SU_FLN1:161/1095;SU_T1.R1:669/1375;SU_T1.R2:291/1096\n",
      "7           BR_LN1:63/368;BR_LN2:86/542;BR_LN3:103/571;SU_FLN1:112/585;SU_T1.R1:312/677;SU_T1.R2:127/528\n",
      "8              BR_LN1:39/333;BR_LN2:43/454;BR_LN3:65/486;SU_FLN1:41/435;SU_T1.R1:270/604;SU_T1.R2:66/431\n",
      "9              BR_LN1:39/291;BR_LN2:45/420;BR_LN3:56/445;SU_FLN1:76/570;SU_T1.R1:225/534;SU_T1.R2:99/437\n",
      "Name: RegionSum, dtype: object\n",
      "Number of unique patients: 126\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>exonic.func</th>\n",
       "      <th>NucleotideChange</th>\n",
       "      <th>RegionSum</th>\n",
       "      <th>MajorCPN_SC</th>\n",
       "      <th>MinorCPN_SC</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>ref</th>\n",
       "      <th>var</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CRUK0010:3:47103798:G</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>SETD2</td>\n",
       "      <td>stopgain SNV</td>\n",
       "      <td>c.C6148T</td>\n",
       "      <td>BR_LN1:40/168;BR_LN2:64/276;BR_LN3:47/235;SU_FLN1:96/518;SU_T1.R1:236/317;SU_T1.R2:80/227</td>\n",
       "      <td>SU_T1.R1:3;SU_T1.R2:2;SU_FLN1:2;BR_LN1:2;BR_LN2:2;BR_LN3:2</td>\n",
       "      <td>SU_T1.R1:0;SU_T1.R2:0;SU_FLN1:0;BR_LN1:0;BR_LN2:0;BR_LN3:0</td>\n",
       "      <td>chr3</td>\n",
       "      <td>47103798</td>\n",
       "      <td>47103798</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRUK0010:6:45459895:C</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>RUNX2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BR_LN1:14/106;BR_LN2:18/155;BR_LN3:18/138;SU_FLN1:12/102;SU_T1.R1:94/153;SU_T1.R2:27/175</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1</td>\n",
       "      <td>SU_T1.R1:0;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>45459895</td>\n",
       "      <td>45459895</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRUK0010:1:156105064:C</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>LMNA</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>c.C561T</td>\n",
       "      <td>BR_LN1:44/191;BR_LN2:55/233;BR_LN3:75/260;SU_FLN1:86/288;SU_T1.R1:148/245;SU_T1.R2:92/270</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_FLN1:3;BR_LN1:2;BR_LN2:2;BR_LN3:3</td>\n",
       "      <td>SU_T1.R1:0;SU_T1.R2:0;SU_FLN1:0;BR_LN1:0;BR_LN2:0;BR_LN3:0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>156105064</td>\n",
       "      <td>156105064</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRUK0010:6:25850091:C</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>SLC17A3</td>\n",
       "      <td>stopgain SNV</td>\n",
       "      <td>c.G979T</td>\n",
       "      <td>BR_LN1:71/525;BR_LN2:61/694;BR_LN3:60/675;SU_FLN1:57/485;SU_T1.R1:385/647;SU_T1.R2:103/654</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:1;SU_FLN1:2;BR_LN1:1;BR_LN2:1;BR_LN3:1</td>\n",
       "      <td>SU_T1.R1:0;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1</td>\n",
       "      <td>chr6</td>\n",
       "      <td>25850091</td>\n",
       "      <td>25850091</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRUK0010:8:16001109:G</td>\n",
       "      <td>CRUK0010</td>\n",
       "      <td>MSR1</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>c.C991G</td>\n",
       "      <td>BR_LN1:16/116;BR_LN2:21/141;BR_LN3:27/180;SU_FLN1:49/281;SU_T1.R1:76/152;SU_T1.R2:31/139</td>\n",
       "      <td>SU_T1.R1:1;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1</td>\n",
       "      <td>SU_T1.R1:0;SU_T1.R2:0;SU_FLN1:0;BR_LN1:0;BR_LN2:0;BR_LN3:0</td>\n",
       "      <td>chr8</td>\n",
       "      <td>16001109</td>\n",
       "      <td>16001109</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136310</th>\n",
       "      <td>CRUK0872:22:40080280:C</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>CACNA1I</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SU_FLN1:18/668;SU_LN1:1/446;SU_T1.R1:0/602;SU_T1.R2:0/501;SU_T1.R3:0/443;SU_T1.R4:0/463</td>\n",
       "      <td>SU_T1.R1:3;SU_T1.R2:3;SU_T1.R3:2;SU_T1.R4:3;SU_FLN1:3</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:3;SU_FLN1:2</td>\n",
       "      <td>chr22</td>\n",
       "      <td>40080280</td>\n",
       "      <td>40080280</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136311</th>\n",
       "      <td>CRUK0872:14:99872776:T</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>SETD3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SU_FLN1:0/573;SU_LN1:0/518;SU_T1.R1:0/609;SU_T1.R2:0/625;SU_T1.R3:0/581;SU_T1.R4:13/510</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>99872776</td>\n",
       "      <td>99872776</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136312</th>\n",
       "      <td>CRUK0872:8:110131566:C</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>TRHR</td>\n",
       "      <td>stopgain SNV</td>\n",
       "      <td>c.C1079A</td>\n",
       "      <td>SU_FLN1:0/562;SU_LN1:3/616;SU_T1.R1:2/716;SU_T1.R2:0/763;SU_T1.R3:0/722;SU_T1.R4:16/650</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2</td>\n",
       "      <td>chr8</td>\n",
       "      <td>110131566</td>\n",
       "      <td>110131566</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136313</th>\n",
       "      <td>CRUK0872:1:51032853:G</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>FAF1</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>c.C1164G</td>\n",
       "      <td>SU_FLN1:0/389;SU_LN1:0/394;SU_T1.R1:0/464;SU_T1.R2:0/484;SU_T1.R3:11/449;SU_T1.R4:0/428</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2</td>\n",
       "      <td>SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>51032853</td>\n",
       "      <td>51032853</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136314</th>\n",
       "      <td>CRUK0872:20:24524120:G</td>\n",
       "      <td>CRUK0872</td>\n",
       "      <td>SYNDIG1</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>c.G387A</td>\n",
       "      <td>SU_FLN1:10/1295;SU_LN1:15/612;SU_T1.R1:0/745;SU_T1.R2:1/716;SU_T1.R3:0/740;SU_T1.R4:0/621</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chr20</td>\n",
       "      <td>24524120</td>\n",
       "      <td>24524120</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136315 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   mutation_id patient_id Hugo_Symbol        exonic.func  \\\n",
       "0        CRUK0010:3:47103798:G   CRUK0010       SETD2       stopgain SNV   \n",
       "1        CRUK0010:6:45459895:C   CRUK0010       RUNX2                NaN   \n",
       "2       CRUK0010:1:156105064:C   CRUK0010        LMNA     synonymous SNV   \n",
       "3        CRUK0010:6:25850091:C   CRUK0010     SLC17A3       stopgain SNV   \n",
       "4        CRUK0010:8:16001109:G   CRUK0010        MSR1  nonsynonymous SNV   \n",
       "...                        ...        ...         ...                ...   \n",
       "136310  CRUK0872:22:40080280:C   CRUK0872     CACNA1I                NaN   \n",
       "136311  CRUK0872:14:99872776:T   CRUK0872       SETD3                NaN   \n",
       "136312  CRUK0872:8:110131566:C   CRUK0872        TRHR       stopgain SNV   \n",
       "136313   CRUK0872:1:51032853:G   CRUK0872        FAF1  nonsynonymous SNV   \n",
       "136314  CRUK0872:20:24524120:G   CRUK0872     SYNDIG1     synonymous SNV   \n",
       "\n",
       "       NucleotideChange  \\\n",
       "0              c.C6148T   \n",
       "1                   NaN   \n",
       "2               c.C561T   \n",
       "3               c.G979T   \n",
       "4               c.C991G   \n",
       "...                 ...   \n",
       "136310              NaN   \n",
       "136311              NaN   \n",
       "136312         c.C1079A   \n",
       "136313         c.C1164G   \n",
       "136314          c.G387A   \n",
       "\n",
       "                                                                                         RegionSum  \\\n",
       "0        BR_LN1:40/168;BR_LN2:64/276;BR_LN3:47/235;SU_FLN1:96/518;SU_T1.R1:236/317;SU_T1.R2:80/227   \n",
       "1         BR_LN1:14/106;BR_LN2:18/155;BR_LN3:18/138;SU_FLN1:12/102;SU_T1.R1:94/153;SU_T1.R2:27/175   \n",
       "2        BR_LN1:44/191;BR_LN2:55/233;BR_LN3:75/260;SU_FLN1:86/288;SU_T1.R1:148/245;SU_T1.R2:92/270   \n",
       "3       BR_LN1:71/525;BR_LN2:61/694;BR_LN3:60/675;SU_FLN1:57/485;SU_T1.R1:385/647;SU_T1.R2:103/654   \n",
       "4         BR_LN1:16/116;BR_LN2:21/141;BR_LN3:27/180;SU_FLN1:49/281;SU_T1.R1:76/152;SU_T1.R2:31/139   \n",
       "...                                                                                            ...   \n",
       "136310     SU_FLN1:18/668;SU_LN1:1/446;SU_T1.R1:0/602;SU_T1.R2:0/501;SU_T1.R3:0/443;SU_T1.R4:0/463   \n",
       "136311     SU_FLN1:0/573;SU_LN1:0/518;SU_T1.R1:0/609;SU_T1.R2:0/625;SU_T1.R3:0/581;SU_T1.R4:13/510   \n",
       "136312     SU_FLN1:0/562;SU_LN1:3/616;SU_T1.R1:2/716;SU_T1.R2:0/763;SU_T1.R3:0/722;SU_T1.R4:16/650   \n",
       "136313     SU_FLN1:0/389;SU_LN1:0/394;SU_T1.R1:0/464;SU_T1.R2:0/484;SU_T1.R3:11/449;SU_T1.R4:0/428   \n",
       "136314   SU_FLN1:10/1295;SU_LN1:15/612;SU_T1.R1:0/745;SU_T1.R2:1/716;SU_T1.R3:0/740;SU_T1.R4:0/621   \n",
       "\n",
       "                                                       MajorCPN_SC  \\\n",
       "0       SU_T1.R1:3;SU_T1.R2:2;SU_FLN1:2;BR_LN1:2;BR_LN2:2;BR_LN3:2   \n",
       "1       SU_T1.R1:2;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1   \n",
       "2       SU_T1.R1:2;SU_T1.R2:2;SU_FLN1:3;BR_LN1:2;BR_LN2:2;BR_LN3:3   \n",
       "3       SU_T1.R1:2;SU_T1.R2:1;SU_FLN1:2;BR_LN1:1;BR_LN2:1;BR_LN3:1   \n",
       "4       SU_T1.R1:1;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1   \n",
       "...                                                            ...   \n",
       "136310       SU_T1.R1:3;SU_T1.R2:3;SU_T1.R3:2;SU_T1.R4:3;SU_FLN1:3   \n",
       "136311       SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2   \n",
       "136312       SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2   \n",
       "136313       SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2   \n",
       "136314                                                         NaN   \n",
       "\n",
       "                                                       MinorCPN_SC    chr  \\\n",
       "0       SU_T1.R1:0;SU_T1.R2:0;SU_FLN1:0;BR_LN1:0;BR_LN2:0;BR_LN3:0   chr3   \n",
       "1       SU_T1.R1:0;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1   chr6   \n",
       "2       SU_T1.R1:0;SU_T1.R2:0;SU_FLN1:0;BR_LN1:0;BR_LN2:0;BR_LN3:0   chr1   \n",
       "3       SU_T1.R1:0;SU_T1.R2:1;SU_FLN1:1;BR_LN1:1;BR_LN2:1;BR_LN3:1   chr6   \n",
       "4       SU_T1.R1:0;SU_T1.R2:0;SU_FLN1:0;BR_LN1:0;BR_LN2:0;BR_LN3:0   chr8   \n",
       "...                                                            ...    ...   \n",
       "136310       SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:3;SU_FLN1:2  chr22   \n",
       "136311       SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2  chr14   \n",
       "136312       SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:2   chr8   \n",
       "136313       SU_T1.R1:2;SU_T1.R2:2;SU_T1.R3:2;SU_T1.R4:2;SU_FLN1:1   chr1   \n",
       "136314                                                         NaN  chr20   \n",
       "\n",
       "            start       stop ref var  \n",
       "0        47103798   47103798   G   A  \n",
       "1        45459895   45459895   C   G  \n",
       "2       156105064  156105064   C   T  \n",
       "3        25850091   25850091   C   A  \n",
       "4        16001109   16001109   G   C  \n",
       "...           ...        ...  ..  ..  \n",
       "136310   40080280   40080280   C   T  \n",
       "136311   99872776   99872776   T   C  \n",
       "136312  110131566  110131566   C   A  \n",
       "136313   51032853   51032853   G   C  \n",
       "136314   24524120   24524120   G   A  \n",
       "\n",
       "[136315 rows x 13 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_data_all = pyreadr.read_r(os.path.join(TRACERX_DATA_DIR, 'mutTableAll.cloneInfo.20220726.rda'))\n",
    "patient_data_all_df = patient_data_all['mutTableAll']\n",
    "print(\"Patient data columns:\\n \", patient_data_all_df.columns)\n",
    "print(patient_data_all_df[:10]['RegionSum'])\n",
    "\n",
    "patient_data_all_df = patient_data_all_df[['mutation_id', 'patient_id', 'Hugo_Symbol','exonic.func','NucleotideChange', 'RegionSum' , 'MajorCPN_SC', 'MinorCPN_SC', 'chr', 'start', 'stop', 'ref', 'var']]\n",
    "unique_patients = patient_data_all_df['patient_id'].unique()\n",
    "print(\"Number of unique patients:\", len(unique_patients))\n",
    "patient_data_all_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "086e3e99",
   "metadata": {},
   "source": [
    "### Prep data for input into tree building methods (CONIPHER, PairTree, Orchard) and migration history methods (Metient and MACHINA)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a71a9666-ed32-4309-aba8-51e340c71702",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NOTE: Redirects are currently not supported in Windows or MacOs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CUDA GPU: False\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "For Matient/MACHINA, need a tsv with minimum columns: ['#sample_index', 'sample_label', 'anatomical_site_index',\n",
    "'anatomical_site_label', 'character_index', 'character_label','ref', 'var']\n",
    "\n",
    "For CONIPHER, need a tsv with minimum columns: ['CASE_ID', 'SAMPLE', 'CHR', 'POS', 'REF', 'ALT', \n",
    "'REF_COUNT', 'VAR_COUNT', 'DEPTH', 'COPY_NUMBER_A', 'COPY_NUMBER_B', 'ACF', 'PLOIDY']\n",
    "'''\n",
    "\n",
    "import re\n",
    "from src.util import pairtree_data_extraction_util as pt_util\n",
    "import numpy as np\n",
    "\n",
    "def remove_region_suffix(input_string):\n",
    "    pattern = r'\\.R\\d+$'\n",
    "    modified_string = re.sub(pattern, '', input_string)\n",
    "    return modified_string\n",
    "\n",
    "def generate_sample_and_mut_info(row):\n",
    "    '''\n",
    "    Given a row in the patient_data_all_df dataframe, yields two lists:\n",
    "        (1) the first for MACHINA/Matient input, \n",
    "        with columns: [sample_name, mutation_name, sample_type_detailed, ref_count,var_count, sample_type, purity, ploidy]\n",
    "        e.g. [\"SETD3:2:14:99872776:T\", \"BR_LN1\", \"LN_SU_FLN1\", 43,270, \"metastasis\", 0.54, 3.19]\n",
    "        \n",
    "        (2) the second for CONIPHER input,\n",
    "        with columns: [patient_id, sample_name, chr, pos, ref, alt, ref_count, var_count, depth,\n",
    "        copy_number_a, copy_number_b, purity, ploidy]\n",
    "        \n",
    "    '''\n",
    "    mut_id = \":\".join(row['mutation_id'].split(\":\")[1:])\n",
    "    mut_name = f\"{row['Hugo_Symbol']}:{mut_id}\"\n",
    "        \n",
    "    # RegionSum  column values need to be split by sample \n",
    "    # (e.g. \"SU_FLN1:0/562;SU_LN1:3/616\")    \n",
    "    sample_name_to_mut_counts = {x.split(\":\")[0]:x.split(\":\")[1] for x in row['RegionSum'].split(\";\")}\n",
    "    # Don't include mutations without copy number information\n",
    "    if isinstance(row['MajorCPN_SC'], float) or isinstance(row['MinorCPN_SC'], float):\n",
    "        yield None\n",
    "        return\n",
    "    sample_name_to_major_cn = {x.split(\":\")[0]:x.split(\":\")[1] for x in row['MajorCPN_SC'].split(\";\")}\n",
    "    sample_name_to_minor_cn = {x.split(\":\")[0]:x.split(\":\")[1] for x in row['MinorCPN_SC'].split(\";\")}\n",
    "\n",
    "\n",
    "    for sample_name in sample_name_to_mut_counts.keys():\n",
    "        mut_count = sample_name_to_mut_counts[sample_name]\n",
    "        # Lookup sample info in sample_overview dataframe\n",
    "        patient_sample_name = row['patient_id']+\"_\"+sample_name\n",
    "        sample_info = sample_info_df[sample_info_df['region']==patient_sample_name]\n",
    "        # Low purity samples\n",
    "        if len(sample_info) == 0:\n",
    "            yield None\n",
    "            continue\n",
    "        \n",
    "        sample_type = sample_info['sampleType'].item()\n",
    "        sample_type_detailed = sample_info['sampleTypeDetail'].item()+\"_\"+sample_name\n",
    "        \n",
    "        # Lookup purity and ploidy info\n",
    "        purity_ploidy_info = purity_ploidy_df[purity_ploidy_df[\"Sample\"]==patient_sample_name]\n",
    "        if len(purity_ploidy_info) == 0:\n",
    "            yield None\n",
    "            continue\n",
    "        purity = purity_ploidy_info[\"final.purity\"].item()\n",
    "        ploidy =  purity_ploidy_info[\"final.ploidy\"].item()\n",
    "        \n",
    "        # Get ref and var counts\n",
    "        var = int(mut_count.split(\"/\")[0])\n",
    "        total = int(mut_count.split(\"/\")[1])\n",
    "        ref = total - var\n",
    "        \n",
    "        # According to TRACERx: muts w/o copy number calls in this dataset are muts that either do not overlap with \n",
    "        # copy number  values – (e.g. on sex chromosomes or mutations from regions with low purity), and therefore \n",
    "        # not included in clustering and phylogenetic tree building.  \n",
    "        if (sample_name not in sample_name_to_major_cn) or (sample_name not in sample_name_to_minor_cn):\n",
    "            yield None\n",
    "            continue\n",
    "        cn_a_count = sample_name_to_major_cn[sample_name]\n",
    "        cn_b_count = sample_name_to_minor_cn[sample_name]\n",
    "\n",
    "        # remove_region_suffix(sample_type_detailed) means samples from multiple regions\n",
    "        # of the same primary get mapped to the same anatomical site label, e.g. primary.T1.R1 \n",
    "        # and primary.T1.R2 get mapped to primary.T1. This means they get pooled together below as well.\n",
    "        mig_hist_input = [mut_name, patient_sample_name, remove_region_suffix(sample_type_detailed), \n",
    "                          ref, var,  sample_type_detailed, sample_type, purity, ploidy]\n",
    "            \n",
    "        conipher_input = [ row['patient_id'], patient_sample_name, row['chr'].replace('chr', ''), row['start'], \n",
    "                          row['ref'], row['var'], ref, var, total,cn_a_count, cn_b_count, purity, ploidy ]\n",
    "        \n",
    "        yield mig_hist_input, conipher_input\n",
    "\n",
    "        \n",
    "def format_mig_hist_df(data):\n",
    "    sample_mut_df = pd.DataFrame(data)\n",
    "    sample_mut_df.columns = [\"character_label\", \"sample_label\", \"anatomical_site_label\", \"ref\",\n",
    "                            \"var\", \"original_anatomical_site_label\", \"sample_type\", \"purity\", \"ploidy\"]\n",
    "    \n",
    "    # Add indices for mutations, samples and anatomical sites as needed for input format\n",
    "    sample_mut_df['#sample_index'] = sample_mut_df.apply(lambda row: list(sample_mut_df['sample_label'].unique()).index(row[\"sample_label\"]), axis=1)    \n",
    "    sample_mut_df['character_index'] = sample_mut_df.apply(lambda row: list(sample_mut_df['character_label'].unique()).index(row[\"character_label\"]), axis=1)\n",
    "    sample_mut_df['anatomical_site_index'] = sample_mut_df.apply(lambda row: list(sample_mut_df['anatomical_site_label'].unique()).index(row[\"anatomical_site_label\"]), axis=1)\n",
    "    \n",
    "    sample_mut_df = sample_mut_df[['#sample_index', \"sample_label\", \"anatomical_site_index\", \"anatomical_site_label\", \n",
    "                                 'character_index', \"character_label\", \"ref\",\"var\", \"original_anatomical_site_label\", \"sample_type\",\"purity\", \"ploidy\"]]\n",
    "\n",
    "    return sample_mut_df\n",
    "\n",
    "def format_conipher_df(data):\n",
    "    sample_mut_df = pd.DataFrame(data)\n",
    "    sample_mut_df.columns = ['CASE_ID', 'SAMPLE', 'CHR', 'POS', 'REF', 'ALT', 'REF_COUNT', 'VAR_COUNT', \n",
    "                             'DEPTH', 'COPY_NUMBER_A', 'COPY_NUMBER_B', 'ACF', 'PLOIDY']\n",
    "    sample_mut_df = sample_mut_df.dropna()\n",
    "    return sample_mut_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "138d342a-22bc-4db7-8156-474a08d35513",
   "metadata": {},
   "source": [
    "### Write all tsvs/json/ssm files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "89f729c1-6c79-4913-a929-8d31261ec4fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CRUK0010\n",
      "242\n",
      "CRUK0013\n",
      "246\n",
      "CRUK0284\n",
      "1331\n",
      "CRUK0361\n",
      "1484\n",
      "CRUK0497\n",
      "1654\n",
      "CRUK0452\n",
      "525\n",
      "CRUK0472\n",
      "988\n",
      "CRUK0484\n",
      "1333\n",
      "CRUK0418\n",
      "11308\n",
      "CRUK0721\n",
      "2856\n",
      "CRUK0702\n",
      "646\n",
      "CRUK0794\n",
      "366\n",
      "CRUK0004\n",
      "550\n",
      "CRUK0027\n",
      "1157\n",
      "CRUK0044\n",
      "577\n",
      "CRUK0052\n",
      "1541\n",
      "CRUK0100\n",
      "986\n",
      "CRUK0036\n",
      "610\n",
      "CRUK0083\n",
      "1125\n",
      "CRUK0178\n",
      "944\n",
      "CRUK0286\n",
      "875\n",
      "CRUK0256\n",
      "1249\n",
      "CRUK0227\n",
      "785\n",
      "CRUK0236\n",
      "884\n",
      "CRUK0352\n",
      "2471\n",
      "CRUK0388\n",
      "199\n",
      "CRUK0343\n",
      "259\n",
      "CRUK0422\n",
      "1075\n",
      "CRUK0485\n",
      "1376\n",
      "CRUK0478\n",
      "1241\n",
      "CRUK0598\n",
      "3231\n",
      "CRUK0590\n",
      "889\n",
      "CRUK0543\n",
      "1273\n",
      "CRUK0524\n",
      "224\n",
      "CRUK0567\n",
      "1408\n",
      "CRUK0620\n",
      "1679\n",
      "CRUK0625\n",
      "606\n",
      "CRUK0691\n",
      "1517\n",
      "CRUK0693\n",
      "3106\n",
      "CRUK0609\n",
      "623\n",
      "CRUK0617\n",
      "766\n",
      "CRUK0698\n",
      "859\n",
      "CRUK0799\n",
      "2330\n",
      "CRUK0766\n",
      "1943\n",
      "CRUK0769\n",
      "1111\n",
      "CRUK0714\n",
      "872\n",
      "CRUK0722\n",
      "1103\n",
      "CRUK0719\n",
      "1046\n",
      "CRUK0003\n",
      "491\n",
      "CRUK0089\n",
      "1102\n",
      "CRUK0290\n",
      "1489\n",
      "CRUK0468\n",
      "1033\n",
      "CRUK0410\n",
      "522\n",
      "CRUK0412\n",
      "372\n",
      "CRUK0415\n",
      "537\n",
      "CRUK0745\n",
      "713\n",
      "CRUK0084\n",
      "1134\n",
      "CRUK0009\n",
      "602\n",
      "CRUK0296\n",
      "274\n",
      "CRUK0097\n",
      "604\n",
      "CRUK0087\n",
      "1122\n",
      "CRUK0245\n",
      "606\n",
      "CRUK0074\n",
      "1330\n",
      "CRUK0022\n",
      "367\n",
      "CRUK0299\n",
      "894\n",
      "CRUK0369\n",
      "2370\n",
      "CRUK0301\n",
      "1086\n",
      "CRUK0337\n",
      "2584\n",
      "CRUK0311\n",
      "1136\n",
      "CRUK0451\n",
      "923\n",
      "CRUK0453\n",
      "1951\n",
      "CRUK0537\n",
      "458\n",
      "CRUK0530\n",
      "902\n",
      "CRUK0730\n",
      "349\n",
      "CRUK0707\n",
      "1187\n",
      "CRUK0325\n",
      "597\n",
      "CRUK0099\n",
      "671\n",
      "CRUK0368\n",
      "536\n",
      "CRUK0465\n",
      "433\n",
      "CRUK0742\n",
      "146\n",
      "CRUK0444\n",
      "562\n",
      "CRUK0510\n",
      "2135\n",
      "CRUK0572\n",
      "348\n",
      "CRUK0090\n",
      "1948\n",
      "CRUK0063\n",
      "657\n",
      "CRUK0092\n",
      "343\n",
      "CRUK0029\n",
      "447\n",
      "CRUK0035\n",
      "501\n",
      "CRUK0183\n",
      "961\n",
      "CRUK0041\n",
      "261\n",
      "CRUK0048\n",
      "582\n",
      "CRUK0242\n",
      "696\n",
      "CRUK0372\n",
      "604\n",
      "CRUK0344\n",
      "610\n",
      "CRUK0487\n",
      "621\n",
      "CRUK0496\n",
      "837\n",
      "CRUK0407\n",
      "312\n",
      "CRUK0467\n",
      "1044\n",
      "CRUK0514\n",
      "4720\n",
      "CRUK0584\n",
      "565\n",
      "CRUK0519\n",
      "2506\n",
      "CRUK0557\n",
      "920\n",
      "CRUK0559\n",
      "923\n",
      "CRUK0516\n",
      "1939\n",
      "CRUK0596\n",
      "640\n",
      "CRUK0587\n",
      "1212\n",
      "CRUK0589\n",
      "1005\n",
      "CRUK0667\n",
      "877\n",
      "CRUK0640\n",
      "1235\n",
      "CRUK0636\n",
      "780\n",
      "CRUK0748\n",
      "925\n",
      "CRUK0762\n",
      "678\n",
      "CRUK0736\n",
      "318\n",
      "CRUK0737\n",
      "554\n",
      "CRUK0733\n",
      "251\n",
      "CRUK0718\n",
      "537\n",
      "CRUK0817\n",
      "612\n",
      "CRUK0810\n",
      "642\n",
      "CRUK0552\n",
      "1184\n",
      "CRUK0206\n",
      "915\n",
      "CRUK0250\n",
      "918\n",
      "CRUK0495\n",
      "1236\n",
      "CRUK0476\n",
      "612\n",
      "CRUK0528\n",
      "1075\n",
      "CRUK0666\n",
      "603\n",
      "CRUK0872\n",
      "474\n"
     ]
    }
   ],
   "source": [
    "for patient_id in unique_patients:\n",
    "    print(patient_id)\n",
    "    output_rows = []\n",
    "    subset = patient_data_all_df[patient_data_all_df['patient_id'] == patient_id]\n",
    "    print(len(subset))\n",
    "    \n",
    "    mig_hist_data = []\n",
    "    conipher_data = []\n",
    "    for _, row in subset.iterrows(): # each row is one mutation\n",
    "        generator = generate_sample_and_mut_info(row)\n",
    "        for data in generator:\n",
    "            if data == None:\n",
    "                continue\n",
    "            mig_hist_input, conipher_input = data\n",
    "            mig_hist_data.append(mig_hist_input)\n",
    "            conipher_data.append(conipher_input)\n",
    "    mig_hist_df = format_mig_hist_df(mig_hist_data)\n",
    "    conipher_df = format_conipher_df(conipher_data)\n",
    "    # Write migration history inputs\n",
    "    mig_hist_df.to_csv(os.path.join(OUTPUT_DIR, f\"{patient_id}_SNVs.tsv\"), sep=\"\\t\")\n",
    "    # Write conipher inputs\n",
    "    conipher_df.to_csv(os.path.join(CONIPHER_DIR, f\"{patient_id}_conipher_SNVs.tsv\"), sep=\"\\t\")\n",
    "    # Write pairtree inputs\n",
    "    pt_util.write_pairtree_inputs(mig_hist_df, patient_id, OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e90b6f69",
   "metadata": {},
   "source": [
    "### Above is all on individual mutations. **After clustering mutations using PyClone**, we have to prep clustered inputs for PairTree/Orchard and Metient"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "715a7b5b-7889-4f9d-820d-4433f21adadd",
   "metadata": {},
   "source": [
    "#### Take PyClone clusters and create new params.json files with these clusters so PairTree/Orchard can be run on mutation clusters and not individual mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fa13352c-f79c-4511-aced-57dd2efddc61",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.util import pairtree_data_extraction_util as pt_util\n",
    "PYCLONE_CLUSTERING_DIR = os.path.join(TRACERX_DATA_DIR, 'conipher_outputs', 'TreeBuilding')\n",
    "for patient_id in unique_patients:\n",
    "    pt_util.write_clustered_params_json_from_pyclone_clusters(os.path.join(PYCLONE_CLUSTERING_DIR, f\"{patient_id}_conipher_SNVstreeTable_cleaned.tsv\"), \n",
    "                                                              os.path.join(OUTPUT_DIR, f\"{patient_id}.ssm\"),\n",
    "                                                              os.path.join(OUTPUT_DIR, f\"{patient_id}.params.json\"),\n",
    "                                                              os.path.join(OUTPUT_DIR, \"pyclone_clustered\"), patient_id)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "645c4a82",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/tracerx_nsclc/patient_data/CRUK0010_clustered.params.json'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 7\u001b[0m\n\u001b[1;32m      4\u001b[0m agg_rules \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mploidy\u001b[39m\u001b[38;5;124m'\u001b[39m:np\u001b[38;5;241m.\u001b[39mmean,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpurity\u001b[39m\u001b[38;5;124m'\u001b[39m:np\u001b[38;5;241m.\u001b[39mmean,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msample_type\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfirst\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124moriginal_anatomical_site_label\u001b[39m\u001b[38;5;124m'\u001b[39m:\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m;\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28mset\u001b[39m(x))}\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m patient_id \u001b[38;5;129;01min\u001b[39;00m unique_patients:\n\u001b[0;32m----> 7\u001b[0m     \u001b[43mpt_util\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_pooled_tsv_from_pairtree_clusters\u001b[49m\u001b[43m(\u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mOUTPUT_DIR\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mpatient_id\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_SNVS.tsv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m                                                    \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mOUTPUT_DIR\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mpatient_id\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m.ssm\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m                                                    \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mOUTPUT_DIR\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mpatient_id\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m_clustered.params.json\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m                                                    \u001b[49m\u001b[43magg_rules\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mOUTPUT_DIR\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpairtree_clustered\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpatient_id\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/util/pairtree_data_extraction_util.py:130\u001b[0m, in \u001b[0;36mwrite_pooled_tsv_from_pairtree_clusters\u001b[0;34m(tsv_fn, ssm_fn, clustered_params_json_fn, aggregation_rules, output_dir, patient_id, cluster_sep)\u001b[0m\n\u001b[1;32m    127\u001b[0m             mut_name_to_mut_id[items[\u001b[38;5;241m1\u001b[39m]] \u001b[38;5;241m=\u001b[39m items[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m    129\u001b[0m \u001b[38;5;66;03m# 2. Get pairtree cluster assignments\u001b[39;00m\n\u001b[0;32m--> 130\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mclustered_params_json_fn\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[1;32m    131\u001b[0m     cluster_json \u001b[38;5;241m=\u001b[39m json\u001b[38;5;241m.\u001b[39mloads(f\u001b[38;5;241m.\u001b[39mread())\n\u001b[1;32m    132\u001b[0m mut_name_to_cluster_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m()\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/tracerx_nsclc/patient_data/CRUK0010_clustered.params.json'"
     ]
    }
   ],
   "source": [
    "# import numpy as np\n",
    "# from src.util import pairtree_data_extraction_util as pt_util\n",
    "\n",
    "# agg_rules = {'ploidy':np.mean,'purity':np.mean,'sample_type':'first', 'original_anatomical_site_label':lambda x: ';'.join(set(x))}\n",
    "\n",
    "# for patient_id in unique_patients:\n",
    "#     pt_util.write_pooled_tsv_from_pairtree_clusters(os.path.join(OUTPUT_DIR, f\"{patient_id}_SNVS.tsv\"), \n",
    "#                                                     os.path.join(OUTPUT_DIR, f\"{patient_id}.ssm\"),\n",
    "#                                                     os.path.join(OUTPUT_DIR, f\"{patient_id}_clustered.params.json\"),\n",
    "#                                                     agg_rules, os.path.join(OUTPUT_DIR, \"pairtree_clustered\"), patient_id)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "725fe812",
   "metadata": {},
   "source": [
    "#### Take PyClone clusters and create pooled tsvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "13b00b9c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from src.util import pairtree_data_extraction_util as pt_util\n",
    "from src.util import data_extraction_util as dutil\n",
    "\n",
    "agg_rules = {'ploidy':np.mean,'purity':np.mean,'sample_type':'first', 'original_anatomical_site_label':lambda x: ';'.join(set(x))}\n",
    "for patient_id in unique_patients:\n",
    "    dutil.write_pooled_tsv_from_pyclone_clusters(os.path.join(OUTPUT_DIR, f\"{patient_id}_SNVS.tsv\"), \n",
    "                                                 os.path.join(PYCLONE_CLUSTERING_DIR, f\"{patient_id}_conipher_SNVstreeTable_cleaned.tsv\"),\n",
    "                                                 agg_rules, os.path.join(OUTPUT_DIR, \"pyclone_clustered\"),  patient_id )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "354210f6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mambaforge-met]",
   "language": "python",
   "name": "conda-env-mambaforge-met-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
