{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aaf0e53c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         E-H-131201-T1-1-D1-1\n",
       "1         E-H-131202-T1-1-D1-1\n",
       "2         E-H-131203-T1-1-D1-1\n",
       "3         E-H-131204-T1-1-D1-1\n",
       "4         E-H-131205-T1-1-D1-1\n",
       "                ...           \n",
       "242    IID_H201574_T01_01_WG01\n",
       "243    IID_H201574_T02_01_WG01\n",
       "244    IID_H201601_T01_01_WG01\n",
       "245    IID_H201601_T02_01_WG01\n",
       "246    IID_H201603_T01_01_WG01\n",
       "Name: DNA_SAMPLE, Length: 247, dtype: object"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import importlib\n",
    "\n",
    "from src.util import pairtree_data_extraction_util as pt_util\n",
    "\n",
    "# Process Gundem et.al. neuroblastoma dataset with putative oncogenic SNVs from WGS data\n",
    "REPO_DIR = os.path.join(os.getcwd(), \"../../\")\n",
    "os.chdir(REPO_DIR)\n",
    "\n",
    "GUNDEM_DATA_DIR = os.path.join(REPO_DIR, 'src/data/gundem_neuroblastoma_2023')\n",
    "OUTPUT_DIR = os.path.join(GUNDEM_DATA_DIR, 'patient_driver_genes')\n",
    "\n",
    "tumor_info = pd.read_excel(os.path.join(GUNDEM_DATA_DIR, \"gundem_neuroblastoma_supp_tables.xlsx\"), sheet_name=1)\n",
    "multimets_patient_info = pd.read_excel(os.path.join(GUNDEM_DATA_DIR, \"gundem_neuroblastoma_supp_tables.xlsx\"), sheet_name=3)\n",
    "\n",
    "# Possible values are (1) MSK-IMPACT (2) RNA-seq (3)WGS\n",
    "wgs_samples = tumor_info[tumor_info['SEQEUNCING_TYPE']==\"WGS\"][\"DNA_SAMPLE\"]\n",
    "wgs_samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4eead8ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1690\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNA_SAMPLE</th>\n",
       "      <th>PATIENT</th>\n",
       "      <th>PATIENT_DISEASE_SUBTYPE</th>\n",
       "      <th>SAMPLE_TYPE</th>\n",
       "      <th>VARIANT_TYPE</th>\n",
       "      <th>VARIANT_EFFECT</th>\n",
       "      <th>VARIANT_LOCUS</th>\n",
       "      <th>VARIANT_CHANGE</th>\n",
       "      <th>CHR1</th>\n",
       "      <th>START1</th>\n",
       "      <th>...</th>\n",
       "      <th>SUBCLONE_TYPE</th>\n",
       "      <th>IS_DRIVER</th>\n",
       "      <th>PATHWAY</th>\n",
       "      <th>tumor_type</th>\n",
       "      <th>anatomical_site_label</th>\n",
       "      <th>anatomical_site_broad</th>\n",
       "      <th>sampling_time_months_from_dx</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>purity</th>\n",
       "      <th>ploidy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>E-H-131203-T1-1-D1-1</td>\n",
       "      <td>H131203</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>SNV</td>\n",
       "      <td>non_synonymous_codon</td>\n",
       "      <td>PTPRT</td>\n",
       "      <td>p.X1398</td>\n",
       "      <td>20</td>\n",
       "      <td>40710659</td>\n",
       "      <td>...</td>\n",
       "      <td>private_met_clonal</td>\n",
       "      <td>clonality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary_disease</td>\n",
       "      <td>retroperitoneal - D</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>0.000</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>0.870</td>\n",
       "      <td>2.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>E-H-131203-T1-1-D1-1</td>\n",
       "      <td>H131203</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>SNV</td>\n",
       "      <td>non_synonymous_codon</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>p.G931V</td>\n",
       "      <td>1</td>\n",
       "      <td>27092771</td>\n",
       "      <td>...</td>\n",
       "      <td>private_met_clonal</td>\n",
       "      <td>clonality</td>\n",
       "      <td>CHROMATIN_MODELING</td>\n",
       "      <td>primary_disease</td>\n",
       "      <td>retroperitoneal - D</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>0.000</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>0.870</td>\n",
       "      <td>2.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>E-H-131203-T1-1-D1-1</td>\n",
       "      <td>H131203</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>Del</td>\n",
       "      <td>inframe_codon_loss</td>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>p.F1234_K1237del</td>\n",
       "      <td>19</td>\n",
       "      <td>11144117</td>\n",
       "      <td>...</td>\n",
       "      <td>shared_met_clonal</td>\n",
       "      <td>yes</td>\n",
       "      <td>CHROMATIN_MODELING</td>\n",
       "      <td>primary_disease</td>\n",
       "      <td>retroperitoneal - D</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>0.000</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>0.870</td>\n",
       "      <td>2.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>E-H-131203-T1-1-D1-1</td>\n",
       "      <td>H131203</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>SNV</td>\n",
       "      <td>non_synonymous_codon</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>p.A1690D</td>\n",
       "      <td>X</td>\n",
       "      <td>76888760</td>\n",
       "      <td>...</td>\n",
       "      <td>trunk</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary_disease</td>\n",
       "      <td>retroperitoneal - D</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>0.000</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>0.870</td>\n",
       "      <td>2.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>E-H-131203-T1-1-D1-1</td>\n",
       "      <td>H131203</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>SNV</td>\n",
       "      <td>non_synonymous_codon</td>\n",
       "      <td>FAT1</td>\n",
       "      <td>p.D2931N</td>\n",
       "      <td>4</td>\n",
       "      <td>187538949</td>\n",
       "      <td>...</td>\n",
       "      <td>trunk</td>\n",
       "      <td>clonality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>primary_disease</td>\n",
       "      <td>retroperitoneal - D</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>0.000</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>0.870</td>\n",
       "      <td>2.800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1836</th>\n",
       "      <td>I-H-135463-T1-2-D1-1</td>\n",
       "      <td>H135463</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>frelapse</td>\n",
       "      <td>Del</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>KMT2D</td>\n",
       "      <td>p.P4314Sfs*19</td>\n",
       "      <td>12</td>\n",
       "      <td>49425546</td>\n",
       "      <td>...</td>\n",
       "      <td>private_met_clonal</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>local_disease</td>\n",
       "      <td>aortocaval lymph node - R</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>12.867</td>\n",
       "      <td>relapse</td>\n",
       "      <td>0.340</td>\n",
       "      <td>2.550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>I-H-135463-T1-2-D1-1</td>\n",
       "      <td>H135463</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>frelapse</td>\n",
       "      <td>SNV</td>\n",
       "      <td>non_synonymous_codon</td>\n",
       "      <td>EPHA5</td>\n",
       "      <td>p.R356S</td>\n",
       "      <td>4</td>\n",
       "      <td>66356429</td>\n",
       "      <td>...</td>\n",
       "      <td>private_met_subclonal</td>\n",
       "      <td>clonality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>local_disease</td>\n",
       "      <td>aortocaval lymph node - R</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>12.867</td>\n",
       "      <td>relapse</td>\n",
       "      <td>0.340</td>\n",
       "      <td>2.550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1838</th>\n",
       "      <td>I-H-135463-T1-2-D1-1</td>\n",
       "      <td>H135463</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>frelapse</td>\n",
       "      <td>SNV</td>\n",
       "      <td>non_synonymous_codon</td>\n",
       "      <td>LATS1</td>\n",
       "      <td>p.K702N</td>\n",
       "      <td>6</td>\n",
       "      <td>150001498</td>\n",
       "      <td>...</td>\n",
       "      <td>private_met_clonal</td>\n",
       "      <td>clonality</td>\n",
       "      <td>NaN</td>\n",
       "      <td>local_disease</td>\n",
       "      <td>aortocaval lymph node - R</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>12.867</td>\n",
       "      <td>relapse</td>\n",
       "      <td>0.340</td>\n",
       "      <td>2.550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1839</th>\n",
       "      <td>I-H-135463-T1-2-D1-1</td>\n",
       "      <td>H135463</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>frelapse</td>\n",
       "      <td>SNV</td>\n",
       "      <td>non_synonymous_codon</td>\n",
       "      <td>ALK</td>\n",
       "      <td>p.F1174S</td>\n",
       "      <td>2</td>\n",
       "      <td>29443696</td>\n",
       "      <td>...</td>\n",
       "      <td>private_met_subclonal</td>\n",
       "      <td>yes</td>\n",
       "      <td>RAS_MAPK</td>\n",
       "      <td>local_disease</td>\n",
       "      <td>aortocaval lymph node - R</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>12.867</td>\n",
       "      <td>relapse</td>\n",
       "      <td>0.340</td>\n",
       "      <td>2.550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1840</th>\n",
       "      <td>I-H-135463-T1-2-D1-1</td>\n",
       "      <td>H135463</td>\n",
       "      <td>ATRX</td>\n",
       "      <td>frelapse</td>\n",
       "      <td>Ins</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>NOTCH2</td>\n",
       "      <td>p.R2298Sfs*15</td>\n",
       "      <td>1</td>\n",
       "      <td>120458453</td>\n",
       "      <td>...</td>\n",
       "      <td>private_met_subclonal</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>local_disease</td>\n",
       "      <td>aortocaval lymph node - R</td>\n",
       "      <td>retroperitoneal</td>\n",
       "      <td>12.867</td>\n",
       "      <td>relapse</td>\n",
       "      <td>0.340</td>\n",
       "      <td>2.550</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>426 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                DNA_SAMPLE  PATIENT PATIENT_DISEASE_SUBTYPE SAMPLE_TYPE  \\\n",
       "14    E-H-131203-T1-1-D1-1  H131203                    ATRX   diagnosis   \n",
       "15    E-H-131203-T1-1-D1-1  H131203                    ATRX   diagnosis   \n",
       "16    E-H-131203-T1-1-D1-1  H131203                    ATRX   diagnosis   \n",
       "17    E-H-131203-T1-1-D1-1  H131203                    ATRX   diagnosis   \n",
       "18    E-H-131203-T1-1-D1-1  H131203                    ATRX   diagnosis   \n",
       "...                    ...      ...                     ...         ...   \n",
       "1836  I-H-135463-T1-2-D1-1  H135463                    ATRX    frelapse   \n",
       "1837  I-H-135463-T1-2-D1-1  H135463                    ATRX    frelapse   \n",
       "1838  I-H-135463-T1-2-D1-1  H135463                    ATRX    frelapse   \n",
       "1839  I-H-135463-T1-2-D1-1  H135463                    ATRX    frelapse   \n",
       "1840  I-H-135463-T1-2-D1-1  H135463                    ATRX    frelapse   \n",
       "\n",
       "     VARIANT_TYPE        VARIANT_EFFECT VARIANT_LOCUS    VARIANT_CHANGE CHR1  \\\n",
       "14            SNV  non_synonymous_codon         PTPRT           p.X1398   20   \n",
       "15            SNV  non_synonymous_codon        ARID1A           p.G931V    1   \n",
       "16            Del    inframe_codon_loss       SMARCA4  p.F1234_K1237del   19   \n",
       "17            SNV  non_synonymous_codon          ATRX          p.A1690D    X   \n",
       "18            SNV  non_synonymous_codon          FAT1          p.D2931N    4   \n",
       "...           ...                   ...           ...               ...  ...   \n",
       "1836          Del    frameshift_variant         KMT2D     p.P4314Sfs*19   12   \n",
       "1837          SNV  non_synonymous_codon         EPHA5           p.R356S    4   \n",
       "1838          SNV  non_synonymous_codon         LATS1           p.K702N    6   \n",
       "1839          SNV  non_synonymous_codon           ALK          p.F1174S    2   \n",
       "1840          Ins    frameshift_variant        NOTCH2     p.R2298Sfs*15    1   \n",
       "\n",
       "         START1  ...          SUBCLONE_TYPE  IS_DRIVER             PATHWAY  \\\n",
       "14     40710659  ...     private_met_clonal  clonality                 NaN   \n",
       "15     27092771  ...     private_met_clonal  clonality  CHROMATIN_MODELING   \n",
       "16     11144117  ...      shared_met_clonal        yes  CHROMATIN_MODELING   \n",
       "17     76888760  ...                  trunk        yes                 NaN   \n",
       "18    187538949  ...                  trunk  clonality                 NaN   \n",
       "...         ...  ...                    ...        ...                 ...   \n",
       "1836   49425546  ...     private_met_clonal        yes                 NaN   \n",
       "1837   66356429  ...  private_met_subclonal  clonality                 NaN   \n",
       "1838  150001498  ...     private_met_clonal  clonality                 NaN   \n",
       "1839   29443696  ...  private_met_subclonal        yes            RAS_MAPK   \n",
       "1840  120458453  ...  private_met_subclonal        yes                 NaN   \n",
       "\n",
       "           tumor_type      anatomical_site_label anatomical_site_broad  \\\n",
       "14    primary_disease        retroperitoneal - D       retroperitoneal   \n",
       "15    primary_disease        retroperitoneal - D       retroperitoneal   \n",
       "16    primary_disease        retroperitoneal - D       retroperitoneal   \n",
       "17    primary_disease        retroperitoneal - D       retroperitoneal   \n",
       "18    primary_disease        retroperitoneal - D       retroperitoneal   \n",
       "...               ...                        ...                   ...   \n",
       "1836    local_disease  aortocaval lymph node - R       retroperitoneal   \n",
       "1837    local_disease  aortocaval lymph node - R       retroperitoneal   \n",
       "1838    local_disease  aortocaval lymph node - R       retroperitoneal   \n",
       "1839    local_disease  aortocaval lymph node - R       retroperitoneal   \n",
       "1840    local_disease  aortocaval lymph node - R       retroperitoneal   \n",
       "\n",
       "      sampling_time_months_from_dx  sample_type  purity ploidy  \n",
       "14                           0.000    diagnosis   0.870  2.800  \n",
       "15                           0.000    diagnosis   0.870  2.800  \n",
       "16                           0.000    diagnosis   0.870  2.800  \n",
       "17                           0.000    diagnosis   0.870  2.800  \n",
       "18                           0.000    diagnosis   0.870  2.800  \n",
       "...                            ...          ...     ...    ...  \n",
       "1836                        12.867      relapse   0.340  2.550  \n",
       "1837                        12.867      relapse   0.340  2.550  \n",
       "1838                        12.867      relapse   0.340  2.550  \n",
       "1839                        12.867      relapse   0.340  2.550  \n",
       "1840                        12.867      relapse   0.340  2.550  \n",
       "\n",
       "[426 rows x 31 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_data = pd.read_excel(os.path.join(GUNDEM_DATA_DIR, \"gundem_neuroblastoma_supp_tables.xlsx\"), sheet_name=2)\n",
    "\n",
    "tumor_type_short_map = {\n",
    "    \"frelapse\":\"FR\",\n",
    "    \"diagnosis\":\"D\",\n",
    "    \"relapse\":\"R\",\n",
    "    \"t-resection\":\"TR\",\n",
    "    \"reresection\":\"RR\"\n",
    "}\n",
    "# Only keep WGS samples\n",
    "variant_data = variant_data[variant_data[\"DNA_SAMPLE\"].isin(wgs_samples)]\n",
    "print(len(variant_data))\n",
    "# Add tumor info for each sample (anatomical site, primary vs. local met vs. distant met)\n",
    "def add_tumor_info(row, key):\n",
    "    return tumor_info[tumor_info[\"DNA_SAMPLE\"] == row[\"DNA_SAMPLE\"]][key].item()\n",
    "def add_site_info(row, site_key, type_key):\n",
    "    site = tumor_info[tumor_info[\"DNA_SAMPLE\"] == row[\"DNA_SAMPLE\"]][site_key].item()\n",
    "    tumor_type = tumor_info[tumor_info[\"DNA_SAMPLE\"] == row[\"DNA_SAMPLE\"]][type_key].item()\n",
    "    return f\"{site} - {tumor_type_short_map[tumor_type]}\"\n",
    "\n",
    "variant_data['tumor_type'] = variant_data.apply(lambda row: add_tumor_info(row, \"TUMOR_SITE_CATEGORY\"), axis=1)\n",
    "variant_data['anatomical_site_label'] = variant_data.apply(lambda row: add_site_info(row, \"TUMOR_SITE_DETAIL\", \"SAMPLE_TYPE\"), axis=1)\n",
    "variant_data['anatomical_site_broad'] = variant_data.apply(lambda row: add_tumor_info(row, \"TUMOR_SITE\"), axis=1)\n",
    "\n",
    "# additional useful patient data\n",
    "variant_data['sampling_time_months_from_dx'] = variant_data.apply(lambda row: add_tumor_info(row, \"SAMPLING_TIME_MONTHS_FROM_DX\"), axis=1)\n",
    "variant_data['sample_type'] = variant_data.apply(lambda row: add_tumor_info(row, \"SAMPLE_TYPE\"), axis=1)\n",
    "variant_data['purity'] = variant_data.apply(lambda row: add_tumor_info(row, \"TUMOR_PURITY\"), axis=1)\n",
    "variant_data['ploidy'] = variant_data.apply(lambda row: add_tumor_info(row, \"TUMOR_PLOIDY\"), axis=1)\n",
    "\n",
    "variant_data = variant_data.dropna(subset=['VAF', 'DEPTH']) # drop any NaNs\n",
    "variant_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d7fad831",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "frelapse       218\n",
       "diagnosis       77\n",
       "relapse         74\n",
       "t-resection     49\n",
       "reresection      8\n",
       "Name: sample_type, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_data[\"sample_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3077189d",
   "metadata": {},
   "source": [
    "## Find patients with at least one met and primary tumor sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4087811d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 patients with primaries and CNS mets\n",
      "10 patients with primaries and distant mets\n",
      "19 patients with primaries and mets (local+distant)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['H132374',\n",
       " 'H134819',\n",
       " 'H133120',\n",
       " 'H116984',\n",
       " 'H132387',\n",
       " 'H132384',\n",
       " 'H132380',\n",
       " 'H134821',\n",
       " 'H132392',\n",
       " 'H132382',\n",
       " 'H133121',\n",
       " 'H132379',\n",
       " 'H132388',\n",
       " 'H132396',\n",
       " 'H103207',\n",
       " 'H116988',\n",
       " 'H118706',\n",
       " 'H112909',\n",
       " 'H116991']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cns_met_patients = set(tumor_info[tumor_info[\"TUMOR_SITE\"] == \"CNS\"][\"PATIENT\"].unique())\n",
    "patients_with_primary_samples = set(variant_data[variant_data[\"tumor_type\"] == \"primary_disease\"][\"PATIENT\"].unique())\n",
    "\n",
    "cns_met_patients = list(cns_met_patients.intersection(patients_with_primary_samples))\n",
    "print(f\"{len(cns_met_patients)} patients with primaries and CNS mets\")\n",
    "\n",
    "distant_met_patients = set(variant_data[(variant_data[\"tumor_type\"] == \"distant_metastasis\")][\"PATIENT\"])\n",
    "distant_met_patients = list(distant_met_patients.intersection(patients_with_primary_samples))\n",
    "print(f\"{len(distant_met_patients)} patients with primaries and distant mets\")\n",
    "\n",
    "met_patients = set(variant_data[(variant_data[\"tumor_type\"] == \"distant_metastasis\") | (variant_data[\"tumor_type\"] == \"local_disease\")][\"PATIENT\"])\n",
    "met_patients = list(met_patients.intersection(patients_with_primary_samples))\n",
    "\n",
    "print(f\"{len(met_patients)} patients with primaries and mets (local+distant)\")\n",
    "met_patients"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c6a9c09",
   "metadata": {},
   "source": [
    "## Prep data for input into PairTree, Metient and MACHINA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1fcb7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_patient_data(patient_id):\n",
    "    '''\n",
    "    Output a tsv for each patient with columns: ['#sample_index', 'sample_label', 'anatomical_site_index',\n",
    "    'anatomical_site_label', 'character_index', 'character_label', 'ref', 'var']\n",
    "    '''\n",
    "    \n",
    "    def label_snv(row):\n",
    "        label = []\n",
    "        label += [str(row['VARIANT_LOCUS']), str(row['VARIANT_CHANGE']),str(row['VARIANT_EFFECT']),f\"{str(row['CHR1'])}:{str(row['START1'])}\"]\n",
    "        return (\"_\").join(label)\n",
    "\n",
    "    patient_data = variant_data[(variant_data[\"PATIENT\"]==patient_id)]\n",
    "    if patient_data.empty:\n",
    "        print(f\"No variant data for {patient_id}\")\n",
    "        return\n",
    "        \n",
    "    patient_data['character_label'] = patient_data.apply(lambda row: label_snv(row), axis=1)\n",
    "    patient_data['var'] = round(patient_data[\"VAF\"]*patient_data[\"DEPTH\"])\n",
    "    patient_data['ref'] = patient_data[\"DEPTH\"] - patient_data['var']\n",
    "\n",
    "    # Find the intersection of variants captured in the data by each sample\n",
    "    all_samples_variants = []\n",
    "    for sample in patient_data[\"DNA_SAMPLE\"].unique():   \n",
    "        subset = patient_data[patient_data[\"DNA_SAMPLE\"]==sample]\n",
    "        variants = subset[\"character_label\"]\n",
    "        all_samples_variants.append(set(variants))\n",
    "    intersecting_variants = list(set.intersection(*all_samples_variants))\n",
    "\n",
    "    patient_data = patient_data[patient_data[\"character_label\"].isin(intersecting_variants)]\n",
    "\n",
    "    patient_data = patient_data.rename(columns={\"DNA_SAMPLE\":\"sample_label\"})\n",
    "    num_variants = len(patient_data[\"character_label\"].unique())\n",
    "    if  (num_variants <= 1):\n",
    "        print(f\"Skipping patient {patient_id} with {num_variants} variant\")\n",
    "        return\n",
    "    # Add indices for mutations, samples and anatomical sites as needed for input format\n",
    "    patient_data['#sample_index'] = patient_data.apply(lambda row: list(patient_data['sample_label'].unique()).index(row[\"sample_label\"]), axis=1)    \n",
    "\n",
    "    # Some of these patients have multiple samples w/ same anatomical site label\n",
    "    if len(patient_data['anatomical_site_label'].unique()) != len(patient_data['sample_label'].unique()):\n",
    "        patient_data['anatomical_site_label'] = patient_data.apply(lambda row: row['anatomical_site_label'] + f\" {row['#sample_index']}\", axis=1)\n",
    "    \n",
    "    patient_data['character_index'] = patient_data.apply(lambda row: list(patient_data['character_label'].unique()).index(row[\"character_label\"]), axis=1)\n",
    "    patient_data['anatomical_site_index'] = patient_data.apply(lambda row: list(patient_data['anatomical_site_label'].unique()).index(row[\"anatomical_site_label\"]), axis=1)\n",
    "    \n",
    "    patient_data = patient_data[['#sample_index', \"sample_label\", \"anatomical_site_index\", \"anatomical_site_label\", \n",
    "                                 'character_index', \"character_label\", \"ref\",\"var\", \"tumor_type\", \"anatomical_site_broad\", \n",
    "                                 \"sample_type\", \"purity\", \"ploidy\", \"sampling_time_months_from_dx\"]]\n",
    "\n",
    "    patient_data.to_csv(os.path.join(OUTPUT_DIR, f\"{patient_id}_SNVs.tsv\"), sep=\"\\t\")\n",
    "    \n",
    "    pt_util.write_pairtree_inputs(patient_data, patient_id, OUTPUT_DIR)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d4165bdb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H132374\n",
      "H134819\n",
      "H133120\n",
      "H116984\n",
      "H132387\n",
      "H132384\n",
      "Skipping patient H132384 with 1 variant\n",
      "H132380\n",
      "Skipping patient H132380 with 1 variant\n",
      "H134821\n",
      "H132392\n",
      "Skipping patient H132392 with 1 variant\n",
      "H132382\n",
      "Skipping patient H132382 with 1 variant\n",
      "H133121\n",
      "H132379\n",
      "Skipping patient H132379 with 1 variant\n",
      "H132388\n",
      "H132396\n",
      "Skipping patient H132396 with 1 variant\n",
      "H103207\n",
      "H116988\n",
      "Skipping patient H116988 with 1 variant\n",
      "H118706\n",
      "H112909\n",
      "Skipping patient H112909 with 1 variant\n",
      "H116991\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/divyakoyyalagunta/mambaforge/envs/metf/lib/python3.7/site-packages/ipykernel_launcher.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  app.launch_new_instance()\n",
      "/Users/divyakoyyalagunta/mambaforge/envs/metf/lib/python3.7/site-packages/ipykernel_launcher.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/divyakoyyalagunta/mambaforge/envs/metf/lib/python3.7/site-packages/ipykernel_launcher.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "for patient_id in met_patients:\n",
    "    output_name = patient_id\n",
    "    print(patient_id)\n",
    "    preprocess_patient_data(patient_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a9c4629",
   "metadata": {},
   "source": [
    "### Above is all on individual mutations. We can also cluster mutations by running PairTree's clustervars, and then pooling together those clustered mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8b43b6f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H133120_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H134821_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H116984_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H132388_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H132388_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H103207_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H118706_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H133120_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H133121_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H134819_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H118706_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H132387_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H116991_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H116991_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H103207_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H132374_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H132374_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H134819_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H116984_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H132387_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H134821_clustered_SNVs.tsv', '/Users/divyakoyyalagunta/Desktop/Cornell_Research/Morris_Lab/met_history_prediction/src/jupyter_notebooks/../../src/data/gundem_neuroblastoma_2023/patient_driver_genes/H133121_clustered_SNVs.tsv']\n",
      "H133120\n",
      "{'ATRX_p.R250*_stop_gained_X:76940000': 'm0', 'ALK_p.R1275Q_non_synonymous_codon_2:29432664': 'm1'}\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'ATRX_p.R250*_stop_gained_X:76940000;ALK_p.R1275Q_non_synonymous_codon_2:29432664'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/35/l3stl61s6n14w337drj2gt9m0000gn/T/ipykernel_19666/4287793027.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0mcluster_assignments\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mmut_name\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'character_label'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m         \u001b[0mcluster_assignments\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_cluster_id\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmut_name_to_id\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmut_name\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcluster_json\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'clusters'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m     \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'cluster'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcluster_assignments\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'ATRX_p.R250*_stop_gained_X:76940000;ALK_p.R1275Q_non_synonymous_codon_2:29432664'"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "import numpy as np\n",
    "import json\n",
    "\n",
    "\n",
    "def get_cluster_id(mut_id, clusters):\n",
    "    '''\n",
    "    Given mut_id = 'm1' and clusters= [['m1'], ['m0', 'm2']], returns 1 \n",
    "    '''\n",
    "    for i, cluster in enumerate(clusters):\n",
    "        if mut_id in cluster:\n",
    "            return i\n",
    "    \n",
    "    print(f\"mut id {mut_id} not found\")\n",
    "    return None\n",
    "\n",
    "for patient_id in met_patients:\n",
    "    patient_tsv = os.path.join(OUTPUT_DIR, f\"{patient_id}_SNVs.tsv)\n",
    "    print(patient_id)\n",
    "    df = pd.read_csv(patient_tsv, delimiter=\"\\t\")\n",
    "    \n",
    "    # 1. Get mapping between mutation names and pairtree mutation_ids\n",
    "    mut_name_to_id = dict()\n",
    "    with open(os.path.join(OUTPUT_DIR, f\"{patient_id}.ssm\")) as f:\n",
    "        for i, line in enumerate(f):\n",
    "            if i == 0: continue\n",
    "            items = line.split(\"\\t\")\n",
    "            if items[1] not in mut_name_to_id:\n",
    "                mut_name_to_id[items[1]] = items[0]\n",
    "                \n",
    "    print(mut_name_to_id)\n",
    "    # 2. Get pairtree cluster assignments\n",
    "    with open(os.path.join(OUTPUT_DIR, f\"{patient_id}_clustered.params.json\")) as f:\n",
    "        cluster_json = json.loads(f.read())\n",
    "    cluster_assignments = []\n",
    "    for mut_name in df['character_label']:\n",
    "        cluster_assignments.append(get_cluster_id(mut_name_to_id[mut_name], cluster_json['clusters']))\n",
    "    df['cluster'] = cluster_assignments\n",
    "    \n",
    "    # 3. Pool reference and variant allele counts from all mutations within a cluster\n",
    "    pooled_df = df.drop(['character_label', 'character_index', '#sample_index', 'anatomical_site_index'], axis=1)\n",
    "    # We can use \"first\" aggregation rules for the string columns since it is information shared by a sample\n",
    "    pooled_df = pooled_df.groupby(['cluster', 'sample_label'], as_index=False).agg({'ref': np.sum, 'var': np.sum, \n",
    "                                                                                    'purity': np.mean, 'ploidy':np.mean, \n",
    "                                                                                    'sampling_time_months_from_dx':np.mean, \n",
    "                                                                                    'anatomical_site_label':'first', \n",
    "                                                                                    'sample_type':'first', 'tumor_type':'first',\n",
    "                                                                                    'anatomical_site_broad':'first'})\n",
    "    # 4. Add new names for clustered mutations\n",
    "    mut_id_to_name = {v:k for k,v in mut_name_to_id.items()}\n",
    "    cluster_id_to_cluster_name = dict()\n",
    "    for i, cluster in enumerate(cluster_json['clusters']):\n",
    "        cluster_comps = []\n",
    "        for mut in cluster:\n",
    "            cluster_comps.append(mut_id_to_name[mut])\n",
    "        cluster_id_to_cluster_name[i] = \";\".join(cluster_comps)\n",
    "    print(cluster_id_to_cluster_name)\n",
    "    pooled_df['character_label'] = pooled_df.apply(lambda row: cluster_id_to_cluster_name[row['cluster']], axis=1)\n",
    "    \n",
    "    # Add indices for mutations, samples and anatomical sites as needed for input format\n",
    "    pooled_df['character_index'] = pooled_df.apply(lambda row: list(pooled_df['character_label'].unique()).index(row[\"character_label\"]), axis=1)\n",
    "    pooled_df['anatomical_site_index'] = pooled_df.apply(lambda row: list(pooled_df['anatomical_site_label'].unique()).index(row[\"anatomical_site_label\"]), axis=1)\n",
    "    pooled_df['#sample_index'] = pooled_df.apply(lambda row: list(pooled_df['sample_label'].unique()).index(row[\"sample_label\"]), axis=1)    \n",
    "    \n",
    "    pooled_df = pooled_df[['#sample_index', \"sample_label\", \"anatomical_site_index\", \"anatomical_site_label\", \n",
    "                             'character_index', \"character_label\", \"ref\",\"var\", \"tumor_type\", \"anatomical_site_broad\", \n",
    "                             \"sample_type\", \"purity\", \"ploidy\", \"sampling_time_months_from_dx\"]]\n",
    "    pooled_df.to_csv(os.path.join(OUTPUT_DIR, f\"{patient_id}_clustered_SNVs.tsv\"), sep=\"\\t\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44c2086b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:metf] *",
   "language": "python",
   "name": "conda-env-metf-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
