{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "04e0eccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import os\n",
    "from metient.util import data_extraction_util as dutil\n",
    "from metient.util import vertex_labeling_util as vutil\n",
    "\n",
    "repo_dir = os.path.join(os.getcwd(), \"../\")\n",
    "data_dir = os.path.join(repo_dir, 'data', 'mcpherson_ovarian_2016')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d36464b",
   "metadata": {},
   "source": [
    "## (1) Read in data from McPherson et.al. supplemental tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ee814dd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient IDs: [ 1  2  3  4  7  9 10]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>malignant</th>\n",
       "      <th>primer_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref_counts</th>\n",
       "      <th>alt_counts</th>\n",
       "      <th>depth</th>\n",
       "      <th>alt_freq</th>\n",
       "      <th>background_average_alt_freq</th>\n",
       "      <th>ref_p_value</th>\n",
       "      <th>alt_p_value</th>\n",
       "      <th>status</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>snpeff_impact</th>\n",
       "      <th>character_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>normal_blood</td>\n",
       "      <td>1</td>\n",
       "      <td>no</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>641</td>\n",
       "      <td>0</td>\n",
       "      <td>641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>normal_sample</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>omentum_site_1</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>449</td>\n",
       "      <td>0</td>\n",
       "      <td>449</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>395</td>\n",
       "      <td>64</td>\n",
       "      <td>459</td>\n",
       "      <td>0.139434</td>\n",
       "      <td>0.002375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.800000e-90</td>\n",
       "      <td>somatic</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>416</td>\n",
       "      <td>0.002404</td>\n",
       "      <td>0.002754</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.825260e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>1</td>\n",
       "      <td>yes</td>\n",
       "      <td>1_b_amplicrazy</td>\n",
       "      <td>1</td>\n",
       "      <td>17090971.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>609</td>\n",
       "      <td>4</td>\n",
       "      <td>613</td>\n",
       "      <td>0.006525</td>\n",
       "      <td>0.002145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.440600e-02</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1:17090971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15293</th>\n",
       "      <td>normal_blood</td>\n",
       "      <td>10</td>\n",
       "      <td>no</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3669</td>\n",
       "      <td>10</td>\n",
       "      <td>3679</td>\n",
       "      <td>0.002718</td>\n",
       "      <td>0.002239</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.130860e-01</td>\n",
       "      <td>normal_sample</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15294</th>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2037</td>\n",
       "      <td>295</td>\n",
       "      <td>2332</td>\n",
       "      <td>0.126501</td>\n",
       "      <td>0.002306</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>somatic</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15295</th>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1369</td>\n",
       "      <td>4</td>\n",
       "      <td>1373</td>\n",
       "      <td>0.002913</td>\n",
       "      <td>0.002442</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.314750e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15296</th>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2675</td>\n",
       "      <td>4</td>\n",
       "      <td>2679</td>\n",
       "      <td>0.001493</td>\n",
       "      <td>0.002189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.364490e-01</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15297</th>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>10</td>\n",
       "      <td>yes</td>\n",
       "      <td>10_b_amplicrazy</td>\n",
       "      <td>X</td>\n",
       "      <td>110000000.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3044</td>\n",
       "      <td>20</td>\n",
       "      <td>3064</td>\n",
       "      <td>0.006527</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.430000e-06</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>RGAG1</td>\n",
       "      <td>LOW</td>\n",
       "      <td>RGAG1:X:110000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15298 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample_id  patient_id malignant        primer_id chrom  \\\n",
       "0            normal_blood           1        no   1_b_amplicrazy     1   \n",
       "1          omentum_site_1           1       yes   1_b_amplicrazy     1   \n",
       "2      right_ovary_site_1           1       yes   1_b_amplicrazy     1   \n",
       "3      right_ovary_site_2           1       yes   1_b_amplicrazy     1   \n",
       "4      right_ovary_site_3           1       yes   1_b_amplicrazy     1   \n",
       "...                   ...         ...       ...              ...   ...   \n",
       "15293        normal_blood          10        no  10_b_amplicrazy     X   \n",
       "15294  right_ovary_site_1          10       yes  10_b_amplicrazy     X   \n",
       "15295  right_ovary_site_2          10       yes  10_b_amplicrazy     X   \n",
       "15296  right_ovary_site_3          10       yes  10_b_amplicrazy     X   \n",
       "15297  right_ovary_site_4          10       yes  10_b_amplicrazy     X   \n",
       "\n",
       "             coord ref alt  ref_counts  alt_counts  depth  alt_freq  \\\n",
       "0       17090971.0   G   A         641           0    641  0.000000   \n",
       "1       17090971.0   G   A         449           0    449  0.000000   \n",
       "2       17090971.0   G   A         395          64    459  0.139434   \n",
       "3       17090971.0   G   A         415           1    416  0.002404   \n",
       "4       17090971.0   G   A         609           4    613  0.006525   \n",
       "...            ...  ..  ..         ...         ...    ...       ...   \n",
       "15293  110000000.0   A   G        3669          10   3679  0.002718   \n",
       "15294  110000000.0   A   G        2037         295   2332  0.126501   \n",
       "15295  110000000.0   A   G        1369           4   1373  0.002913   \n",
       "15296  110000000.0   A   G        2675           4   2679  0.001493   \n",
       "15297  110000000.0   A   G        3044          20   3064  0.006527   \n",
       "\n",
       "       background_average_alt_freq  ref_p_value   alt_p_value         status  \\\n",
       "0                         0.002034          0.0  1.000000e+00  normal_sample   \n",
       "1                         0.001597          0.0  1.000000e+00       wildtype   \n",
       "2                         0.002375          0.0  7.800000e-90        somatic   \n",
       "3                         0.002754          0.0  6.825260e-01       wildtype   \n",
       "4                         0.002145          0.0  4.440600e-02       wildtype   \n",
       "...                            ...          ...           ...            ...   \n",
       "15293                     0.002239          0.0  3.130860e-01  normal_sample   \n",
       "15294                     0.002306          0.0  0.000000e+00        somatic   \n",
       "15295                     0.002442          0.0  4.314750e-01       wildtype   \n",
       "15296                     0.002189          0.0  8.364490e-01       wildtype   \n",
       "15297                     0.001969          0.0  5.430000e-06       wildtype   \n",
       "\n",
       "      gene_name snpeff_impact    character_label  \n",
       "0           nan           NaN         1:17090971  \n",
       "1           nan           NaN         1:17090971  \n",
       "2           nan           NaN         1:17090971  \n",
       "3           nan           NaN         1:17090971  \n",
       "4           nan           NaN         1:17090971  \n",
       "...         ...           ...                ...  \n",
       "15293     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15294     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15295     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15296     RGAG1           LOW  RGAG1:X:110000000  \n",
       "15297     RGAG1           LOW  RGAG1:X:110000000  \n",
       "\n",
       "[15298 rows x 19 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First, read in targeted deep sequencing read data\n",
    "# Taken directly from supplementary table 10 of McPherson et.al.\n",
    "reads_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_10.csv'))\n",
    "patient_ids = reads_df['patient_id'].unique()\n",
    "print(\"patient IDs:\", patient_ids)\n",
    "reads_df['gene_name'] = reads_df['gene_name'].astype(str)\n",
    "def label_snv(row):\n",
    "    label = []\n",
    "    if row['gene_name'] != \"nan\":\n",
    "        label.append(row['gene_name'])\n",
    "    label += [str(row['chrom']), str(int(row['coord']))]\n",
    "    return (\":\").join(label)\n",
    "reads_df['character_label'] = reads_df.apply(lambda row: label_snv(row), axis=1)\n",
    "reads_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "76e24850",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>clone_id</th>\n",
       "      <th>pyclone_cluster_id</th>\n",
       "      <th>present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>10</td>\n",
       "      <td>F</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>10</td>\n",
       "      <td>F</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>10</td>\n",
       "      <td>F</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>10</td>\n",
       "      <td>F</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>10</td>\n",
       "      <td>F</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>377 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     patient_id clone_id  pyclone_cluster_id  present\n",
       "0             1        A                   2        1\n",
       "1             1        A                   3        0\n",
       "2             1        A                   4        0\n",
       "3             1        A                   5        1\n",
       "4             1        A                   6        1\n",
       "..          ...      ...                 ...      ...\n",
       "372          10        F                   3        1\n",
       "373          10        F                   7        1\n",
       "374          10        F                  11        1\n",
       "375          10        F                  12        0\n",
       "376          10        F                  13        1\n",
       "\n",
       "[377 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyclone_to_clone_id_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_8.csv'))\n",
    "pyclone_to_clone_id_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eefb25a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>paper_id</th>\n",
       "      <th>malignant</th>\n",
       "      <th>discovery_sample</th>\n",
       "      <th>tissue_source</th>\n",
       "      <th>anatomy</th>\n",
       "      <th>anatomical_site_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>appendix_site_c1</td>\n",
       "      <td>ApC1</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>ffpe</td>\n",
       "      <td>Appendix</td>\n",
       "      <td>Appendix</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>left_fallopian_tube_site_b4</td>\n",
       "      <td>LFTB4</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>ffpe</td>\n",
       "      <td>Left Fallopian Tube</td>\n",
       "      <td>Left Fallopian Tube</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>left_ovary_site_b2</td>\n",
       "      <td>LOvB2</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>ffpe</td>\n",
       "      <td>Left Ovary</td>\n",
       "      <td>Left Ovary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>right_fallopian_tube_site_a16</td>\n",
       "      <td>RFTA16</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>ffpe</td>\n",
       "      <td>Right Fallopian Tube</td>\n",
       "      <td>Right Fallopian Tube</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>right_ovary_site_a4</td>\n",
       "      <td>ROvA4</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>ffpe</td>\n",
       "      <td>Right Ovary</td>\n",
       "      <td>Right Ovary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_a9</td>\n",
       "      <td>ROvA9</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>ffpe</td>\n",
       "      <td>Right Ovary</td>\n",
       "      <td>Right Ovary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>fresh_frozen</td>\n",
       "      <td>Right Ovary Site 1</td>\n",
       "      <td>Right Ovary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>ROv2</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>fresh_frozen</td>\n",
       "      <td>Right Ovary Site 2</td>\n",
       "      <td>Right Ovary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>ROv3</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>fresh_frozen</td>\n",
       "      <td>Right Ovary Site 3</td>\n",
       "      <td>Right Ovary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>ROv4</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>fresh_frozen</td>\n",
       "      <td>Right Ovary Site 4</td>\n",
       "      <td>Right Ovary</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    patient_id                      sample_id paper_id malignant  \\\n",
       "1            1               appendix_site_c1     ApC1       yes   \n",
       "2            1    left_fallopian_tube_site_b4    LFTB4       yes   \n",
       "3            1             left_ovary_site_b2    LOvB2       yes   \n",
       "4            1  right_fallopian_tube_site_a16   RFTA16       yes   \n",
       "5            1            right_ovary_site_a4    ROvA4       yes   \n",
       "..         ...                            ...      ...       ...   \n",
       "70          10            right_ovary_site_a9    ROvA9       yes   \n",
       "71          10             right_ovary_site_1     ROv1       yes   \n",
       "72          10             right_ovary_site_2     ROv2       yes   \n",
       "73          10             right_ovary_site_3     ROv3       yes   \n",
       "74          10             right_ovary_site_4     ROv4       yes   \n",
       "\n",
       "   discovery_sample tissue_source               anatomy anatomical_site_label  \n",
       "1                no          ffpe              Appendix              Appendix  \n",
       "2                no          ffpe   Left Fallopian Tube   Left Fallopian Tube  \n",
       "3                no          ffpe            Left Ovary            Left Ovary  \n",
       "4                no          ffpe  Right Fallopian Tube  Right Fallopian Tube  \n",
       "5                no          ffpe           Right Ovary           Right Ovary  \n",
       "..              ...           ...                   ...                   ...  \n",
       "70               no          ffpe           Right Ovary           Right Ovary  \n",
       "71              yes  fresh_frozen    Right Ovary Site 1           Right Ovary  \n",
       "72              yes  fresh_frozen    Right Ovary Site 2           Right Ovary  \n",
       "73              yes  fresh_frozen    Right Ovary Site 3           Right Ovary  \n",
       "74              yes  fresh_frozen    Right Ovary Site 4           Right Ovary  \n",
       "\n",
       "[68 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df = pd.read_csv(os.path.join(data_dir, 'supplemental_table_1.csv'))\n",
    "sample_df = sample_df[sample_df['sample_id']!='normal_blood']\n",
    "sample_df['anatomical_site_label'] = sample_df.apply(lambda row: ''.join([i for i in row['anatomy'] if not i.isdigit()]).replace(\"Site\", \"\").strip(), axis=1)\n",
    "sample_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "5d4bd56c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Appendix', 'Left Fallopian Tube', 'Left Ovary',\n",
       "       'Right Fallopian Tube', 'Right Ovary', 'Small Bowel', 'Omentum',\n",
       "       'Cul de Sac', 'Sigmoid Colon Deposit',\n",
       "       'Left Fallopian Tube Fimbriae', 'Adnexa', 'Left Ovary Surface',\n",
       "       'Left Pelvic Sidewall', 'Right Pelvic Sidewall',\n",
       "       'Brain Metastasis', 'Bowel Implant', 'Right Uterosacral',\n",
       "       'Righ Pelvic Mass'], dtype=object)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df['anatomical_site_label'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b4c9ceaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>paper_id</th>\n",
       "      <th>clone_id</th>\n",
       "      <th>prevalence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ApC1</td>\n",
       "      <td>A</td>\n",
       "      <td>0.055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ApC1</td>\n",
       "      <td>B</td>\n",
       "      <td>0.936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ApC1</td>\n",
       "      <td>C</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ApC1</td>\n",
       "      <td>D</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ApC1</td>\n",
       "      <td>E</td>\n",
       "      <td>0.001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>10</td>\n",
       "      <td>ROvA9</td>\n",
       "      <td>B</td>\n",
       "      <td>0.241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>10</td>\n",
       "      <td>ROvA9</td>\n",
       "      <td>C</td>\n",
       "      <td>0.246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>10</td>\n",
       "      <td>ROvA9</td>\n",
       "      <td>D</td>\n",
       "      <td>0.009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>10</td>\n",
       "      <td>ROvA9</td>\n",
       "      <td>E</td>\n",
       "      <td>0.229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>10</td>\n",
       "      <td>ROvA9</td>\n",
       "      <td>F</td>\n",
       "      <td>0.188</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>409 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     patient_id paper_id clone_id  prevalence\n",
       "0             1     ApC1        A       0.055\n",
       "1             1     ApC1        B       0.936\n",
       "2             1     ApC1        C       0.003\n",
       "3             1     ApC1        D       0.002\n",
       "4             1     ApC1        E       0.001\n",
       "..          ...      ...      ...         ...\n",
       "404          10    ROvA9        B       0.241\n",
       "405          10    ROvA9        C       0.246\n",
       "406          10    ROvA9        D       0.009\n",
       "407          10    ROvA9        E       0.229\n",
       "408          10    ROvA9        F       0.188\n",
       "\n",
       "[409 rows x 4 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevalences_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_9.csv'))\n",
    "prevalences_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24174bef",
   "metadata": {},
   "source": [
    "From McPherson et.al.:\n",
    "Presence of each clone in each site was calculated as follows. First, samples were grouped into more broad\n",
    "anatomic locations, for instance ROv1, 2 etc. were grouped as the ROv site. A clone was said to be present in a\n",
    "sample if it was assigned a clonal prevalence greater than 0.01. Parent clones are often predicted to coexist with child\n",
    "clones at small proportions, though based on single cell data, such a scenario is likely an artifact. Parent and child\n",
    "clones are usually distinguished by an absence of mutations in the parent. If the mutations that identify the child\n",
    "clone are estimated to have slightly lower prevalence than more ancestral mutations, that slight deviation will results\n",
    "in prediction of a minor population of the parent clone in the sample. On the other hand, a sample composed of 95%\n",
    "parent and 5% child is more likely to represent a true mixture. To account for these issues, we use an additional rule\n",
    "for determining whether a parent clone co-exists with its child in a sample. A predicted cellular prevalence of non-leaf\n",
    "clone X must be at least 50% of the combined cellular prevalence of X and its descendants to be considered present in\n",
    "the sample. A clone is said to be present in a site if the clone, by the above rules, is present in any of the samples of\n",
    "that site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "516b6809",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A' 'B' 'C' 'D' 'E' 'F' 'G' 'H' 'I'] ['Appendix' 'Left Fallopian Tube' 'Left Ovary' 'Right Fallopian Tube'\n",
      " 'Right Ovary' 'Small Bowel' 'Omentum']\n",
      "A Left Fallopian Tube\n",
      "A Right Fallopian Tube\n",
      "B Appendix\n",
      "B Left Fallopian Tube\n",
      "B Left Ovary\n",
      "B Small Bowel\n",
      "B Omentum\n",
      "C Small Bowel\n",
      "D Left Ovary\n",
      "E Right Ovary\n",
      "G Right Ovary\n",
      "H Left Ovary\n",
      "H Right Ovary\n",
      "I Right Ovary\n",
      "   anatomical_site_label  cluster_index cluster_label  present site_category  \\\n",
      "0               Appendix              0             A        0    metastasis   \n",
      "1    Left Fallopian Tube              0             A        1    metastasis   \n",
      "2             Left Ovary              0             A        0       primary   \n",
      "3   Right Fallopian Tube              0             A        1    metastasis   \n",
      "4            Right Ovary              0             A        0       primary   \n",
      "..                   ...            ...           ...      ...           ...   \n",
      "58            Left Ovary              8             I        0       primary   \n",
      "59  Right Fallopian Tube              8             I        0    metastasis   \n",
      "60           Right Ovary              8             I        1       primary   \n",
      "61           Small Bowel              8             I        0    metastasis   \n",
      "62               Omentum              8             I        0    metastasis   \n",
      "\n",
      "    num_mutations  \n",
      "0               1  \n",
      "1               1  \n",
      "2               1  \n",
      "3               1  \n",
      "4               1  \n",
      "..            ...  \n",
      "58              1  \n",
      "59              1  \n",
      "60              1  \n",
      "61              1  \n",
      "62              1  \n",
      "\n",
      "[63 rows x 6 columns]\n",
      "['A' 'B' 'C' 'D' 'E' 'F'] ['Left Ovary' 'Omentum' 'Right Fallopian Tube' 'Right Ovary']\n",
      "B Omentum\n",
      "D Right Ovary\n",
      "F Right Ovary\n",
      "   anatomical_site_label  cluster_index cluster_label  present site_category  \\\n",
      "0             Left Ovary              0             A        0       primary   \n",
      "1                Omentum              0             A        0    metastasis   \n",
      "2   Right Fallopian Tube              0             A        0    metastasis   \n",
      "3            Right Ovary              0             A        0       primary   \n",
      "4             Left Ovary              1             B        0       primary   \n",
      "5                Omentum              1             B        1    metastasis   \n",
      "6   Right Fallopian Tube              1             B        0    metastasis   \n",
      "7            Right Ovary              1             B        0       primary   \n",
      "8             Left Ovary              2             C        0       primary   \n",
      "9                Omentum              2             C        0    metastasis   \n",
      "10  Right Fallopian Tube              2             C        0    metastasis   \n",
      "11           Right Ovary              2             C        0       primary   \n",
      "12            Left Ovary              3             D        0       primary   \n",
      "13               Omentum              3             D        0    metastasis   \n",
      "14  Right Fallopian Tube              3             D        0    metastasis   \n",
      "15           Right Ovary              3             D        1       primary   \n",
      "16            Left Ovary              4             E        0       primary   \n",
      "17               Omentum              4             E        0    metastasis   \n",
      "18  Right Fallopian Tube              4             E        0    metastasis   \n",
      "19           Right Ovary              4             E        0       primary   \n",
      "20            Left Ovary              5             F        0       primary   \n",
      "21               Omentum              5             F        0    metastasis   \n",
      "22  Right Fallopian Tube              5             F        0    metastasis   \n",
      "23           Right Ovary              5             F        1       primary   \n",
      "\n",
      "    num_mutations  \n",
      "0               1  \n",
      "1               1  \n",
      "2               1  \n",
      "3               1  \n",
      "4               1  \n",
      "5               1  \n",
      "6               1  \n",
      "7               1  \n",
      "8               1  \n",
      "9               1  \n",
      "10              1  \n",
      "11              1  \n",
      "12              1  \n",
      "13              1  \n",
      "14              1  \n",
      "15              1  \n",
      "16              1  \n",
      "17              1  \n",
      "18              1  \n",
      "19              1  \n",
      "20              1  \n",
      "21              1  \n",
      "22              1  \n",
      "23              1  \n",
      "['A' 'B' 'C' 'D' 'E' 'F' 'G'] ['Cul de Sac' 'Sigmoid Colon Deposit' 'Left Fallopian Tube Fimbriae'\n",
      " 'Left Ovary' 'Omentum' 'Right Fallopian Tube' 'Right Ovary' 'Adnexa']\n",
      "A Left Ovary\n",
      "A Right Fallopian Tube\n",
      "B Cul de Sac\n",
      "B Sigmoid Colon Deposit\n",
      "B Omentum\n",
      "C Sigmoid Colon Deposit\n",
      "C Left Ovary\n",
      "D Cul de Sac\n",
      "D Sigmoid Colon Deposit\n",
      "D Left Fallopian Tube Fimbriae\n",
      "D Left Ovary\n",
      "D Omentum\n",
      "D Right Fallopian Tube\n",
      "D Right Ovary\n",
      "E Right Ovary\n",
      "F Left Fallopian Tube Fimbriae\n",
      "F Left Ovary\n",
      "F Omentum\n",
      "G Adnexa\n",
      "           anatomical_site_label  cluster_index cluster_label  present  \\\n",
      "0                     Cul de Sac              0             A        0   \n",
      "1          Sigmoid Colon Deposit              0             A        0   \n",
      "2   Left Fallopian Tube Fimbriae              0             A        0   \n",
      "3                     Left Ovary              0             A        1   \n",
      "4                        Omentum              0             A        0   \n",
      "5           Right Fallopian Tube              0             A        1   \n",
      "6                    Right Ovary              0             A        0   \n",
      "7                         Adnexa              0             A        0   \n",
      "8                     Cul de Sac              1             B        1   \n",
      "9          Sigmoid Colon Deposit              1             B        1   \n",
      "10  Left Fallopian Tube Fimbriae              1             B        0   \n",
      "11                    Left Ovary              1             B        0   \n",
      "12                       Omentum              1             B        1   \n",
      "13          Right Fallopian Tube              1             B        0   \n",
      "14                   Right Ovary              1             B        0   \n",
      "15                        Adnexa              1             B        0   \n",
      "16                    Cul de Sac              2             C        0   \n",
      "17         Sigmoid Colon Deposit              2             C        1   \n",
      "18  Left Fallopian Tube Fimbriae              2             C        0   \n",
      "19                    Left Ovary              2             C        1   \n",
      "20                       Omentum              2             C        0   \n",
      "21          Right Fallopian Tube              2             C        0   \n",
      "22                   Right Ovary              2             C        0   \n",
      "23                        Adnexa              2             C        0   \n",
      "24                    Cul de Sac              3             D        1   \n",
      "25         Sigmoid Colon Deposit              3             D        1   \n",
      "26  Left Fallopian Tube Fimbriae              3             D        1   \n",
      "27                    Left Ovary              3             D        1   \n",
      "28                       Omentum              3             D        1   \n",
      "29          Right Fallopian Tube              3             D        1   \n",
      "30                   Right Ovary              3             D        1   \n",
      "31                        Adnexa              3             D        0   \n",
      "32                    Cul de Sac              4             E        0   \n",
      "33         Sigmoid Colon Deposit              4             E        0   \n",
      "34  Left Fallopian Tube Fimbriae              4             E        0   \n",
      "35                    Left Ovary              4             E        0   \n",
      "36                       Omentum              4             E        0   \n",
      "37          Right Fallopian Tube              4             E        0   \n",
      "38                   Right Ovary              4             E        1   \n",
      "39                        Adnexa              4             E        0   \n",
      "40                    Cul de Sac              5             F        0   \n",
      "41         Sigmoid Colon Deposit              5             F        0   \n",
      "42  Left Fallopian Tube Fimbriae              5             F        1   \n",
      "43                    Left Ovary              5             F        1   \n",
      "44                       Omentum              5             F        1   \n",
      "45          Right Fallopian Tube              5             F        0   \n",
      "46                   Right Ovary              5             F        0   \n",
      "47                        Adnexa              5             F        0   \n",
      "48                    Cul de Sac              6             G        0   \n",
      "49         Sigmoid Colon Deposit              6             G        0   \n",
      "50  Left Fallopian Tube Fimbriae              6             G        0   \n",
      "51                    Left Ovary              6             G        0   \n",
      "52                       Omentum              6             G        0   \n",
      "53          Right Fallopian Tube              6             G        0   \n",
      "54                   Right Ovary              6             G        0   \n",
      "55                        Adnexa              6             G        1   \n",
      "\n",
      "   site_category  num_mutations  \n",
      "0     metastasis              1  \n",
      "1     metastasis              1  \n",
      "2     metastasis              1  \n",
      "3        primary              1  \n",
      "4     metastasis              1  \n",
      "5     metastasis              1  \n",
      "6        primary              1  \n",
      "7     metastasis              1  \n",
      "8     metastasis              1  \n",
      "9     metastasis              1  \n",
      "10    metastasis              1  \n",
      "11       primary              1  \n",
      "12    metastasis              1  \n",
      "13    metastasis              1  \n",
      "14       primary              1  \n",
      "15    metastasis              1  \n",
      "16    metastasis              1  \n",
      "17    metastasis              1  \n",
      "18    metastasis              1  \n",
      "19       primary              1  \n",
      "20    metastasis              1  \n",
      "21    metastasis              1  \n",
      "22       primary              1  \n",
      "23    metastasis              1  \n",
      "24    metastasis              1  \n",
      "25    metastasis              1  \n",
      "26    metastasis              1  \n",
      "27       primary              1  \n",
      "28    metastasis              1  \n",
      "29    metastasis              1  \n",
      "30       primary              1  \n",
      "31    metastasis              1  \n",
      "32    metastasis              1  \n",
      "33    metastasis              1  \n",
      "34    metastasis              1  \n",
      "35       primary              1  \n",
      "36    metastasis              1  \n",
      "37    metastasis              1  \n",
      "38       primary              1  \n",
      "39    metastasis              1  \n",
      "40    metastasis              1  \n",
      "41    metastasis              1  \n",
      "42    metastasis              1  \n",
      "43       primary              1  \n",
      "44    metastasis              1  \n",
      "45    metastasis              1  \n",
      "46       primary              1  \n",
      "47    metastasis              1  \n",
      "48    metastasis              1  \n",
      "49    metastasis              1  \n",
      "50    metastasis              1  \n",
      "51       primary              1  \n",
      "52    metastasis              1  \n",
      "53    metastasis              1  \n",
      "54       primary              1  \n",
      "55    metastasis              1  \n",
      "['A' 'B' 'C' 'D' 'E' 'F' 'G' 'H'] ['Left Ovary Surface' 'Left Pelvic Sidewall' 'Right Ovary'\n",
      " 'Right Pelvic Sidewall']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B Left Ovary Surface\n",
      "B Left Pelvic Sidewall\n",
      "B Right Ovary\n",
      "B Right Pelvic Sidewall\n",
      "D Right Ovary\n",
      "E Right Ovary\n",
      "F Right Ovary\n",
      "G Right Ovary\n",
      "H Right Ovary\n",
      "    anatomical_site_label  cluster_index cluster_label  present site_category  \\\n",
      "0      Left Ovary Surface              0             A        0       primary   \n",
      "1    Left Pelvic Sidewall              0             A        0    metastasis   \n",
      "2             Right Ovary              0             A        0       primary   \n",
      "3   Right Pelvic Sidewall              0             A        0    metastasis   \n",
      "4      Left Ovary Surface              1             B        1       primary   \n",
      "5    Left Pelvic Sidewall              1             B        1    metastasis   \n",
      "6             Right Ovary              1             B        1       primary   \n",
      "7   Right Pelvic Sidewall              1             B        1    metastasis   \n",
      "8      Left Ovary Surface              2             C        0       primary   \n",
      "9    Left Pelvic Sidewall              2             C        0    metastasis   \n",
      "10            Right Ovary              2             C        0       primary   \n",
      "11  Right Pelvic Sidewall              2             C        0    metastasis   \n",
      "12     Left Ovary Surface              3             D        0       primary   \n",
      "13   Left Pelvic Sidewall              3             D        0    metastasis   \n",
      "14            Right Ovary              3             D        1       primary   \n",
      "15  Right Pelvic Sidewall              3             D        0    metastasis   \n",
      "16     Left Ovary Surface              4             E        0       primary   \n",
      "17   Left Pelvic Sidewall              4             E        0    metastasis   \n",
      "18            Right Ovary              4             E        1       primary   \n",
      "19  Right Pelvic Sidewall              4             E        0    metastasis   \n",
      "20     Left Ovary Surface              5             F        0       primary   \n",
      "21   Left Pelvic Sidewall              5             F        0    metastasis   \n",
      "22            Right Ovary              5             F        1       primary   \n",
      "23  Right Pelvic Sidewall              5             F        0    metastasis   \n",
      "24     Left Ovary Surface              6             G        0       primary   \n",
      "25   Left Pelvic Sidewall              6             G        0    metastasis   \n",
      "26            Right Ovary              6             G        1       primary   \n",
      "27  Right Pelvic Sidewall              6             G        0    metastasis   \n",
      "28     Left Ovary Surface              7             H        0       primary   \n",
      "29   Left Pelvic Sidewall              7             H        0    metastasis   \n",
      "30            Right Ovary              7             H        1       primary   \n",
      "31  Right Pelvic Sidewall              7             H        0    metastasis   \n",
      "\n",
      "    num_mutations  \n",
      "0               1  \n",
      "1               1  \n",
      "2               1  \n",
      "3               1  \n",
      "4               1  \n",
      "5               1  \n",
      "6               1  \n",
      "7               1  \n",
      "8               1  \n",
      "9               1  \n",
      "10              1  \n",
      "11              1  \n",
      "12              1  \n",
      "13              1  \n",
      "14              1  \n",
      "15              1  \n",
      "16              1  \n",
      "17              1  \n",
      "18              1  \n",
      "19              1  \n",
      "20              1  \n",
      "21              1  \n",
      "22              1  \n",
      "23              1  \n",
      "24              1  \n",
      "25              1  \n",
      "26              1  \n",
      "27              1  \n",
      "28              1  \n",
      "29              1  \n",
      "30              1  \n",
      "31              1  \n",
      "['A' 'B' 'C' 'D' 'E'] ['Brain Metastasis' 'Bowel Implant' 'Left Ovary' 'Right Ovary'\n",
      " 'Right Uterosacral' 'Righ Pelvic Mass']\n",
      "A Right Ovary\n",
      "B Brain Metastasis\n",
      "B Right Uterosacral\n",
      "C Left Ovary\n",
      "C Right Ovary\n",
      "C Right Uterosacral\n",
      "D Bowel Implant\n",
      "D Right Uterosacral\n",
      "D Righ Pelvic Mass\n",
      "E Left Ovary\n",
      "   anatomical_site_label  cluster_index cluster_label  present site_category  \\\n",
      "0       Brain Metastasis              0             A        0    metastasis   \n",
      "1          Bowel Implant              0             A        0    metastasis   \n",
      "2             Left Ovary              0             A        0       primary   \n",
      "3            Right Ovary              0             A        1       primary   \n",
      "4      Right Uterosacral              0             A        0       primary   \n",
      "5       Righ Pelvic Mass              0             A        0    metastasis   \n",
      "6       Brain Metastasis              1             B        1    metastasis   \n",
      "7          Bowel Implant              1             B        0    metastasis   \n",
      "8             Left Ovary              1             B        0       primary   \n",
      "9            Right Ovary              1             B        0       primary   \n",
      "10     Right Uterosacral              1             B        1       primary   \n",
      "11      Righ Pelvic Mass              1             B        0    metastasis   \n",
      "12      Brain Metastasis              2             C        0    metastasis   \n",
      "13         Bowel Implant              2             C        0    metastasis   \n",
      "14            Left Ovary              2             C        1       primary   \n",
      "15           Right Ovary              2             C        1       primary   \n",
      "16     Right Uterosacral              2             C        1       primary   \n",
      "17      Righ Pelvic Mass              2             C        0    metastasis   \n",
      "18      Brain Metastasis              3             D        0    metastasis   \n",
      "19         Bowel Implant              3             D        1    metastasis   \n",
      "20            Left Ovary              3             D        0       primary   \n",
      "21           Right Ovary              3             D        0       primary   \n",
      "22     Right Uterosacral              3             D        1       primary   \n",
      "23      Righ Pelvic Mass              3             D        1    metastasis   \n",
      "24      Brain Metastasis              4             E        0    metastasis   \n",
      "25         Bowel Implant              4             E        0    metastasis   \n",
      "26            Left Ovary              4             E        1       primary   \n",
      "27           Right Ovary              4             E        0       primary   \n",
      "28     Right Uterosacral              4             E        0       primary   \n",
      "29      Righ Pelvic Mass              4             E        0    metastasis   \n",
      "\n",
      "    num_mutations  \n",
      "0               1  \n",
      "1               1  \n",
      "2               1  \n",
      "3               1  \n",
      "4               1  \n",
      "5               1  \n",
      "6               1  \n",
      "7               1  \n",
      "8               1  \n",
      "9               1  \n",
      "10              1  \n",
      "11              1  \n",
      "12              1  \n",
      "13              1  \n",
      "14              1  \n",
      "15              1  \n",
      "16              1  \n",
      "17              1  \n",
      "18              1  \n",
      "19              1  \n",
      "20              1  \n",
      "21              1  \n",
      "22              1  \n",
      "23              1  \n",
      "24              1  \n",
      "25              1  \n",
      "26              1  \n",
      "27              1  \n",
      "28              1  \n",
      "29              1  \n",
      "['A' 'B' 'C'] ['Left Ovary' 'Omentum' 'Right Ovary']\n",
      "B Left Ovary\n",
      "C Left Ovary\n",
      "C Omentum\n",
      "C Right Ovary\n",
      "  anatomical_site_label  cluster_index cluster_label  present site_category  \\\n",
      "0            Left Ovary              0             A        0       primary   \n",
      "1               Omentum              0             A        0    metastasis   \n",
      "2           Right Ovary              0             A        0       primary   \n",
      "3            Left Ovary              1             B        1       primary   \n",
      "4               Omentum              1             B        0    metastasis   \n",
      "5           Right Ovary              1             B        0       primary   \n",
      "6            Left Ovary              2             C        1       primary   \n",
      "7               Omentum              2             C        1    metastasis   \n",
      "8           Right Ovary              2             C        1       primary   \n",
      "\n",
      "   num_mutations  \n",
      "0              1  \n",
      "1              1  \n",
      "2              1  \n",
      "3              1  \n",
      "4              1  \n",
      "5              1  \n",
      "6              1  \n",
      "7              1  \n",
      "8              1  \n",
      "['A' 'B' 'C' 'D' 'E' 'F'] ['Left Fallopian Tube' 'Omentum' 'Right Ovary']\n",
      "A Left Fallopian Tube\n",
      "A Omentum\n",
      "B Left Fallopian Tube\n",
      "B Right Ovary\n",
      "D Right Ovary\n",
      "E Omentum\n",
      "E Right Ovary\n",
      "F Omentum\n",
      "F Right Ovary\n",
      "   anatomical_site_label  cluster_index cluster_label  present site_category  \\\n",
      "0    Left Fallopian Tube              0             A        1    metastasis   \n",
      "1                Omentum              0             A        1    metastasis   \n",
      "2            Right Ovary              0             A        0       primary   \n",
      "3    Left Fallopian Tube              1             B        1    metastasis   \n",
      "4                Omentum              1             B        0    metastasis   \n",
      "5            Right Ovary              1             B        1       primary   \n",
      "6    Left Fallopian Tube              2             C        0    metastasis   \n",
      "7                Omentum              2             C        0    metastasis   \n",
      "8            Right Ovary              2             C        0       primary   \n",
      "9    Left Fallopian Tube              3             D        0    metastasis   \n",
      "10               Omentum              3             D        0    metastasis   \n",
      "11           Right Ovary              3             D        1       primary   \n",
      "12   Left Fallopian Tube              4             E        0    metastasis   \n",
      "13               Omentum              4             E        1    metastasis   \n",
      "14           Right Ovary              4             E        1       primary   \n",
      "15   Left Fallopian Tube              5             F        0    metastasis   \n",
      "16               Omentum              5             F        1    metastasis   \n",
      "17           Right Ovary              5             F        1       primary   \n",
      "\n",
      "    num_mutations  \n",
      "0               1  \n",
      "1               1  \n",
      "2               1  \n",
      "3               1  \n",
      "4               1  \n",
      "5               1  \n",
      "6               1  \n",
      "7               1  \n",
      "8               1  \n",
      "9               1  \n",
      "10              1  \n",
      "11              1  \n",
      "12              1  \n",
      "13              1  \n",
      "14              1  \n",
      "15              1  \n",
      "16              1  \n",
      "17              1  \n"
     ]
    }
   ],
   "source": [
    "cluster_labels = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I']\n",
    "\n",
    "def get_combined_prevalence(pid, clone_ids, sample):\n",
    "    total_prevalence = 0.0\n",
    "    for clone_id in clone_ids:\n",
    "        subset_df = prevalences_df[(prevalences_df['patient_id']==pid)&(prevalences_df['clone_id']==clone_id)&(prevalences_df['paper_id']==sample)]\n",
    "        if len(subset_df) > 0:\n",
    "            prevalence = subset_df['prevalence'].item()\n",
    "            total_prevalence += float(prevalence)\n",
    "    return total_prevalence\n",
    "\n",
    "def get_presence_in_site(tree, path_matrix, pid, clone_id, site):\n",
    "    pid_clone_df = prevalences_df[(prevalences_df['patient_id']==pid)&(prevalences_df['clone_id']==clone_id)]\n",
    "    site_samples = sample_df[sample_df['anatomical_site_label']==site]['paper_id'].unique()\n",
    "    \n",
    "    descendants = [cluster_labels[i[0].item()] for i in path_matrix[cluster_labels.index(clone_id)].nonzero()]\n",
    "    present_in_site = False\n",
    "    for sample in site_samples:\n",
    "        clone_prevalence = get_combined_prevalence(pid, [clone_id], sample)\n",
    "        descendants_prevalence = get_combined_prevalence(pid, [clone_id]+descendants, sample)\n",
    "#         print(clone_id, descendants, sample, clone_prevalence, descendants_prevalence)\n",
    "        if clone_prevalence > 0.01 and clone_prevalence > 0.5*descendants_prevalence:\n",
    "            present_in_site = True\n",
    "    return present_in_site\n",
    "\n",
    "def get_parent(tree, node_index):\n",
    "    # Look for a '1' in the column corresponding to the node_index\n",
    "    parent_index = (tree[:, node_index] == 1).nonzero(as_tuple=True)[0]\n",
    "    return parent_index\n",
    "\n",
    "def get_num_mutations(tree, clone_id)\n",
    "\n",
    "cols = ['anatomical_site_label', 'cluster_index', 'cluster_label', 'present', 'site_category', 'num_mutations']\n",
    "for pid in patient_ids:\n",
    "    tree = dutil.get_adjacency_matrix_from_txt_edge_list(os.path.join(data_dir, f\"patient{pid}_tree.txt\"))\n",
    "    path_matrix = vutil.get_path_matrix(tree, remove_self_loops=True)\n",
    "    clones = cellular_prevalences_df[cellular_prevalences_df['patient_id']==pid]['clone_id'].unique()\n",
    "    unique_sites = sample_df[sample_df['patient_id']==pid]['anatomical_site_label'].unique()\n",
    "    print(clones, unique_sites)\n",
    "    data = []\n",
    "    for clone in clones:\n",
    "        for site in unique_sites:\n",
    "            present_in_site = get_presence_in_site(tree, path_matrix, pid, clone, site)\n",
    "            if present_in_site:\n",
    "                print(clone, site)\n",
    "            site_category = 'primary' if 'Ovary' in site or 'Uterosacral' in site else 'metastasis'\n",
    "            data.append([site, cluster_labels.index(clone), clone, 1 if present_in_site else 0, site_category, 1])\n",
    "    # Build a dataframe for each patient w/ the information in cols\n",
    "    df = pd.DataFrame(data, columns=cols)\n",
    "    print(df)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "69008f87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>paper_id</th>\n",
       "      <th>clone_id</th>\n",
       "      <th>prevalence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ApC1</td>\n",
       "      <td>A</td>\n",
       "      <td>0.055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>LFTB4</td>\n",
       "      <td>A</td>\n",
       "      <td>0.543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1</td>\n",
       "      <td>LOvB2</td>\n",
       "      <td>A</td>\n",
       "      <td>0.300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1</td>\n",
       "      <td>Om1</td>\n",
       "      <td>A</td>\n",
       "      <td>0.024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>RFTA16</td>\n",
       "      <td>A</td>\n",
       "      <td>0.986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1</td>\n",
       "      <td>ROv1</td>\n",
       "      <td>A</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>1</td>\n",
       "      <td>ROv2</td>\n",
       "      <td>A</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>1</td>\n",
       "      <td>ROv3</td>\n",
       "      <td>A</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>1</td>\n",
       "      <td>ROv4</td>\n",
       "      <td>A</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>1</td>\n",
       "      <td>ROvA4</td>\n",
       "      <td>A</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>1</td>\n",
       "      <td>SBwl</td>\n",
       "      <td>A</td>\n",
       "      <td>0.044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1</td>\n",
       "      <td>SBwlE4</td>\n",
       "      <td>A</td>\n",
       "      <td>0.324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    patient_id paper_id clone_id  prevalence\n",
       "0            1     ApC1        A       0.055\n",
       "9            1    LFTB4        A       0.543\n",
       "18           1    LOvB2        A       0.300\n",
       "27           1      Om1        A       0.024\n",
       "36           1   RFTA16        A       0.986\n",
       "45           1     ROv1        A       0.004\n",
       "54           1     ROv2        A       0.002\n",
       "63           1     ROv3        A       0.002\n",
       "72           1     ROv4        A       0.002\n",
       "81           1    ROvA4        A       0.002\n",
       "90           1     SBwl        A       0.044\n",
       "99           1   SBwlE4        A       0.324"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prevalences_df[(prevalences_df['patient_id']==1)&(prevalences_df['clone_id']=='A')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4f5763dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>coord</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>primer_set</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>ci_length</th>\n",
       "      <th>ml_loss</th>\n",
       "      <th>ml_origin</th>\n",
       "      <th>ml_presence</th>\n",
       "      <th>deletion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5516</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>10</td>\n",
       "      <td>101648657</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>7</td>\n",
       "      <td>0.996462</td>\n",
       "      <td>0.003303</td>\n",
       "      <td>0.009970</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5517</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>10</td>\n",
       "      <td>123969955</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>7</td>\n",
       "      <td>0.989874</td>\n",
       "      <td>0.049450</td>\n",
       "      <td>0.018570</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5518</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>10</td>\n",
       "      <td>127350427</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>7</td>\n",
       "      <td>0.965689</td>\n",
       "      <td>0.095702</td>\n",
       "      <td>0.263609</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5519</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>10</td>\n",
       "      <td>31815848</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>7</td>\n",
       "      <td>0.995950</td>\n",
       "      <td>0.014094</td>\n",
       "      <td>0.010017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5520</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>11</td>\n",
       "      <td>247341</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>7</td>\n",
       "      <td>0.996430</td>\n",
       "      <td>0.004301</td>\n",
       "      <td>0.009970</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6495</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>6</td>\n",
       "      <td>53142417</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>10_b_1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.408940</td>\n",
       "      <td>0.114234</td>\n",
       "      <td>0.385519</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6496</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_1</td>\n",
       "      <td>3</td>\n",
       "      <td>142681930</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>8</td>\n",
       "      <td>0.814847</td>\n",
       "      <td>0.231407</td>\n",
       "      <td>0.654302</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6497</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_3</td>\n",
       "      <td>3</td>\n",
       "      <td>142681930</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>8</td>\n",
       "      <td>0.685740</td>\n",
       "      <td>0.355491</td>\n",
       "      <td>0.986055</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6498</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_2</td>\n",
       "      <td>3</td>\n",
       "      <td>142681930</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>8</td>\n",
       "      <td>0.650817</td>\n",
       "      <td>0.371326</td>\n",
       "      <td>0.967168</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499</th>\n",
       "      <td>10</td>\n",
       "      <td>right_ovary_site_4</td>\n",
       "      <td>3</td>\n",
       "      <td>142681930</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>amplicrazy</td>\n",
       "      <td>8</td>\n",
       "      <td>0.771171</td>\n",
       "      <td>0.292289</td>\n",
       "      <td>0.905343</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>984 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      patient_id           sample_id chrom      coord ref alt  primer_set  \\\n",
       "5516          10  right_ovary_site_1    10  101648657   C   A  amplicrazy   \n",
       "5517          10  right_ovary_site_1    10  123969955   C   T  amplicrazy   \n",
       "5518          10  right_ovary_site_1    10  127350427   T   C  amplicrazy   \n",
       "5519          10  right_ovary_site_1    10   31815848   G   A  amplicrazy   \n",
       "5520          10  right_ovary_site_1    11     247341   G   A  amplicrazy   \n",
       "...          ...                 ...   ...        ...  ..  ..         ...   \n",
       "6495          10  right_ovary_site_4     6   53142417   T   C      10_b_1   \n",
       "6496          10  right_ovary_site_1     3  142681930   C   T  amplicrazy   \n",
       "6497          10  right_ovary_site_3     3  142681930   C   T  amplicrazy   \n",
       "6498          10  right_ovary_site_2     3  142681930   C   T  amplicrazy   \n",
       "6499          10  right_ovary_site_4     3  142681930   C   T  amplicrazy   \n",
       "\n",
       "      cluster_id      mean       std  ci_length  ml_loss  ml_origin  \\\n",
       "5516           7  0.996462  0.003303   0.009970        0          0   \n",
       "5517           7  0.989874  0.049450   0.018570        0          0   \n",
       "5518           7  0.965689  0.095702   0.263609        0          0   \n",
       "5519           7  0.995950  0.014094   0.010017        0          0   \n",
       "5520           7  0.996430  0.004301   0.009970        0          0   \n",
       "...          ...       ...       ...        ...      ...        ...   \n",
       "6495          13  0.408940  0.114234   0.385519        0          1   \n",
       "6496           8  0.814847  0.231407   0.654302        0          0   \n",
       "6497           8  0.685740  0.355491   0.986055        0          0   \n",
       "6498           8  0.650817  0.371326   0.967168        0          0   \n",
       "6499           8  0.771171  0.292289   0.905343        0          0   \n",
       "\n",
       "      ml_presence  deletion  \n",
       "5516            1         0  \n",
       "5517            1         0  \n",
       "5518            1         0  \n",
       "5519            1         0  \n",
       "5520            1         0  \n",
       "...           ...       ...  \n",
       "6495            1         0  \n",
       "6496            1         0  \n",
       "6497            1         0  \n",
       "6498            1         0  \n",
       "6499            1         0  \n",
       "\n",
       "[984 rows x 15 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyclone_cluster_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_16.csv'))\n",
    "pyclone_cluster_df[pyclone_cluster_df['patient_id']==10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9aa47046",
   "metadata": {},
   "outputs": [],
   "source": [
    "pyclone_cluster_df\n",
    "def get_clone_id(row):\n",
    "    pid = row['patient_id']\n",
    "    pyclone_cluster_id = row['cluster_id']\n",
    "    pt_cluster_df =  pyclone_to_clone_id_df[(pyclone_to_clone_id_df['pyclone_cluster_id']==pyclone_cluster_id)&(pyclone_to_clone_id_df['patient_id']==pid)]\n",
    "    if len(pt_cluster_df) != 0:\n",
    "        return pt_cluster_df['clone_id'].unique()\n",
    "    return None\n",
    "pyclone_cluster_df['clone_id'] = pyclone_cluster_df.apply(lambda row:get_clone_id(row), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ecb03acb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>clone_id</th>\n",
       "      <th>pyclone_cluster_id</th>\n",
       "      <th>present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>10</td>\n",
       "      <td>B</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>10</td>\n",
       "      <td>C</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>10</td>\n",
       "      <td>D</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>10</td>\n",
       "      <td>E</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>10</td>\n",
       "      <td>F</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     patient_id clone_id  pyclone_cluster_id  present\n",
       "343          10        A                   7        1\n",
       "349          10        B                   7        1\n",
       "355          10        C                   7        1\n",
       "361          10        D                   7        1\n",
       "367          10        E                   7        1\n",
       "373          10        F                   7        1"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pyclone_to_clone_id_df[(pyclone_to_clone_id_df['pyclone_cluster_id']==7)&(pyclone_to_clone_id_df['patient_id']==10)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b195c6ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['left_fallopian_tube_site_b2', 'normal_blood', 'omentum_site_c1',\n",
       "       'right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3',\n",
       "       'right_ovary_site_4', 'right_ovary_site_a4', 'right_ovary_site_a9'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads_df[reads_df['patient_id']==10]['sample_id'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ab653e9",
   "metadata": {},
   "source": [
    "## (2) Generate inputs for PyClone clustering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b201a3fd",
   "metadata": {},
   "source": [
    "### (2a) Create input for PyClone clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "376d6f8b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        patient_id           sample_id chrom        start          end  major  \\\n",
      "0                1  small_bowel_site_1     1          1.0      71283.0     16   \n",
      "1                1  small_bowel_site_1     1      73560.0      94700.0      7   \n",
      "2                1  small_bowel_site_1     1     101640.0     110479.0     13   \n",
      "3                1  small_bowel_site_1     1     110479.0     141357.0     22   \n",
      "4                1  small_bowel_site_1     1     144162.0     247926.0     24   \n",
      "...            ...                 ...   ...          ...          ...    ...   \n",
      "135154          10  right_ovary_site_4     X  153000000.0  153000000.0      0   \n",
      "135155          10  right_ovary_site_4     X  153000000.0  154000000.0      2   \n",
      "135156          10  right_ovary_site_4     X  154000000.0  155000000.0      2   \n",
      "135157          10  right_ovary_site_4     X  155000000.0  155000000.0      2   \n",
      "135158          10  right_ovary_site_4     X  155000000.0  155000000.0      2   \n",
      "\n",
      "        minor  major_sub  minor_sub  subclonal  \n",
      "0           7          0          0        NaN  \n",
      "1           4          0          0        NaN  \n",
      "2           9          0          0        NaN  \n",
      "3          14          0          0        NaN  \n",
      "4          24          0          0        NaN  \n",
      "...       ...        ...        ...        ...  \n",
      "135154      0          0          0        NaN  \n",
      "135155      0          0          0       0.22  \n",
      "135156      0          1          0       0.99  \n",
      "135157      0          0          0        NaN  \n",
      "135158      0          1          0       0.97  \n",
      "\n",
      "[135159 rows x 10 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# get CNA info from supplement of McPherson et. al. this is only available for the discovery samples\n",
    "cna_df = pd.read_csv(os.path.join(data_dir, 'supplement_table_7.csv'))\n",
    "print(cna_df)\n",
    "\n",
    "def subset_cna(row):\n",
    "    chrom = row['chrom']\n",
    "    coord = row['coord']\n",
    "    patient_id = row['patient_id']\n",
    "    sample_id = row['sample_id']\n",
    "    df = cna_df[(cna_df['patient_id'] == patient_id) & (cna_df['sample_id'] == sample_id) & (cna_df['chrom'] == chrom) & (cna_df['start'] <= coord) & (cna_df['end'] >= coord)]\n",
    "    return df\n",
    "\n",
    "def major_cn(row):\n",
    "    coord_df = subset_cna(row)\n",
    "    if len(coord_df) != 1:\n",
    "        #print(\"Not found in CNA:\\n\", row['chrom'], row['coord'])\n",
    "        return 0\n",
    "    return int(coord_df['major'])\n",
    "\n",
    "def minor_cn(row):\n",
    "    coord_df = subset_cna(row)\n",
    "    if len(coord_df) != 1:\n",
    "        return 0\n",
    "    return int(coord_df['minor'])\n",
    "\n",
    "\n",
    "# Add CNAs to reads_df\n",
    "cna_reads_df = reads_df.copy(deep=True)\n",
    "cna_reads_df['major_cn'] =  cna_reads_df.apply(lambda row: major_cn(row), axis=1)\n",
    "cna_reads_df['minor_cn'] =  cna_reads_df.apply(lambda row: minor_cn(row), axis=1)\n",
    "cna_reads_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebd57557",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do some pre-processing to remove any mutations where CNA info is not available or\n",
    "# the major allele CN is 0 (if the malignant cells have no copies of the region \n",
    "# overlapping the mutation, the mutation cannot exist.) or minor allele CN > major allele CN\n",
    "cna_reads_df = cna_reads_df[cna_reads_df['major_cn'] != 0]\n",
    "cna_reads_df = cna_reads_df[(cna_reads_df['minor_cn'] <= cna_reads_df['major_cn'])]\n",
    "cna_reads_df['normal_cn'] = 2 # copy number of the locus in normal cells is 2 (no male chromosomes here)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5be21d1",
   "metadata": {},
   "source": [
    "## (4a) Take PyClone generated clusters and create csvs with ref and var counts pooled by cluster "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9d3ac3fe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient 1 13 clusters 177 mutations 6 samples\n",
      "['omentum_site_1', 'right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3', 'right_ovary_site_4', 'small_bowel_site_1']\n",
      "['omentum_site', 'small_bowel_site', 'right_ovary_site']\n",
      "patient 2 6 clusters 120 mutations 4 samples\n",
      "['omentum_site_1', 'omentum_site_2', 'right_ovary_site_1', 'right_ovary_site_2']\n",
      "['omentum_site', 'right_ovary_site']\n",
      "patient 3 9 clusters 168 mutations 4 samples\n",
      "['adnexa_site_1', 'omentum_site_1', 'right_ovary_site_1', 'right_ovary_site_2']\n",
      "['omentum_site', 'adnexa_site', 'right_ovary_site']\n",
      "patient 4 10 clusters 161 mutations 5 samples\n",
      "['left_pelvic_sidewall_site_1', 'right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3', 'right_ovary_site_4']\n",
      "['right_ovary_site', 'left_pelvic_sidewall_site']\n",
      "patient 7 8 clusters 221 mutations 3 samples\n",
      "['brain_metastasis', 'left_ovary_site_1', 'righ_pelvic_mass']\n",
      "['righ_pelvic_mass', 'left_ovary_site', 'brain_metastasis']\n",
      "patient 9 5 clusters 132 mutations 5 samples\n",
      "['left_ovary_site_1', 'left_ovary_site_2', 'omentum_site_1', 'omentum_site_2', 'right_ovary_site_1']\n",
      "['omentum_site', 'left_ovary_site', 'right_ovary_site']\n",
      "patient 10 3 clusters 189 mutations 4 samples\n",
      "['right_ovary_site_1', 'right_ovary_site_2', 'right_ovary_site_3', 'right_ovary_site_4']\n",
      "['right_ovary_site']\n"
     ]
    }
   ],
   "source": [
    "# Need a tsv for each patient with ['#sample_index', 'sample_label', 'anatomical_site_index','anatomical_site_label', 'character_index', 'character_label', 'ref', 'var']\n",
    "import re\n",
    "import numpy as np\n",
    "from metient.util import data_extraction_util as dutil\n",
    "pyclone_dir = os.path.join(data_dir, \"pyclone_analysis\")\n",
    "\n",
    "\n",
    "cols = [ 'anatomical_site_index','anatomical_site_label',  'cluster_index', 'character_index', 'character_label', 'ref', 'var', 'var_read_prob', 'site_category']\n",
    "#agg_rules = {'sample_label': lambda x: ';'.join(set(x))}\n",
    "\n",
    "for patient_id in patient_ids:\n",
    "    patient_pyclone_dir = os.path.join(pyclone_dir, f\"patient_{patient_id}\", \"tables\", f\"loci.tsv\")\n",
    "    mut_name_to_clstr_id, clstr_id_to_name, mutation_names = dutil.get_mut_to_cluster_map_from_pyclone_output(patient_pyclone_dir, min_mut_thres=MIN_MUT_THRES)\n",
    "  \n",
    "    patient_subset = cna_reads_df[cna_reads_df['patient_id'] == patient_id]\n",
    "    discovery_samples = discovery_samples_df[discovery_samples_df['patient_id']==patient_id]['sample_id']\n",
    "    patient_subset = patient_subset[(patient_subset['sample_id'].isin(discovery_samples)) & (patient_subset['sample_id'] != 'normal_blood')]   \n",
    "    sample_names = list(patient_subset['sample_id'].unique())\n",
    "    print(\"patient\", patient_id, len(clstr_id_to_name), \"clusters\", len(mutation_names), \"mutations\", len(sample_names), \"samples\")\n",
    "    print(sample_names)\n",
    "    data = []\n",
    "    \n",
    "    anat_sites = list(set([re.sub(r'_[0-9]+$', '', sample) for sample in sample_names]))\n",
    "    print(anat_sites)\n",
    "    for midx, mut in enumerate(mutation_names):\n",
    "        mut_patient_subset = patient_subset[patient_subset['character_label'] == mut]\n",
    "        for sidx, sample in enumerate(sample_names):\n",
    "            mut_patient_sample = mut_patient_subset[mut_patient_subset['sample_id'] == sample]\n",
    "            var = mut_patient_sample['alt_counts'].values[0]\n",
    "            ref = mut_patient_sample['ref_counts'].values[0]\n",
    "            tumour_prop_subset = discovery_samples_df[(discovery_samples_df['patient_id']==patient_id) & (discovery_samples_df['sample_id']==sample)]\n",
    "            p = tumour_prop_subset['tumour_cell_proportion'].item() \n",
    "            major_cn = mut_patient_sample['major_cn'].values[0]\n",
    "            minor_cn = mut_patient_sample['minor_cn'].values[0]\n",
    "            var_read_prob = dutil.calc_var_read_prob(major_cn, minor_cn, p)\n",
    "            \n",
    "            site_label = re.sub(r'_[0-9]+$', '', sample)\n",
    "            #print(sample, site_label, anat_sites.index(site_label))\n",
    "            category = 'primary' if 'ovary_site' in site_label else 'metastasis'\n",
    "            data.append([anat_sites.index(site_label), site_label, mut_name_to_clstr_id[mut], midx, mut.split(\":\")[0], ref, var, var_read_prob, category])\n",
    "            \n",
    "    patient_df = pd.DataFrame(data, columns=cols)\n",
    "    patient_df.to_csv(os.path.join(data_dir,\"pyclone_clustered_tsvs\", f\"{patient_id}_SNVs.tsv\"), sep=\"\\t\", index=False)\n",
    "\n",
    "#     dutil.write_pooled_tsv_from_clusters(patient_df, mut_name_to_clstr_id, clstr_id_to_name, agg_rules, \n",
    "#                                          os.path.join(data_dir,\"pyclone_clustered_tsvs\"), f\"patient{patient_id}\", \";\", \":\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4962e0c5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:met]",
   "language": "python",
   "name": "conda-env-met-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
